--- a/net/minecraft/entity/projectile/EvokerFangsEntity.java
+++ b/net/minecraft/entity/projectile/EvokerFangsEntity.java
@@ -15,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class EvokerFangsEntity extends Entity {
    private int warmupDelayTicks;
@@ -92,7 +_,9 @@
       } else if (--this.warmupDelayTicks < 0) {
          if (this.warmupDelayTicks == -8) {
             for(LivingEntity livingentity : this.level.getEntitiesOfClass(LivingEntity.class, this.getBoundingBox().inflate(0.2D, 0.0D, 0.2D))) {
+               CraftEventFactory.entityDamage = this; // CraftBukkit
                this.dealDamageTo(livingentity);
+               CraftEventFactory.entityDamage = null; // CraftBukkit
             }
          }
 
