--- a/net/minecraft/entity/projectile/EyeOfEnderEntity.java
+++ b/net/minecraft/entity/projectile/EyeOfEnderEntity.java
@@ -28,11 +_,11 @@
 )
 public class EyeOfEnderEntity extends Entity implements IRendersAsItem {
    private static final DataParameter<ItemStack> DATA_ITEM_STACK = EntityDataManager.defineId(EyeOfEnderEntity.class, DataSerializers.ITEM_STACK);
-   private double tx;
-   private double ty;
-   private double tz;
-   private int life;
-   private boolean surviveAfterDeath;
+   public double tx;
+   public double ty;
+   public double tz;
+   public int life;
+   public boolean surviveAfterDeath;
 
    public EyeOfEnderEntity(EntityType<? extends EyeOfEnderEntity> p_i50169_1_, World p_i50169_2_) {
       super(p_i50169_1_, p_i50169_2_);
@@ -45,7 +_,7 @@
    }
 
    public void setItem(ItemStack p_213863_1_) {
-      if (p_213863_1_.getItem() != Items.ENDER_EYE || p_213863_1_.hasTag()) {
+      if (true || p_213863_1_.getItem() != Items.ENDER_EYE || p_213863_1_.hasTag()) { // CraftBukkit - always allow item changing
          this.getEntityData().set(DATA_ITEM_STACK, Util.make(p_213863_1_.copy(), (p_213862_0_) -> {
             p_213862_0_.setCount(1);
          }));
@@ -174,7 +_,7 @@
 
    public void readAdditionalSaveData(CompoundNBT p_70037_1_) {
       ItemStack itemstack = ItemStack.of(p_70037_1_.getCompound("Item"));
-      this.setItem(itemstack);
+      if (!itemstack.isEmpty()) this.setItem(itemstack); // CraftBukkit - SPIGOT-6103 summon, see also SPIGOT-5474
    }
 
    public float getBrightness() {
