--- a/net/minecraft/entity/projectile/FireballEntity.java
+++ b/net/minecraft/entity/projectile/FireballEntity.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class FireballEntity extends AbstractFireballEntity {
    public int explosionPower = 1;
@@ -23,17 +_,26 @@
    @OnlyIn(Dist.CLIENT)
    public FireballEntity(World p_i1768_1_, double p_i1768_2_, double p_i1768_4_, double p_i1768_6_, double p_i1768_8_, double p_i1768_10_, double p_i1768_12_) {
       super(EntityType.FIREBALL, p_i1768_2_, p_i1768_4_, p_i1768_6_, p_i1768_8_, p_i1768_10_, p_i1768_12_, p_i1768_1_);
+      isIncendiary = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING); // CraftBukkit
    }
 
    public FireballEntity(World p_i1769_1_, LivingEntity p_i1769_2_, double p_i1769_3_, double p_i1769_5_, double p_i1769_7_) {
       super(EntityType.FIREBALL, p_i1769_2_, p_i1769_3_, p_i1769_5_, p_i1769_7_, p_i1769_1_);
+      isIncendiary = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING); // CraftBukkit
    }
 
    protected void onHit(RayTraceResult p_70227_1_) {
       super.onHit(p_70227_1_);
       if (!this.level.isClientSide) {
-         boolean flag = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
-         this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), (float)this.explosionPower, flag, flag ? Explosion.Mode.DESTROY : Explosion.Mode.NONE);
+         boolean flag = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level, this.getOwner());
+         // CraftBukkit start - fire ExplosionPrimeEvent
+         ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+         this.level.getCBServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            // give 'this' instead of (Entity) null so we know what causes the damage
+            this.level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), flag ? Explosion.Mode.DESTROY : Explosion.Mode.NONE);
+         }
+         // CraftBukkit end
          this.remove();
       }
 
@@ -60,7 +_,8 @@
    public void readAdditionalSaveData(CompoundNBT p_70037_1_) {
       super.readAdditionalSaveData(p_70037_1_);
       if (p_70037_1_.contains("ExplosionPower", 99)) {
-         this.explosionPower = p_70037_1_.getInt("ExplosionPower");
+         // CraftBukkit - set bukkitYield when setting explosionPower
+         bukkitYield = this.explosionPower = p_70037_1_.getInt("ExplosionPower");
       }
 
    }
