--- a/net/minecraft/entity/projectile/PotionEntity.java
+++ b/net/minecraft/entity/projectile/PotionEntity.java
@@ -1,9 +_,12 @@
 package net.minecraft.entity.projectile;
 
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
 import java.util.function.Predicate;
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
 import net.minecraft.block.CampfireBlock;
 import net.minecraft.entity.AreaEffectCloudEntity;
 import net.minecraft.entity.Entity;
@@ -11,6 +_,7 @@
 import net.minecraft.entity.IRendersAsItem;
 import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.Items;
@@ -31,6 +_,9 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 @OnlyIn(
    value = Dist.CLIENT,
@@ -90,7 +_,7 @@
          boolean flag = potion == Potions.WATER && list.isEmpty();
          if (flag) {
             this.applyWater();
-         } else if (!list.isEmpty()) {
+         } else if (true || !list.isEmpty()) { // CraftBukkit - Call event even if no effects to apply
             if (this.isLingering()) {
                this.makeAreaOfEffectCloud(itemstack, potion);
             } else {
@@ -121,6 +_,8 @@
    private void applySplash(List<EffectInstance> p_213888_1_, @Nullable Entity p_213888_2_) {
       AxisAlignedBB axisalignedbb = this.getBoundingBox().inflate(4.0D, 2.0D, 4.0D);
       List<LivingEntity> list = this.level.getEntitiesOfClass(LivingEntity.class, axisalignedbb);
+      Map<org.bukkit.entity.LivingEntity, Double> affected = new HashMap<>(); // CraftBukkit
+
       if (!list.isEmpty()) {
          for(LivingEntity livingentity : list) {
             if (livingentity.isAffectedByPotions()) {
@@ -131,23 +_,49 @@
                      d1 = 1.0D;
                   }
 
-                  for(EffectInstance effectinstance : p_213888_1_) {
-                     Effect effect = effectinstance.getEffect();
-                     if (effect.isInstantenous()) {
-                        effect.applyInstantenousEffect(this, this.getOwner(), livingentity, effectinstance.getAmplifier(), d1);
-                     } else {
-                        int i = (int)(d1 * (double)effectinstance.getDuration() + 0.5D);
-                        if (i > 20) {
-                           livingentity.addEffect(new EffectInstance(effect, i, effectinstance.getAmplifier(), effectinstance.isAmbient(), effectinstance.isVisible()));
-                        }
-                     }
-                  }
-               }
-            }
-         }
-      }
-
+                  // CraftBukkit start
+                  affected.put((org.bukkit.entity.LivingEntity) livingentity.getBukkitEntity(), d1);
+               }
+            }
+         }
+      }
+
+      org.bukkit.event.entity.PotionSplashEvent event = CraftEventFactory.callPotionSplashEvent(this, affected);
+      if (!event.isCancelled() && list != null && !list.isEmpty()) { // do not process effects if there are no effects to process
+         for (org.bukkit.entity.LivingEntity victim : event.getAffectedEntities()) {
+            if (!(victim instanceof CraftLivingEntity)) {
+               continue;
+            }
+
+            LivingEntity livingentity = ((CraftLivingEntity) victim).getHandle();
+            double dl = event.getIntensity(victim);
+            // CraftBukkit end
+
+            for (EffectInstance effectinstance : p_213888_1_) {
+               Effect effect = effectinstance.getEffect();
+               // CraftBukkit start - Abide by PVP settings - for players only!
+               if (!this.level.pvpMode && this.getOwner() instanceof ServerPlayerEntity && livingentity instanceof ServerPlayerEntity && livingentity != this.getOwner()) {
+                  int i = Effect.getId(effect);
+                  // Block SLOWER_MOVEMENT, SLOWER_DIG, HARM, BLINDNESS, HUNGER, WEAKNESS and POISON potions
+                  if (i == 2 || i == 4 || i == 7 || i == 15 || i == 17 || i == 18 || i == 19) {
+                     continue;
+                  }
+               }
+               // CraftBukkit end
+
+               if (effect.isInstantenous()) {
+                  effect.applyInstantenousEffect(this, this.getOwner(), livingentity, effectinstance.getAmplifier(), dl);
+               } else {
+                  int i = (int) (dl * (double) effectinstance.getDuration() + 0.5D);
+                  if (i > 20) {
+                     livingentity.addEffect(new EffectInstance(effect, i, effectinstance.getAmplifier(), effectinstance.isAmbient(), effectinstance.isVisible()), Cause.POTION_SPLASH); // CraftBukkit
+                  }
+               }
+            }
+         }
+      }
    }
+
 
    private void makeAreaOfEffectCloud(ItemStack p_190542_1_, Potion p_190542_2_) {
       AreaEffectCloudEntity areaeffectcloudentity = new AreaEffectCloudEntity(this.level, this.getX(), this.getY(), this.getZ());
@@ -171,7 +_,14 @@
          areaeffectcloudentity.setFixedColor(compoundnbt.getInt("CustomPotionColor"));
       }
 
-      this.level.addFreshEntity(areaeffectcloudentity);
+      // CraftBukkit start
+      org.bukkit.event.entity.LingeringPotionSplashEvent event = CraftEventFactory.callLingeringPotionSplashEvent(this, areaeffectcloudentity);
+      if (!(event.isCancelled() || areaeffectcloudentity.removed)) {
+         this.level.addFreshEntity(areaeffectcloudentity);
+      } else {
+         areaeffectcloudentity.removed = true;
+      }
+      // CraftBukkit end
    }
 
    private boolean isLingering() {
@@ -181,11 +_,19 @@
    private void dowseFire(BlockPos p_184542_1_, Direction p_184542_2_) {
       BlockState blockstate = this.level.getBlockState(p_184542_1_);
       if (blockstate.is(BlockTags.FIRE)) {
-         this.level.removeBlock(p_184542_1_, false);
+         // CraftBukkit start
+         if (!CraftEventFactory.callEntityChangeBlockEvent(this, p_184542_1_, Blocks.AIR.defaultBlockState()).isCancelled()) {
+            this.level.removeBlock(p_184542_1_, false);
+         }
+         // CraftBukkit end
       } else if (CampfireBlock.isLitCampfire(blockstate)) {
-         this.level.levelEvent((PlayerEntity)null, 1009, p_184542_1_, 0);
-         CampfireBlock.dowse(this.level, p_184542_1_, blockstate);
-         this.level.setBlockAndUpdate(p_184542_1_, blockstate.setValue(CampfireBlock.LIT, Boolean.valueOf(false)));
+         // CraftBukkit start
+         if (!CraftEventFactory.callEntityChangeBlockEvent(this, p_184542_1_, blockstate.setValue(CampfireBlock.LIT, false)).isCancelled()) { // PAIL rename LIT
+            this.level.levelEvent((PlayerEntity) null, 1009, p_184542_1_, 0);
+            CampfireBlock.dowse(this.level, p_184542_1_, blockstate);
+            this.level.setBlockAndUpdate(p_184542_1_, blockstate.setValue(CampfireBlock.LIT, false));
+         }
+         // CraftBukkit end
       }
 
    }
