--- a/net/minecraft/entity/projectile/ProjectileEntity.java
+++ b/net/minecraft/entity/projectile/ProjectileEntity.java
@@ -2,6 +_,8 @@
 
 import java.util.UUID;
 import javax.annotation.Nullable;
+
+import com.ibm.icu.text.RawCollationKey;
 import net.minecraft.block.BlockState;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
@@ -15,12 +_,17 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.projectiles.ProjectileSource;
 
 public abstract class ProjectileEntity extends Entity {
    private UUID ownerUUID;
    private int ownerNetworkId;
    private boolean leftOwner;
 
+   // CraftBukkit Start
+   private boolean hitCancelled = false;
+   // CraftBukkit end
+
    ProjectileEntity(EntityType<? extends ProjectileEntity> p_i231584_1_, World p_i231584_2_) {
       super(p_i231584_1_, p_i231584_2_);
    }
@@ -29,7 +_,9 @@
       if (p_212361_1_ != null) {
          this.ownerUUID = p_212361_1_.getUUID();
          this.ownerNetworkId = p_212361_1_.getId();
+
       }
+      this.projectileSource = (p_212361_1_ != null && p_212361_1_.getBukkitEntity() instanceof ProjectileSource) ? (ProjectileSource) p_212361_1_.getBukkitEntity() : null; // CraftBukkit
 
    }
 
@@ -103,7 +_,20 @@
       this.setDeltaMovement(this.getDeltaMovement().add(vector3d.x, p_234612_1_.isOnGround() ? 0.0D : vector3d.y, vector3d.z));
    }
 
+   // CraftBukkit Start - call projectile hit event
+   protected void preOnHit(RayTraceResult movingobjectposition)
+   {
+      org.bukkit.event.entity.ProjectileHitEvent event = org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callProjectileHitEvent(this, movingobjectposition);
+      this.hitCancelled = event != null && event.isCancelled();
+      if(movingobjectposition.getType() == RayTraceResult.Type.BLOCK || !this.hitCancelled)
+      {
+         this.onHit(movingobjectposition); // PAIL rename onHit
+      }
+   }
+
+
    protected void onHit(RayTraceResult p_70227_1_) {
+      org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callProjectileHitEvent(this, p_70227_1_); // CraftBukkit - Call event
       RayTraceResult.Type raytraceresult$type = p_70227_1_.getType();
       if (raytraceresult$type == RayTraceResult.Type.ENTITY) {
          this.onHitEntity((EntityRayTraceResult)p_70227_1_);
