--- a/net/minecraft/entity/projectile/ProjectileItemEntity.java
+++ b/net/minecraft/entity/projectile/ProjectileItemEntity.java
@@ -36,7 +_,9 @@
    public void setItem(ItemStack p_213884_1_) {
       if (p_213884_1_.getItem() != this.getDefaultItem() || p_213884_1_.hasTag()) {
          this.getEntityData().set(DATA_ITEM_STACK, Util.make(p_213884_1_.copy(), (p_213883_0_) -> {
-            p_213883_0_.setCount(1);
+            if (!p_213883_0_.isEmpty()) {
+               p_213883_0_.setCount(1); // CraftBukkit
+            }
          }));
       }
 
@@ -44,7 +_,13 @@
 
    protected abstract Item getDefaultItem();
 
-   protected ItemStack getItemRaw() {
+   // CraftBukkit start
+   public Item getDefaultItemPublic() {
+      return getDefaultItem();
+   }
+   // CraftBukkit end
+
+   public ItemStack getItemRaw() {
       return this.getEntityData().get(DATA_ITEM_STACK);
    }
 
