--- a/net/minecraft/entity/projectile/ShulkerBulletEntity.java
+++ b/net/minecraft/entity/projectile/ShulkerBulletEntity.java
@@ -29,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ShulkerBulletEntity extends ProjectileEntity {
    private Entity finalTarget;
@@ -64,12 +_,25 @@
       this.finalTarget = p_i46772_3_;
       this.currentMoveDirection = Direction.UP;
       this.selectNextMoveDirection(p_i46772_4_);
+      projectileSource = (org.bukkit.entity.LivingEntity) p_i46772_2_.getBukkitEntity(); // CraftBukkit
    }
 
    public SoundCategory getSoundSource() {
       return SoundCategory.HOSTILE;
    }
 
+   // CraftBukkit start
+   public Entity getTarget() {
+      return this.finalTarget;
+   }
+
+   public void setTarget(Entity e) {
+      this.finalTarget = e;
+      this.currentMoveDirection = Direction.UP;
+      this.selectNextMoveDirection(Direction.Axis.X);
+   }
+   // CraftBukkit end
+
    protected void addAdditionalSaveData(CompoundNBT p_213281_1_) {
       super.addAdditionalSaveData(p_213281_1_);
       if (this.finalTarget != null) {
@@ -213,8 +_,8 @@
          }
 
          RayTraceResult raytraceresult = ProjectileHelper.getHitResult(this, this::canHitEntity);
-         if (raytraceresult.getType() != RayTraceResult.Type.MISS) {
-            this.onHit(raytraceresult);
+            if (raytraceresult.getType() != RayTraceResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
+            this.preOnHit(raytraceresult);
          }
       }
 
@@ -274,7 +_,7 @@
       if (flag) {
          this.doEnchantDamageEffects(livingentity, entity);
          if (entity instanceof LivingEntity) {
-            ((LivingEntity)entity).addEffect(new EffectInstance(Effects.LEVITATION, 200));
+            ((LivingEntity) entity).addEffect(new EffectInstance(Effects.LEVITATION, 200), Cause.ATTACK); // CraftBukkit
          }
       }
 
@@ -296,6 +_,11 @@
    }
 
    public boolean hurt(DamageSource p_70097_1_, float p_70097_2_) {
+      // CraftBukkit start
+      if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_, false)) {
+         return false;
+      }
+      // CraftBukkit end
       if (!this.level.isClientSide) {
          this.playSound(SoundEvents.SHULKER_BULLET_HURT, 1.0F, 1.0F);
          ((ServerWorld)this.level).sendParticles(ParticleTypes.CRIT, this.getX(), this.getY(), this.getZ(), 15, 0.2D, 0.2D, 0.2D, 0.0D);
