--- a/net/minecraft/entity/projectile/SmallFireballEntity.java
+++ b/net/minecraft/entity/projectile/SmallFireballEntity.java
@@ -12,6 +_,7 @@
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityCombustByEntityEvent;
 
 public class SmallFireballEntity extends AbstractFireballEntity {
    public SmallFireballEntity(EntityType<? extends SmallFireballEntity> p_i50160_1_, World p_i50160_2_) {
@@ -20,6 +_,11 @@
 
    public SmallFireballEntity(World p_i1771_1_, LivingEntity p_i1771_2_, double p_i1771_3_, double p_i1771_5_, double p_i1771_7_) {
       super(EntityType.SMALL_FIREBALL, p_i1771_2_, p_i1771_3_, p_i1771_5_, p_i1771_7_, p_i1771_1_);
+      // CraftBukkit start
+      if (this.getOwner() != null && this.getOwner() instanceof MobEntity) {
+         isIncendiary = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
+      }
+      // CraftBukkit end
    }
 
    public SmallFireballEntity(World p_i1772_1_, double p_i1772_2_, double p_i1772_4_, double p_i1772_6_, double p_i1772_8_, double p_i1772_10_, double p_i1772_12_) {
@@ -33,7 +_,15 @@
          if (!entity.fireImmune()) {
             Entity entity1 = this.getOwner();
             int i = entity.getRemainingFireTicks();
-            entity.setSecondsOnFire(5);
+            // CraftBukkit start - Entity damage by entity event + combust event
+            EntityCombustByEntityEvent event = new EntityCombustByEntityEvent((org.bukkit.entity.Projectile) this.getBukkitEntity(), entity.getBukkitEntity(), 5);
+            entity.level.getCBServer().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               entity.putCallCombustEvent(false);
+               entity.setSecondsOnFire(event.getDuration());
+               entity.putCallCombustEvent(true);
+            }
+            // CraftBukkit end
             boolean flag = entity.hurt(DamageSource.fireball(this, entity1), 5.0F);
             if (!flag) {
                entity.setRemainingFireTicks(i);
@@ -49,9 +_,9 @@
       super.onHitBlock(p_230299_1_);
       if (!this.level.isClientSide) {
          Entity entity = this.getOwner();
-         if (entity == null || !(entity instanceof MobEntity) || this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+         if (isIncendiary) { // CraftBukkit
             BlockPos blockpos = p_230299_1_.getBlockPos().relative(p_230299_1_.getDirection());
-            if (this.level.isEmptyBlock(blockpos)) {
+            if (this.level.isEmptyBlock(blockpos) && !org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockIgniteEvent(level, blockpos, this).isCancelled()) { // CraftBukkit
                this.level.setBlockAndUpdate(blockpos, AbstractFireBlock.getState(this.level, blockpos));
             }
          }
