--- a/net/minecraft/entity/projectile/TridentEntity.java
+++ b/net/minecraft/entity/projectile/TridentEntity.java
@@ -24,11 +_,12 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.weather.LightningStrikeEvent.Cause;
 
 public class TridentEntity extends AbstractArrowEntity {
    private static final DataParameter<Byte> ID_LOYALTY = EntityDataManager.defineId(TridentEntity.class, DataSerializers.BYTE);
    private static final DataParameter<Boolean> ID_FOIL = EntityDataManager.defineId(TridentEntity.class, DataSerializers.BOOLEAN);
-   private ItemStack tridentItem = new ItemStack(Items.TRIDENT);
+   public ItemStack tridentItem = new ItemStack(Items.TRIDENT);
    private boolean dealtDamage;
    public int clientSideReturnTridentTickCount;
 
@@ -148,7 +_,7 @@
             LightningBoltEntity lightningboltentity = EntityType.LIGHTNING_BOLT.create(this.level);
             lightningboltentity.moveTo(Vector3d.atBottomCenterOf(blockpos));
             lightningboltentity.setCause(entity1 instanceof ServerPlayerEntity ? (ServerPlayerEntity)entity1 : null);
-            this.level.addFreshEntity(lightningboltentity);
+            ((ServerWorld)this.level).strikeLightning(lightningboltentity, Cause.TRIDENT); // CraftBukkit
             soundevent = SoundEvents.TRIDENT_THUNDER;
             f1 = 5.0F;
          }
