--- a/net/minecraft/world/server/ServerBossInfo.java
+++ b/net/minecraft/world/server/ServerBossInfo.java
@@ -15,7 +_,7 @@
 public class ServerBossInfo extends BossInfo {
    private final Set<ServerPlayerEntity> players = Sets.newHashSet();
    private final Set<ServerPlayerEntity> unmodifiablePlayers = Collections.unmodifiableSet(this.players);
-   private boolean visible = true;
+   public boolean visible = true;
 
    public ServerBossInfo(ITextComponent p_i46839_1_, BossInfo.Color p_i46839_2_, BossInfo.Overlay p_i46839_3_) {
       super(MathHelper.createInsecureUUID(), p_i46839_1_, p_i46839_2_, p_i46839_3_);
@@ -80,7 +_,7 @@
 
    }
 
-   private void broadcast(SUpdateBossInfoPacket.Operation p_186759_1_) {
+   public void broadcast(SUpdateBossInfoPacket.Operation p_186759_1_) {
       if (this.visible) {
          SUpdateBossInfoPacket supdatebossinfopacket = new SUpdateBossInfoPacket(p_186759_1_, this);
 
