--- a/net/minecraft/entity/player/PlayerAbilities.java
+++ b/net/minecraft/entity/player/PlayerAbilities.java
@@ -10,8 +_,8 @@
    public boolean mayfly;
    public boolean instabuild;
    public boolean mayBuild = true;
-   private float flyingSpeed = 0.05F;
-   private float walkingSpeed = 0.1F;
+   public float flyingSpeed = 0.05F;
+   public float walkingSpeed = 0.1F;
 
    public void addSaveData(CompoundNBT p_75091_1_) {
       CompoundNBT compoundnbt = new CompoundNBT();
