--- a/net/minecraft/entity/monster/piglin/AbstractPiglinEntity.java
+++ b/net/minecraft/entity/monster/piglin/AbstractPiglinEntity.java
@@ -24,7 +_,7 @@
 
 public abstract class AbstractPiglinEntity extends MonsterEntity {
    protected static final DataParameter<Boolean> DATA_IMMUNE_TO_ZOMBIFICATION = EntityDataManager.defineId(AbstractPiglinEntity.class, DataSerializers.BOOLEAN);
-   protected int timeInOverworld = 0;
+   public int timeInOverworld = 0;
 
    public AbstractPiglinEntity(EntityType<? extends AbstractPiglinEntity> p_i241915_1_, World p_i241915_2_) {
       super(p_i241915_1_, p_i241915_2_);
@@ -47,7 +_,7 @@
       this.getEntityData().set(DATA_IMMUNE_TO_ZOMBIFICATION, p_242340_1_);
    }
 
-   protected boolean isImmuneToZombification() {
+   public boolean isImmuneToZombification() {
       return this.getEntityData().get(DATA_IMMUNE_TO_ZOMBIFICATION);
    }
 
@@ -83,7 +_,7 @@
          this.timeInOverworld = 0;
       }
 
-      if (this.timeInOverworld > 300) {
+      if (this.timeInOverworld > 300 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOMBIFIED_PIGLIN, (timer) -> this.timeInOverworld = timer)) {
          this.playConvertedSound();
          this.finishConversion((ServerWorld)this.level);
       }
@@ -95,9 +_,10 @@
    }
 
    protected void finishConversion(ServerWorld p_234416_1_) {
-      ZombifiedPiglinEntity zombifiedpiglinentity = this.convertTo(EntityType.ZOMBIFIED_PIGLIN, true);
+      ZombifiedPiglinEntity zombifiedpiglinentity = this.convertTo(EntityType.ZOMBIFIED_PIGLIN, true, org.bukkit.event.entity.EntityTransformEvent.TransformReason.PIGLIN_ZOMBIFIED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.PIGLIN_ZOMBIFIED); // CraftBukkit - add spawn and transform reasons
       if (zombifiedpiglinentity != null) {
          zombifiedpiglinentity.addEffect(new EffectInstance(Effects.CONFUSION, 200, 0));
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zombifiedpiglinentity);
       }
 
    }
