--- a/net/minecraft/item/crafting/FurnaceRecipe.java
+++ b/net/minecraft/item/crafting/FurnaceRecipe.java
@@ -3,6 +_,11 @@
 import net.minecraft.block.Blocks;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.ResourceLocation;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftFurnaceRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftRecipe;
+import org.bukkit.craftbukkit.v1_16_R3.util.CraftNamespacedKey;
+import org.bukkit.inventory.Recipe;
 
 public class FurnaceRecipe extends AbstractCookingRecipe {
    public FurnaceRecipe(ResourceLocation p_i48715_1_, String p_i48715_2_, Ingredient p_i48715_3_, ItemStack p_i48715_4_, float p_i48715_5_, int p_i48715_6_) {
@@ -16,4 +_,14 @@
    public IRecipeSerializer<?> getSerializer() {
       return IRecipeSerializer.SMELTING_RECIPE;
    }
+
+   // CraftBukkit start
+    @Override
+    public Recipe toBukkitRecipe() {
+      CraftItemStack result = CraftItemStack.asCraftCopy(CraftItemStack.asBukkitCopy(this.getResultItem()));
+      CraftFurnaceRecipe recipe = new CraftFurnaceRecipe(CraftNamespacedKey.fromMinecraft(this.id), result, CraftRecipe.toBukkit(this.ingredient), this.experience, this.cookingTime);
+      recipe.setGroup(this.group);
+      return recipe;
+    }
+   // CraftBukkit end
 }
