--- a/net/minecraft/item/crafting/RecipeBook.java
+++ b/net/minecraft/item/crafting/RecipeBook.java
@@ -9,7 +_,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class RecipeBook {
-   protected final Set<ResourceLocation> known = Sets.newHashSet();
+   public final Set<ResourceLocation> known = Sets.newHashSet();
    protected final Set<ResourceLocation> highlight = Sets.newHashSet();
    private final RecipeBookStatus bookSettings = new RecipeBookStatus();
 
