--- a/net/minecraft/item/crafting/ServerRecipeBook.java
+++ b/net/minecraft/item/crafting/ServerRecipeBook.java
@@ -16,6 +_,7 @@
 import net.minecraft.util.ResourceLocationException;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ServerRecipeBook extends RecipeBook {
    private static final Logger LOGGER = LogManager.getLogger();
@@ -26,7 +_,7 @@
 
       for(IRecipe<?> irecipe : p_197926_1_) {
          ResourceLocation resourcelocation = irecipe.getId();
-         if (!this.known.contains(resourcelocation) && !irecipe.isSpecial()) {
+         if (!this.known.contains(resourcelocation) && !irecipe.isSpecial() && CraftEventFactory.handlePlayerRecipeListUpdateEvent(p_197926_2_, resourcelocation)) { // CraftBukkit
             this.add(resourcelocation);
             this.addHighlight(resourcelocation);
             list.add(resourcelocation);
@@ -57,6 +_,7 @@
    }
 
    private void sendRecipes(SRecipeBookPacket.State p_194081_1_, ServerPlayerEntity p_194081_2_, List<ResourceLocation> p_194081_3_) {
+      if (p_194081_2_.connection == null) return; // SPIGOT-4478 during PlayerLoginEvent
       p_194081_2_.connection.send(new SRecipeBookPacket(p_194081_1_, p_194081_3_, Collections.emptyList(), this.getBookSettings()));
    }
 
