--- a/net/minecraft/util/CooldownTracker.java
+++ b/net/minecraft/util/CooldownTracker.java
@@ -10,8 +_,8 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class CooldownTracker {
-   private final Map<Item, CooldownTracker.Cooldown> cooldowns = Maps.newHashMap();
-   private int tickCount;
+   public final Map<Item, CooldownTracker.Cooldown> cooldowns = Maps.newHashMap();
+   public int tickCount;
 
    public boolean isOnCooldown(Item p_185141_1_) {
       return this.getCooldownPercent(p_185141_1_, 0.0F) > 0.0F;
@@ -61,9 +_,9 @@
    protected void onCooldownEnded(Item p_185146_1_) {
    }
 
-   class Cooldown {
+   public class Cooldown {
       private final int startTime;
-      private final int endTime;
+      public final int endTime;
 
       private Cooldown(int p_i47037_2_, int p_i47037_3_) {
          this.startTime = p_i47037_2_;
