--- a/net/minecraft/util/EntityPredicates.java
+++ b/net/minecraft/util/EntityPredicates.java
@@ -41,7 +_,7 @@
       Team team = p_200823_0_.getTeam();
       Team.CollisionRule team$collisionrule = team == null ? Team.CollisionRule.ALWAYS : team.getCollisionRule();
       return (Predicate<Entity>)(team$collisionrule == Team.CollisionRule.NEVER ? Predicates.alwaysFalse() : NO_SPECTATORS.and((p_233586_3_) -> {
-         if (!p_233586_3_.isPushable()) {
+         if (!p_233586_3_.canCollideWithCB(p_200823_0_) || !p_200823_0_.canCollideWithCB(p_233586_3_)) { // CraftBukkit - collidable API
             return false;
          } else if (!p_200823_0_.level.isClientSide || p_233586_3_ instanceof PlayerEntity && ((PlayerEntity)p_233586_3_).isLocalPlayer()) {
             Team team1 = p_233586_3_.getTeam();
