--- a/net/minecraft/util/IndirectEntityDamageSource.java
+++ b/net/minecraft/util/IndirectEntityDamageSource.java
@@ -32,4 +_,11 @@
       String s1 = s + ".item";
       return !itemstack.isEmpty() && itemstack.hasCustomHoverName() ? new TranslationTextComponent(s1, p_151519_1_.getDisplayName(), itextcomponent, itemstack.getDisplayName()) : new TranslationTextComponent(s, p_151519_1_.getDisplayName(), itextcomponent);
    }
+
+   // CraftBukkit start
+   public Entity getProximateDamageSource() {
+      return super.getEntity();
+   }
+   // CraftBukkit end
+
 }
