--- a/net/minecraft/util/Util.java
+++ b/net/minecraft/util/Util.java
@@ -104,6 +_,10 @@
             ForkJoinWorkerThread forkjoinworkerthread = new ForkJoinWorkerThread(p_240981_1_) {
                protected void onTermination(Throwable p_onTermination_1_) {
                   if (p_onTermination_1_ != null) {
+                     // CatServer start
+                     p_onTermination_1_.printStackTrace();
+                     catserver.server.utils.WorkerExceptionLogger.addException(p_onTermination_1_);
+                     // CatServer end
                      Util.LOGGER.warn("{} died", this.getName(), p_onTermination_1_);
                   } else {
                      Util.LOGGER.debug("{} shutdown", (Object)this.getName());
@@ -200,7 +_,7 @@
       try {
          type = DataFixesManager.getDataFixer().getSchema(DataFixUtils.makeKey(SharedConstants.getCurrentVersion().getWorldVersion())).getChoiceType(p_240990_0_, p_240990_1_);
       } catch (IllegalArgumentException illegalargumentexception) {
-         LOGGER.error("No data fixer registered for {}", (Object)p_240990_1_);
+         LOGGER.debug("No data fixer registered for {}", (Object)p_240990_1_);
          if (SharedConstants.IS_RUNNING_IN_IDE) {
             throw illegalargumentexception;
          }
