--- a/net/minecraft/entity/item/EnderPearlEntity.java
+++ b/net/minecraft/entity/item/EnderPearlEntity.java
@@ -19,6 +_,10 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.player.PlayerTeleportEvent;
+import org.bukkit.event.player.PlayerTeleportEvent.TeleportCause;
 
 public class EnderPearlEntity extends ProjectileItemEntity {
    public EnderPearlEntity(EntityType<? extends EnderPearlEntity> p_i50153_1_, World p_i50153_2_) {
@@ -55,20 +_,37 @@
          if (entity instanceof ServerPlayerEntity) {
             ServerPlayerEntity serverplayerentity = (ServerPlayerEntity)entity;
             if (serverplayerentity.connection.getConnection().isConnected() && serverplayerentity.level == this.level && !serverplayerentity.isSleeping()) {
-               if (this.random.nextFloat() < 0.05F && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
-                  EndermiteEntity endermiteentity = EntityType.ENDERMITE.create(this.level);
-                  endermiteentity.setPlayerSpawned(true);
-                  endermiteentity.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.yRot, entity.xRot);
-                  this.level.addFreshEntity(endermiteentity);
-               }
-
-               if (entity.isPassenger()) {
-                  entity.stopRiding();
-               }
-
-               entity.teleportTo(this.getX(), this.getY(), this.getZ());
-               entity.fallDistance = 0.0F;
-               entity.hurt(DamageSource.FALL, 5.0F);
+               net.minecraftforge.event.entity.living.EntityTeleportEvent.EnderPearl event = net.minecraftforge.event.ForgeEventFactory.onEnderPearlLand(serverplayerentity, this.getX(), this.getY(), this.getZ(), this, 5.0F);
+               if (!event.isCanceled()) { // Don't indent to lower patch size
+                  // CraftBukkit start - Fire PlayerTeleportEvent
+                  org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer player = serverplayerentity.getBukkitEntity();
+                  org.bukkit.Location location = getBukkitEntity().getLocation();
+                  location.setPitch(player.getLocation().getPitch());
+                  location.setYaw(player.getLocation().getYaw());
+
+                  PlayerTeleportEvent teleportEvent = new PlayerTeleportEvent(player, player.getLocation(), location, TeleportCause.ENDER_PEARL);
+                  Bukkit.getPluginManager().callEvent(teleportEvent);
+
+                  if (!teleportEvent.isAsynchronous() && !serverplayerentity.connection.isDisconnected()) {
+                     if (this.random.nextFloat() < 0.05F && this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING)) {
+                        EndermiteEntity endermiteentity = EntityType.ENDERMITE.create(this.level);
+                        endermiteentity.setPlayerSpawned(true);
+                        endermiteentity.moveTo(entity.getX(), entity.getY(), entity.getZ(), entity.yRot, entity.xRot);
+                        this.level.addFreshEntity(endermiteentity);
+                     }
+
+                     if (entity.isPassenger()) {
+                        entity.stopRiding();
+                     }
+
+                     serverplayerentity.connection.teleport(teleportEvent.getTo());
+                     entity.fallDistance = 0.0f;
+                     CraftEventFactory.entityDamage = this;
+                     entity.hurt(DamageSource.FALL, event.getAttackDamage());
+                     CraftEventFactory.entityDamage = null;
+                  }
+                  // CraftBukkit end
+               } //Forge: End
             }
          } else if (entity != null) {
             entity.teleportTo(this.getX(), this.getY(), this.getZ());
@@ -91,12 +_,12 @@
    }
 
    @Nullable
-   public Entity changeDimension(ServerWorld p_241206_1_) {
+   public Entity changeDimension(ServerWorld p_241206_1_, net.minecraftforge.common.util.ITeleporter teleporter) {
       Entity entity = this.getOwner();
-      if (entity != null && entity.level.dimension() != p_241206_1_.dimension()) {
+      if (entity != null && p_241206_1_ != null && entity.level.dimension() != p_241206_1_.dimension()) { // CraftBukkit - SPIGOT-6113
          this.setOwner((Entity)null);
       }
 
-      return super.changeDimension(p_241206_1_);
+      return super.changeDimension(p_241206_1_, teleporter);
    }
 }
