--- a/net/minecraft/entity/item/EnderCrystalEntity.java
+++ b/net/minecraft/entity/item/EnderCrystalEntity.java
@@ -21,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class EnderCrystalEntity extends Entity {
    private static final DataParameter<Optional<BlockPos>> DATA_BEAM_TARGET = EntityDataManager.defineId(EnderCrystalEntity.class, DataSerializers.OPTIONAL_BLOCK_POS);
@@ -52,7 +_,11 @@
       if (this.level instanceof ServerWorld) {
          BlockPos blockpos = this.blockPosition();
          if (((ServerWorld)this.level).dragonFight() != null && this.level.getBlockState(blockpos).isAir()) {
-            this.level.setBlockAndUpdate(blockpos, AbstractFireBlock.getState(this.level, blockpos));
+            // CraftBukkit start
+            if (!CraftEventFactory.callBlockIgniteEvent(this.level, blockpos, this).isCancelled()) {
+               this.level.setBlockAndUpdate(blockpos, AbstractFireBlock.getState(this.level, blockpos));
+            }
+            // CraftBukkit end
          }
       }
 
@@ -88,9 +_,24 @@
          return false;
       } else {
          if (!this.removed && !this.level.isClientSide) {
+            // CraftBukkit start - All non-living entities need this
+            if (CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_70097_1_, p_70097_2_, false)) {
+               return false;
+            }
+            // CraftBukkit end
             this.remove();
             if (!p_70097_1_.isExplosion()) {
-               this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), 6.0F, Explosion.Mode.DESTROY);
+               // CraftBukkit start
+               ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), 6.0F, false);
+               this.level.getCBServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  this.removed = false;
+                  return false;
+               }
+               boolean oldFire = this.level.pehkuiExplosionFire;
+               this.level.pehkuiExplosionFire = event.getFire();
+               this.level.explode((Entity)null, this.getX(), this.getY(), this.getZ(), event.getRadius(), Explosion.Mode.DESTROY);
+               this.level.pehkuiExplosionFire = oldFire;
             }
 
             this.onDestroyedBy(p_70097_1_);
