--- a/net/minecraft/entity/item/ExperienceBottleEntity.java
+++ b/net/minecraft/entity/item/ExperienceBottleEntity.java
@@ -9,6 +_,7 @@
 import net.minecraft.potion.Potions;
 import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ExperienceBottleEntity extends ProjectileItemEntity {
    public ExperienceBottleEntity(EntityType<? extends ExperienceBottleEntity> p_i50152_1_, World p_i50152_2_) {
@@ -34,8 +_,17 @@
    protected void onHit(RayTraceResult p_70227_1_) {
       super.onHit(p_70227_1_);
       if (!this.level.isClientSide) {
-         this.level.levelEvent(2002, this.blockPosition(), PotionUtils.getColor(Potions.WATER));
+         // CraftBukkit - moved to after event
+         // this.world.playEvent(2002, this.getPosition(), PotionUtils.getPotionColor(Potions.WATER));
          int i = 3 + this.level.random.nextInt(5) + this.level.random.nextInt(5);
+
+         // CraftBukkit start
+         org.bukkit.event.entity.ExpBottleEvent event = CraftEventFactory.callExpBottleEvent(this, i);
+         i = event.getExperience();
+         if (event.getShowEffect()) {
+            this.level.levelEvent(2002, this.blockPosition(), PotionUtils.getColor(Potions.WATER));
+         }
+         // CraftBukkit end
 
          while(i > 0) {
             int j = ExperienceOrbEntity.getExperienceValue(i);
