--- a/net/minecraft/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/entity/item/FallingBlockEntity.java
@@ -2,13 +_,13 @@
 
 import com.google.common.collect.Lists;
 import java.util.List;
+
 import net.minecraft.block.AnvilBlock;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.Blocks;
 import net.minecraft.block.ConcretePowderBlock;
 import net.minecraft.block.FallingBlock;
-import net.minecraft.block.ITileEntityProvider;
 import net.minecraft.crash.CrashReportCategory;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
@@ -40,13 +_,14 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class FallingBlockEntity extends Entity {
    private BlockState blockState = Blocks.SAND.defaultBlockState();
    public int time;
    public boolean dropItem = true;
    private boolean cancelDrop;
-   private boolean hurtEntities;
+   public boolean hurtEntities; // private->public CraftBukkit
    private int fallDamageMax = 40;
    private float fallDamageAmount = 2.0F;
    public CompoundNBT blockData;
@@ -100,7 +_,7 @@
          Block block = this.blockState.getBlock();
          if (this.time++ == 0) {
             BlockPos blockpos = this.blockPosition();
-            if (this.level.getBlockState(blockpos).is(block)) {
+            if (this.level.getBlockState(blockpos).is(block) && !CraftEventFactory.callEntityChangeBlockEvent(this, blockpos, Blocks.AIR.defaultBlockState()).isCancelled()) { // CraftBukkit
                this.level.removeBlock(blockpos, false);
             } else if (!this.level.isClientSide) {
                this.remove();
@@ -148,12 +_,18 @@
                            this.blockState = this.blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true));
                         }
 
+                        // CraftBukkit start
+                        if(CraftEventFactory.callEntityChangeBlockEvent(this,blockpos1, this.blockState).isCancelled()){
+                           return;
+                        }
+                        // CraftBukkit end
+
                         if (this.level.setBlock(blockpos1, this.blockState, 3)) {
                            if (block instanceof FallingBlock) {
                               ((FallingBlock)block).onLand(this.level, blockpos1, this.blockState, blockstate, this);
                            }
 
-                           if (this.blockData != null && block instanceof ITileEntityProvider) {
+                           if (this.blockData != null && this.blockState.hasTileEntity()) {
                               TileEntity tileentity = this.level.getBlockEntity(blockpos1);
                               if (tileentity != null) {
                                  CompoundNBT compoundnbt = tileentity.save(new CompoundNBT());
@@ -195,7 +_,9 @@
             DamageSource damagesource = flag ? DamageSource.ANVIL : DamageSource.FALLING_BLOCK;
 
             for(Entity entity : list) {
+               CraftEventFactory.entityDamage = this; // CraftBukkit
                entity.hurt(damagesource, (float)Math.min(MathHelper.floor((float)i * this.fallDamageAmount), this.fallDamageMax));
+               CraftEventFactory.entityDamage = null; // CraftBukkit
             }
 
             if (flag && (double)this.random.nextFloat() < (double)0.05F + (double)i * 0.05D) {
