--- a/net/minecraft/entity/item/TNTEntity.java
+++ b/net/minecraft/entity/item/TNTEntity.java
@@ -16,12 +_,15 @@
 import net.minecraft.particles.ParticleTypes;
 import net.minecraft.world.Explosion;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 public class TNTEntity extends Entity {
    private static final DataParameter<Integer> DATA_FUSE_ID = EntityDataManager.defineId(TNTEntity.class, DataSerializers.INT);
    @Nullable
-   private LivingEntity owner;
+   public LivingEntity owner;
    private int life = 80;
+   public float yield = 4; // CraftBukkit - add field
+   public boolean isIncendiary = false; // CraftBukkit - add field
 
    public TNTEntity(EntityType<? extends TNTEntity> p_i50216_1_, World p_i50216_2_) {
       super(p_i50216_1_, p_i50216_2_);
@@ -53,6 +_,7 @@
    }
 
    public void tick() {
+      if (level.spigotConfig.currentPrimedTnt++ > level.spigotConfig.maxTntTicksPerTick) { return; } // Spigot
       if (!this.isNoGravity()) {
          this.setDeltaMovement(this.getDeltaMovement().add(0.0D, -0.04D, 0.0D));
       }
@@ -65,10 +_,13 @@
 
       --this.life;
       if (this.life <= 0) {
-         this.remove();
+         // CraftBukkit start - Need to reverse the order of the explosion and the entity death so we have a location for the event
+    //    this.remove();
          if (!this.level.isClientSide) {
             this.explode();
          }
+         this.remove();
+         // CraftBukkit end
       } else {
          this.updateInWaterStateAndDoFluidPushing();
          if (this.level.isClientSide) {
@@ -79,8 +_,13 @@
    }
 
    protected void explode() {
-      float f = 4.0F;
-      this.level.explode(this, this.getX(), this.getY(0.0625D), this.getZ(), 4.0F, Explosion.Mode.BREAK);
+      // CraftBukkit start
+      ExplosionPrimeEvent event = new ExplosionPrimeEvent((org.bukkit.entity.Explosive) this.getBukkitEntity());
+      this.level.getCBServer().getPluginManager().callEvent(event);
+      if (!event.isCancelled()) {
+         this.level.explode(this, this.getX(), this.getY(0.0625D), this.getZ(), event.getRadius(), event.getFire(), Explosion.Mode.BREAK);
+      }
+      // CraftBukkit end
    }
 
    protected void addAdditionalSaveData(CompoundNBT p_213281_1_) {
