--- a/net/minecraft/command/arguments/EntitySelector.java
+++ b/net/minecraft/command/arguments/EntitySelector.java
@@ -72,7 +_,7 @@
    }
 
    private void checkPermissions(CommandSource p_210324_1_) throws CommandSyntaxException {
-      if (this.usesSelector && !p_210324_1_.hasPermission(2)) {
+      if (this.usesSelector && !p_210324_1_.hasPermission(2, "minecraft.command.selector")) { // CraftBukkit
          throw EntityArgument.ERROR_SELECTORS_NOT_ALLOWED.create();
       }
    }
