--- a/net/minecraft/command/arguments/EntitySelectorParser.java
+++ b/net/minecraft/command/arguments/EntitySelectorParser.java
@@ -188,7 +_,13 @@
    }
 
    protected void parseSelector() throws CommandSyntaxException {
-      this.usesSelectors = true;
+      parseSelector(true);
+   }
+
+   // CraftBukkit start
+   protected void parseSelector(boolean overridePermissions) throws CommandSyntaxException {
+      this.usesSelectors = !overridePermissions;
+      // CraftBukkit end
       this.suggestions = this::suggestSelector;
       if (!this.reader.canRead()) {
          throw ERROR_MISSING_SELECTOR_TYPE.createWithContext(this.reader);
@@ -440,6 +_,12 @@
    }
 
    public EntitySelector parse() throws CommandSyntaxException {
+      // CraftBukkit start
+      return parse(false);
+   }
+
+   public EntitySelector parse(boolean overridePermissions) throws CommandSyntaxException {
+      // CraftBukkit end
       this.startPosition = this.reader.getCursor();
       this.suggestions = this::suggestNameOrSelector;
       if (this.reader.canRead() && this.reader.peek() == '@') {
@@ -448,7 +_,10 @@
          }
 
          this.reader.skip();
-         this.parseSelector();
+         EntitySelector forgeSelector = net.minecraftforge.common.command.EntitySelectorManager.parseSelector(this);
+         if (forgeSelector != null)
+            return forgeSelector;
+         this.parseSelector(overridePermissions); // CraftBukkit
       } else {
          this.parseNameOrUUID();
       }
@@ -463,6 +_,7 @@
       p_210326_0_.suggest("@r", new TranslationTextComponent("argument.entity.selector.randomPlayer"));
       p_210326_0_.suggest("@s", new TranslationTextComponent("argument.entity.selector.self"));
       p_210326_0_.suggest("@e", new TranslationTextComponent("argument.entity.selector.allEntities"));
+      net.minecraftforge.common.command.EntitySelectorManager.fillSelectorSuggestions(p_210326_0_);
    }
 
    private CompletableFuture<Suggestions> suggestNameOrSelector(SuggestionsBuilder p_201981_1_, Consumer<SuggestionsBuilder> p_201981_2_) {
