--- a/net/minecraft/scoreboard/ServerScoreboard.java
+++ b/net/minecraft/scoreboard/ServerScoreboard.java
@@ -26,7 +_,7 @@
    public void onScoreChanged(Score p_96536_1_) {
       super.onScoreChanged(p_96536_1_);
       if (this.trackedObjectives.contains(p_96536_1_.getObjective())) {
-         this.server.getPlayerList().broadcastAll(new SUpdateScorePacket(ServerScoreboard.Action.CHANGE, p_96536_1_.getObjective().getName(), p_96536_1_.getOwner(), p_96536_1_.getScore()));
+         this.sendAll(new SUpdateScorePacket(ServerScoreboard.Action.CHANGE, p_96536_1_.getObjective().getName(), p_96536_1_.getOwner(), p_96536_1_.getScore())); // CraftBukkit
       }
 
       this.setDirty();
@@ -34,14 +_,14 @@
 
    public void onPlayerRemoved(String p_96516_1_) {
       super.onPlayerRemoved(p_96516_1_);
-      this.server.getPlayerList().broadcastAll(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, (String)null, p_96516_1_, 0));
+      this.sendAll(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, (String)null, p_96516_1_, 0)); // CraftBukkit
       this.setDirty();
    }
 
    public void onPlayerScoreRemoved(String p_178820_1_, ScoreObjective p_178820_2_) {
       super.onPlayerScoreRemoved(p_178820_1_, p_178820_2_);
       if (this.trackedObjectives.contains(p_178820_2_)) {
-         this.server.getPlayerList().broadcastAll(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, p_178820_2_.getName(), p_178820_1_, 0));
+         this.sendAll(new SUpdateScorePacket(ServerScoreboard.Action.REMOVE, p_178820_2_.getName(), p_178820_1_, 0)); // CraftBukkit
       }
 
       this.setDirty();
@@ -52,7 +_,7 @@
       super.setDisplayObjective(p_96530_1_, p_96530_2_);
       if (scoreobjective != p_96530_2_ && scoreobjective != null) {
          if (this.getObjectiveDisplaySlotCount(scoreobjective) > 0) {
-            this.server.getPlayerList().broadcastAll(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_));
+            this.sendAll(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_)); // CraftBukkit
          } else {
             this.stopTrackingObjective(scoreobjective);
          }
@@ -60,7 +_,7 @@
 
       if (p_96530_2_ != null) {
          if (this.trackedObjectives.contains(p_96530_2_)) {
-            this.server.getPlayerList().broadcastAll(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_));
+            this.sendAll(new SDisplayObjectivePacket(p_96530_1_, p_96530_2_)); // CraftBukkit
          } else {
             this.startTrackingObjective(p_96530_2_);
          }
@@ -71,7 +_,7 @@
 
    public boolean addPlayerToTeam(String p_197901_1_, ScorePlayerTeam p_197901_2_) {
       if (super.addPlayerToTeam(p_197901_1_, p_197901_2_)) {
-         this.server.getPlayerList().broadcastAll(new STeamsPacket(p_197901_2_, Arrays.asList(p_197901_1_), 3));
+         this.sendAll(new STeamsPacket(p_197901_2_, Arrays.asList(p_197901_1_), 3)); // CraftBukkit
          this.setDirty();
          return true;
       } else {
@@ -81,7 +_,7 @@
 
    public void removePlayerFromTeam(String p_96512_1_, ScorePlayerTeam p_96512_2_) {
       super.removePlayerFromTeam(p_96512_1_, p_96512_2_);
-      this.server.getPlayerList().broadcastAll(new STeamsPacket(p_96512_2_, Arrays.asList(p_96512_1_), 4));
+      this.sendAll(new STeamsPacket(p_96512_2_, Arrays.asList(p_96512_1_), 4)); // CraftBukkit
       this.setDirty();
    }
 
@@ -93,7 +_,7 @@
    public void onObjectiveChanged(ScoreObjective p_199869_1_) {
       super.onObjectiveChanged(p_199869_1_);
       if (this.trackedObjectives.contains(p_199869_1_)) {
-         this.server.getPlayerList().broadcastAll(new SScoreboardObjectivePacket(p_199869_1_, 2));
+         this.sendAll(new SScoreboardObjectivePacket(p_199869_1_, 2)); // CraftBukkit
       }
 
       this.setDirty();
@@ -110,19 +_,19 @@
 
    public void onTeamAdded(ScorePlayerTeam p_96523_1_) {
       super.onTeamAdded(p_96523_1_);
-      this.server.getPlayerList().broadcastAll(new STeamsPacket(p_96523_1_, 0));
+      this.sendAll(new STeamsPacket(p_96523_1_, 0)); // CraftBukkit
       this.setDirty();
    }
 
    public void onTeamChanged(ScorePlayerTeam p_96538_1_) {
       super.onTeamChanged(p_96538_1_);
-      this.server.getPlayerList().broadcastAll(new STeamsPacket(p_96538_1_, 2));
+      this.sendAll(new STeamsPacket(p_96538_1_, 2)); // CraftBukkit
       this.setDirty();
    }
 
    public void onTeamRemoved(ScorePlayerTeam p_96513_1_) {
       super.onTeamRemoved(p_96513_1_);
-      this.server.getPlayerList().broadcastAll(new STeamsPacket(p_96513_1_, 1));
+      this.sendAll(new STeamsPacket(p_96513_1_, 1)); // CraftBukkit
       this.setDirty();
    }
 
@@ -159,6 +_,7 @@
       List<IPacket<?>> list = this.getStartTrackingPackets(p_96549_1_);
 
       for(ServerPlayerEntity serverplayerentity : this.server.getPlayerList().getPlayers()) {
+         if (serverplayerentity.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
          for(IPacket<?> ipacket : list) {
             serverplayerentity.connection.send(ipacket);
          }
@@ -184,6 +_,7 @@
       List<IPacket<?>> list = this.getStopTrackingPackets(p_96546_1_);
 
       for(ServerPlayerEntity serverplayerentity : this.server.getPlayerList().getPlayers()) {
+         if (serverplayerentity.getBukkitEntity().getScoreboard().getHandle() != this) continue; // CraftBukkit - Only players on this board
          for(IPacket<?> ipacket : list) {
             serverplayerentity.connection.send(ipacket);
          }
@@ -203,6 +_,16 @@
 
       return i;
    }
+
+   // CraftBukkit start - Send to players
+   private void sendAll(IPacket packet) {
+      for (ServerPlayerEntity entityplayer : this.server.getPlayerList().players) {
+         if (entityplayer.getBukkitEntity().getScoreboard().getHandle() == this) {
+            entityplayer.connection.send(packet);
+         }
+      }
+   }
+   // CraftBukkit end
 
    public static enum Action {
       CHANGE,
