--- a/net/minecraft/item/ArmorStandItem.java
+++ b/net/minecraft/item/ArmorStandItem.java
@@ -45,11 +_,15 @@
                   return ActionResultType.FAIL;
                }
 
-               serverworld.addFreshEntityWithPassengers(armorstandentity);
                float f = (float)MathHelper.floor((MathHelper.wrapDegrees(p_195939_1_.getRotation() - 180.0F) + 22.5F) / 45.0F) * 45.0F;
                armorstandentity.moveTo(armorstandentity.getX(), armorstandentity.getY(), armorstandentity.getZ(), f, 0.0F);
                this.randomizePose(armorstandentity, world.random);
-               world.addFreshEntity(armorstandentity);
+               // CraftBukkit start
+               if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityPlaceEvent(p_195939_1_, armorstandentity).isCancelled()) {
+                  return ActionResultType.FAIL;
+               }
+               // CraftBukkit end
+               serverworld.addFreshEntityWithPassengers(armorstandentity);
                world.playSound((PlayerEntity)null, armorstandentity.getX(), armorstandentity.getY(), armorstandentity.getZ(), SoundEvents.ARMOR_STAND_PLACE, SoundCategory.BLOCKS, 0.75F, 0.8F);
             }
 
