--- a/net/minecraft/item/ArmorItem.java
+++ b/net/minecraft/item/ArmorItem.java
@@ -23,6 +_,8 @@
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseArmorEvent;
 
 public class ArmorItem extends Item implements IArmorVanishable {
    private static final UUID[] ARMOR_MODIFIER_UUID_PER_SLOT = new UUID[]{UUID.fromString("845DB27C-C624-495F-8C9F-6020A9A58B6B"), UUID.fromString("D8499B04-0E66-4726-AB29-64469D734E0D"), UUID.fromString("9F3D476D-C118-4544-8365-64846904B48E"), UUID.fromString("2AD3F246-FEE1-4E67-B886-69FD380BB150")};
@@ -47,6 +_,29 @@
          LivingEntity livingentity = list.get(0);
          EquipmentSlotType equipmentslottype = MobEntity.getEquipmentSlotForItem(p_226626_1_);
          ItemStack itemstack = p_226626_1_.split(1);
+         // CraftBukkit start
+         World world = p_226626_0_.getLevel();
+         org.bukkit.block.Block block = world.getWorld().getBlockAt(p_226626_0_.getPos().getX(), p_226626_0_.getPos().getY(), p_226626_0_.getPos().getZ());
+         CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_226626_1_);
+         BlockDispenseArmorEvent event = new BlockDispenseArmorEvent(block, craftItem.clone(), (org.bukkit.craftbukkit.v1_16_R3.entity.CraftLivingEntity) livingentity.getBukkitEntity());
+         if (!DispenserBlock.eventFired) {
+            world.getCBServer().getPluginManager().callEvent(event);
+         }
+         if (event.isCancelled()) {
+            itemstack.grow(1);
+            return false;
+         }
+         if (!event.getItem().equals(craftItem)) {
+            itemstack.grow(1);
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            IDispenseItemBehavior idispensebehavior = (IDispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack.getItem());
+            if (idispensebehavior != IDispenseItemBehavior.NOOP && idispensebehavior != ArmorItem.DISPENSE_ITEM_BEHAVIOR) {
+               idispensebehavior.dispense(p_226626_0_, eventStack);
+               return true;
+            }
+         }
+         // CraftBukkit end
          livingentity.setItemSlot(equipmentslottype, itemstack);
          if (livingentity instanceof MobEntity) {
             ((MobEntity)livingentity).setDropChance(equipmentslottype, 2.0F);
@@ -69,7 +_,7 @@
       UUID uuid = ARMOR_MODIFIER_UUID_PER_SLOT[p_i48534_2_.getIndex()];
       builder.put(Attributes.ARMOR, new AttributeModifier(uuid, "Armor modifier", (double)this.defense, AttributeModifier.Operation.ADDITION));
       builder.put(Attributes.ARMOR_TOUGHNESS, new AttributeModifier(uuid, "Armor toughness", (double)this.toughness, AttributeModifier.Operation.ADDITION));
-      if (p_i48534_1_ == ArmorMaterial.NETHERITE) {
+      if (this.knockbackResistance > 0) {
          builder.put(Attributes.KNOCKBACK_RESISTANCE, new AttributeModifier(uuid, "Armor knockback resistance", (double)this.knockbackResistance, AttributeModifier.Operation.ADDITION));
       }
 
