--- a/net/minecraft/item/BoneMealItem.java
+++ b/net/minecraft/item/BoneMealItem.java
@@ -27,10 +_,16 @@
    }
 
    public ActionResultType useOn(ItemUseContext p_195939_1_) {
-      World world = p_195939_1_.getLevel();
-      BlockPos blockpos = p_195939_1_.getClickedPos();
-      BlockPos blockpos1 = blockpos.relative(p_195939_1_.getClickedFace());
-      if (growCrop(p_195939_1_.getItemInHand(), world, blockpos)) {
+      // CraftBukkit start - extract bonemeal application logic to separate, static method
+      return applyBonemeal(p_195939_1_);
+   }
+
+   public static ActionResultType applyBonemeal(ItemUseContext context) {
+      // CraftBukkit end
+      World world = context.getLevel();
+      BlockPos blockpos = context.getClickedPos();
+      BlockPos blockpos1 = blockpos.relative(context.getClickedFace());
+      if (applyBonemeal(context.getItemInHand(), world, blockpos, context.getPlayer())) {
          if (!world.isClientSide) {
             world.levelEvent(2005, blockpos, 0);
          }
@@ -38,8 +_,8 @@
          return ActionResultType.sidedSuccess(world.isClientSide);
       } else {
          BlockState blockstate = world.getBlockState(blockpos);
-         boolean flag = blockstate.isFaceSturdy(world, blockpos, p_195939_1_.getClickedFace());
-         if (flag && growWaterPlant(p_195939_1_.getItemInHand(), world, blockpos1, p_195939_1_.getClickedFace())) {
+         boolean flag = blockstate.isFaceSturdy(world, blockpos, context.getClickedFace());
+         if (flag && growWaterPlant(context.getItemInHand(), world, blockpos1, context.getClickedFace())) {
             if (!world.isClientSide) {
                world.levelEvent(2005, blockpos1, 0);
             }
@@ -51,8 +_,17 @@
       }
    }
 
+   @Deprecated //Forge: Use Player/Hand version
    public static boolean growCrop(ItemStack p_195966_0_, World p_195966_1_, BlockPos p_195966_2_) {
+      if (p_195966_1_ instanceof net.minecraft.world.server.ServerWorld)
+         return applyBonemeal(p_195966_0_, p_195966_1_, p_195966_2_, net.minecraftforge.common.util.FakePlayerFactory.getMinecraft((net.minecraft.world.server.ServerWorld)p_195966_1_));
+      return false;
+   }
+
+   public static boolean applyBonemeal(ItemStack p_195966_0_, World p_195966_1_, BlockPos p_195966_2_, net.minecraft.entity.player.PlayerEntity player) {
       BlockState blockstate = p_195966_1_.getBlockState(p_195966_2_);
+      int hook = net.minecraftforge.event.ForgeEventFactory.onApplyBonemeal(player, p_195966_1_, p_195966_2_, blockstate, p_195966_0_);
+      if (hook != 0) return hook > 0;
       if (blockstate.getBlock() instanceof IGrowable) {
          IGrowable igrowable = (IGrowable)blockstate.getBlock();
          if (igrowable.isValidBonemealTarget(p_195966_1_, p_195966_2_, blockstate, p_195966_1_.isClientSide)) {
@@ -128,7 +_,7 @@
       }
 
       BlockState blockstate = p_195965_0_.getBlockState(p_195965_1_);
-      if (!blockstate.isAir()) {
+      if (!blockstate.isAir(p_195965_0_, p_195965_1_)) {
          double d0 = 0.5D;
          double d1;
          if (blockstate.is(Blocks.WATER)) {
