--- a/net/minecraft/item/ChorusFruitItem.java
+++ b/net/minecraft/item/ChorusFruitItem.java
@@ -8,6 +_,9 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class ChorusFruitItem extends Item {
    public ChorusFruitItem(Item.Properties p_i50053_1_) {
@@ -25,11 +_,34 @@
             double d3 = p_77654_3_.getX() + (p_77654_3_.getRandom().nextDouble() - 0.5D) * 16.0D;
             double d4 = MathHelper.clamp(p_77654_3_.getY() + (double)(p_77654_3_.getRandom().nextInt(16) - 8), 0.0D, (double)(p_77654_2_.getHeight() - 1));
             double d5 = p_77654_3_.getZ() + (p_77654_3_.getRandom().nextDouble() - 0.5D) * 16.0D;
+
+            // CraftBukkit start
+            if (p_77654_3_ instanceof PlayerEntity) {
+               Player player = (Player) p_77654_3_.getBukkitEntity();
+               PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), new Location(player.getWorld(), d3, d4, d5), PlayerTeleportEvent.TeleportCause.CHORUS_FRUIT);
+               p_77654_2_.getCBServer().getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  break;
+               }
+               d3 = teleEvent.getTo().getX();
+               d4 = teleEvent.getTo().getY();
+               d5 = teleEvent.getTo().getZ();
+            }
+            // CraftBukkit end
+
             if (p_77654_3_.isPassenger()) {
                p_77654_3_.stopRiding();
             }
 
-            if (p_77654_3_.randomTeleport(d3, d4, d5, true)) {
+            net.minecraftforge.event.entity.living.EntityTeleportEvent.ChorusFruit event = net.minecraftforge.event.ForgeEventFactory.onChorusFruitTeleport(p_77654_3_, d3, d4, d5);
+            if (event.isCanceled()) return itemstack;
+            java.util.Optional<Boolean> status = p_77654_3_.safeTeleport(event.getTargetX(), event.getTargetY(), event.getTargetZ(), true, PlayerTeleportEvent.TeleportCause.CHORUS_FRUIT);
+            if (!status.isPresent()) {
+               // teleport event was canceled, no more tries
+               break;
+            }
+            if (status.get()){
+               // CraftBukkit end
                SoundEvent soundevent = p_77654_3_ instanceof FoxEntity ? SoundEvents.FOX_TELEPORT : SoundEvents.CHORUS_FRUIT_TELEPORT;
                p_77654_2_.playSound((PlayerEntity)null, d0, d1, d2, soundevent, SoundCategory.PLAYERS, 1.0F, 1.0F);
                p_77654_3_.playSound(soundevent, 1.0F, 1.0F);
