--- a/net/minecraft/item/DebugStickItem.java
+++ b/net/minecraft/item/DebugStickItem.java
@@ -49,7 +_,7 @@
    }
 
    private void handleInteraction(PlayerEntity p_195958_1_, BlockState p_195958_2_, IWorld p_195958_3_, BlockPos p_195958_4_, boolean p_195958_5_, ItemStack p_195958_6_) {
-      if (p_195958_1_.canUseGameMasterBlocks()) {
+      if (p_195958_1_.canUseGameMasterBlocks() || (p_195958_1_.abilities.instabuild && p_195958_1_.getBukkitEntity().hasPermission("minecraft.debugstick")) || p_195958_1_.getBukkitEntity().hasPermission("minecraft.debugstick.always")) { // Spigot
          Block block = p_195958_2_.getBlock();
          StateContainer<Block, BlockState> statecontainer = block.getStateDefinition();
          Collection<Property<?>> collection = statecontainer.getProperties();
