--- a/net/minecraft/item/DyeItem.java
+++ b/net/minecraft/item/DyeItem.java
@@ -8,6 +_,9 @@
 import net.minecraft.util.ActionResultType;
 import net.minecraft.util.Hand;
 
+import org.bukkit.Color;
+import org.bukkit.event.entity.SheepDyeWoolEvent;
+
 public class DyeItem extends Item {
    private static final Map<DyeColor, DyeItem> ITEM_BY_COLOR = Maps.newEnumMap(DyeColor.class);
    private final DyeColor dyeColor;
@@ -23,7 +_,15 @@
          SheepEntity sheepentity = (SheepEntity)p_111207_3_;
          if (sheepentity.isAlive() && !sheepentity.isSheared() && sheepentity.getColor() != this.dyeColor) {
             if (!p_111207_2_.level.isClientSide) {
-               sheepentity.setColor(this.dyeColor);
+               // CraftBukkit start
+               byte bColor = (byte) this.dyeColor.getId();
+               SheepDyeWoolEvent event = new SheepDyeWoolEvent((org.bukkit.entity.Sheep) sheepentity.getBukkitEntity(), org.bukkit.DyeColor.getByWoolData(bColor));
+               sheepentity.level.getCBServer().getPluginManager().callEvent(event);
+               if (event.isCancelled()) {
+                  return ActionResultType.PASS;
+               }
+               sheepentity.setColor(DyeColor.byId(event.getColor().getWoolData()));
+               // CraftBukkit end
                p_111207_1_.shrink(1);
             }
 
