--- a/net/minecraft/item/EnderEyeItem.java
+++ b/net/minecraft/item/EnderEyeItem.java
@@ -9,6 +_,7 @@
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.EyeOfEnderEntity;
+import net.minecraft.network.play.server.SPlaySoundEventPacket;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
 import net.minecraft.util.ActionResultType;
@@ -52,7 +_,25 @@
                   }
                }
 
-               world.globalLevelEvent(1038, blockpos1.offset(1, 0, 1), 0);
+               // CraftBukkit start - Use relative location for far away sounds
+               //  world.playBroadcastSound(1038, blockpos1.add(1, 0, 1), 0);
+               int viewDistance = world.getCBServer().getViewDistance() * 16;
+               BlockPos soundPos = blockpos1.offset(1, 0, 1);
+               for (ServerPlayerEntity player : world.getServer().getServer().getPlayerList().players) {
+                  double deltaX = soundPos.getX() - player.getX();
+                  double deltaZ = soundPos.getZ() - player.getZ();
+                  double distanceSquared = deltaX * deltaX + deltaZ * deltaZ;
+                  if (world.spigotConfig.endPortalSoundRadius > 0 && distanceSquared > world.spigotConfig.endPortalSoundRadius * world.spigotConfig.endPortalSoundRadius) continue; // Spigot
+                  if (distanceSquared > viewDistance * viewDistance) {
+                     double deltaLength = Math.sqrt(distanceSquared);
+                     double relativeX = player.getX() + (deltaX / deltaLength) * viewDistance;
+                     double relativeZ = player.getZ() + (deltaZ / deltaLength) * viewDistance;
+                     player.connection.send(new SPlaySoundEventPacket(1038, new BlockPos((int) relativeX, (int) soundPos.getY(), (int) relativeZ), 0, true));
+                  } else {
+                     player.connection.send(new SPlaySoundEventPacket(1038, soundPos, 0, true));
+                  }
+               }
+               // CraftBukkit end
             }
 
             return ActionResultType.CONSUME;
@@ -75,7 +_,11 @@
                EyeOfEnderEntity eyeofenderentity = new EyeOfEnderEntity(p_77659_1_, p_77659_2_.getX(), p_77659_2_.getY(0.5D), p_77659_2_.getZ());
                eyeofenderentity.setItem(itemstack);
                eyeofenderentity.signalTo(blockpos);
-               p_77659_1_.addFreshEntity(eyeofenderentity);
+               // CraftBukkit start
+               if (!p_77659_1_.addFreshEntity(eyeofenderentity)) {
+                  return new ActionResult<>(ActionResultType.FAIL, itemstack);
+               }
+               // CraftBukkit end
                if (p_77659_2_ instanceof ServerPlayerEntity) {
                   CriteriaTriggers.USED_ENDER_EYE.trigger((ServerPlayerEntity)p_77659_2_, blockpos);
                }
