--- a/net/minecraft/item/EnderPearlItem.java
+++ b/net/minecraft/item/EnderPearlItem.java
@@ -2,8 +_,10 @@
 
 import net.minecraft.entity.item.EnderPearlEntity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.stats.Stats;
 import net.minecraft.util.ActionResult;
+import net.minecraft.util.ActionResultType;
 import net.minecraft.util.Hand;
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvents;
@@ -16,14 +_,22 @@
 
    public ActionResult<ItemStack> use(World p_77659_1_, PlayerEntity p_77659_2_, Hand p_77659_3_) {
       ItemStack itemstack = p_77659_2_.getItemInHand(p_77659_3_);
-      p_77659_1_.playSound((PlayerEntity)null, p_77659_2_.getX(), p_77659_2_.getY(), p_77659_2_.getZ(), SoundEvents.ENDER_PEARL_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F));
-      p_77659_2_.getCooldowns().addCooldown(this, 20);
+
+      // CraftBukkit start - change order
       if (!p_77659_1_.isClientSide) {
          EnderPearlEntity enderpearlentity = new EnderPearlEntity(p_77659_1_, p_77659_2_);
          enderpearlentity.setItem(itemstack);
          enderpearlentity.shootFromRotation(p_77659_2_, p_77659_2_.xRot, p_77659_2_.yRot, 0.0F, 1.5F, 1.0F);
-         p_77659_1_.addFreshEntity(enderpearlentity);
-      }
+         if (!p_77659_1_.addFreshEntity(enderpearlentity)) {
+            if (p_77659_2_ instanceof ServerPlayerEntity) {
+               ((ServerPlayerEntity)p_77659_2_).getBukkitEntity().updateInventory();
+            }
+            return new ActionResult<ItemStack>(ActionResultType.FAIL, itemstack);
+      }
+      }
+      p_77659_1_.playSound((PlayerEntity)null, p_77659_2_.getX(), p_77659_2_.getY(), p_77659_2_.getZ(), SoundEvents.ENDER_PEARL_THROW, SoundCategory.NEUTRAL, 0.5F, 0.4F / (random.nextFloat() * 0.4F + 0.8F));
+      p_77659_2_.getCooldowns().addCooldown(this, 20);
+      // CraftBukkit end
 
       p_77659_2_.awardStat(Stats.ITEM_USED.get(this));
       if (!p_77659_2_.abilities.instabuild) {
