--- a/net/minecraft/item/FilledMapItem.java
+++ b/net/minecraft/item/FilledMapItem.java
@@ -33,6 +_,8 @@
 import net.minecraft.world.storage.MapData;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.Bukkit;
+import org.bukkit.event.server.MapInitializeEvent;
 
 public class FilledMapItem extends AbstractMapItem {
    public FilledMapItem(Item.Properties p_i48482_1_) {
@@ -52,6 +_,22 @@
 
    @Nullable
    public static MapData getOrCreateSavedData(ItemStack p_195950_0_, World p_195950_1_) {
+      // FORGE: Add instance method for mods to override
+      Item map = p_195950_0_.getItem();
+      if (map instanceof FilledMapItem) {
+        return ((FilledMapItem)map).getCustomMapData(p_195950_0_, p_195950_1_);
+      }
+      return null;
+   }
+
+   @Nullable
+   protected MapData getCustomMapData(ItemStack p_195950_0_, World p_195950_1_) {
+      // LoliServer start
+      return getCustomMapData_static(p_195950_0_, p_195950_1_);
+   }
+
+   @Nullable
+   public static MapData getCustomMapData_static(ItemStack p_195950_0_, World p_195950_1_) {
       MapData mapdata = getSavedData(p_195950_0_, p_195950_1_);
       if (mapdata == null && p_195950_1_ instanceof ServerWorld) {
          mapdata = createAndStoreSavedData(p_195950_0_, p_195950_1_, p_195950_1_.getLevelData().getXSpawn(), p_195950_1_.getLevelData().getZSpawn(), 3, false, false, p_195950_1_.dimension());
@@ -59,10 +_,11 @@
 
       return mapdata;
    }
+   // LoliServer end
 
    public static int getMapId(ItemStack p_195949_0_) {
       CompoundNBT compoundnbt = p_195949_0_.getTag();
-      return compoundnbt != null && compoundnbt.contains("map", 99) ? compoundnbt.getInt("map") : 0;
+      return compoundnbt != null && compoundnbt.contains("map", 99) ? compoundnbt.getInt("map") : -1; // CraftBukkit - make new maps for no tag
    }
 
    private static MapData createAndStoreSavedData(ItemStack p_195951_0_, World p_195951_1_, int p_195951_2_, int p_195951_3_, int p_195951_4_, boolean p_195951_5_, boolean p_195951_6_, RegistryKey<World> p_195951_7_) {
@@ -71,6 +_,11 @@
       mapdata.setProperties(p_195951_2_, p_195951_3_, p_195951_4_, p_195951_5_, p_195951_6_, p_195951_7_);
       p_195951_1_.setMapData(mapdata);
       p_195951_0_.getOrCreateTag().putInt("map", i);
+
+      // CraftBukkit start
+      MapInitializeEvent event = new MapInitializeEvent(mapdata.mapView);
+      Bukkit.getServer().getPluginManager().callEvent(event);
+      // CraftBukkit end
       return mapdata;
    }
 
@@ -226,7 +_,7 @@
 
             for(int l = 0; l < 128 * i; ++l) {
                for(int i1 = 0; i1 < 128 * i; ++i1) {
-                  abiome[l * 128 * i + i1] = p_226642_0_.getBiome(new BlockPos((j / i - 64) * i + i1, 0, (k / i - 64) * i + l));
+                  abiome[l * 128 * i + i1] = p_226642_0_.getBiomeBySeed((j / i - 64) * i + i1, 0, (k / i - 64) * i + l); // Paper
                }
             }
 
