--- a/net/minecraft/item/FireChargeItem.java
+++ b/net/minecraft/item/FireChargeItem.java
@@ -9,6 +_,8 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.block.BlockIgniteEvent;
 
 public class FireChargeItem extends Item {
    public FireChargeItem(Item.Properties p_i48499_1_) {
@@ -21,12 +_,28 @@
       BlockState blockstate = world.getBlockState(blockpos);
       boolean flag = false;
       if (CampfireBlock.canLight(blockstate)) {
+         // CraftBukkit start - fire BlockIgniteEvent
+         if (CraftEventFactory.callBlockIgniteEvent(world, blockpos, BlockIgniteEvent.IgniteCause.FIREBALL, p_195939_1_.getPlayer()).isCancelled()) {
+            if (!p_195939_1_.getPlayer().abilities.instabuild) {
+               p_195939_1_.getItemInHand().shrink(1);
+            }
+            return ActionResultType.PASS;
+         }
+         // CraftBukkit end
          this.playSound(world, blockpos);
          world.setBlockAndUpdate(blockpos, blockstate.setValue(CampfireBlock.LIT, Boolean.valueOf(true)));
          flag = true;
       } else {
          blockpos = blockpos.relative(p_195939_1_.getClickedFace());
          if (AbstractFireBlock.canBePlacedAt(world, blockpos, p_195939_1_.getHorizontalDirection())) {
+            // CraftBukkit start - fire BlockIgniteEvent
+            if (CraftEventFactory.callBlockIgniteEvent(world, blockpos, BlockIgniteEvent.IgniteCause.FIREBALL, p_195939_1_.getPlayer()).isCancelled()) {
+               if (!p_195939_1_.getPlayer().abilities.instabuild) {
+                  p_195939_1_.getItemInHand().shrink(1);
+               }
+               return ActionResultType.PASS;
+            }
+            // CraftBukkit end
             this.playSound(world, blockpos);
             world.setBlockAndUpdate(blockpos, AbstractFireBlock.getState(world, blockpos));
             flag = true;
