--- a/net/minecraft/item/FlintAndSteelItem.java
+++ b/net/minecraft/item/FlintAndSteelItem.java
@@ -23,7 +_,7 @@
       World world = p_195939_1_.getLevel();
       BlockPos blockpos = p_195939_1_.getClickedPos();
       BlockState blockstate = world.getBlockState(blockpos);
-      if (CampfireBlock.canLight(blockstate)) {
+      if (CampfireBlock.canLight(blockstate) && !org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockIgniteEvent(world, blockpos, playerentity).isCancelled()) { // CraftBukkit
          world.playSound(playerentity, blockpos, SoundEvents.FLINTANDSTEEL_USE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.4F + 0.8F);
          world.setBlock(blockpos, blockstate.setValue(BlockStateProperties.LIT, Boolean.valueOf(true)), 11);
          if (playerentity != null) {
@@ -36,6 +_,12 @@
       } else {
          BlockPos blockpos1 = blockpos.relative(p_195939_1_.getClickedFace());
          if (AbstractFireBlock.canBePlacedAt(world, blockpos1, p_195939_1_.getHorizontalDirection())) {
+            // CraftBukkit start - Store the clicked block
+            if (org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callBlockIgniteEvent(world, blockpos1, org.bukkit.event.block.BlockIgniteEvent.IgniteCause.FLINT_AND_STEEL, playerentity).isCancelled()) {
+               p_195939_1_.getItemInHand().hurtAndBreak(1, playerentity, (entityhuman1) -> entityhuman1.broadcastBreakEvent(p_195939_1_.getHand()));
+               return ActionResultType.PASS;
+            }
+            // CraftBukkit end
             world.playSound(playerentity, blockpos1, SoundEvents.FLINTANDSTEEL_USE, SoundCategory.BLOCKS, 1.0F, random.nextFloat() * 0.4F + 0.8F);
             BlockState blockstate1 = AbstractFireBlock.getState(world, blockpos1);
             world.setBlock(blockpos1, blockstate1, 11);
