--- a/net/minecraft/item/PotionItem.java
+++ b/net/minecraft/item/PotionItem.java
@@ -42,7 +_,7 @@
             if (effectinstance.getEffect().isInstantenous()) {
                effectinstance.getEffect().applyInstantenousEffect(playerentity, playerentity, p_77654_3_, effectinstance.getAmplifier(), 1.0D);
             } else {
-               p_77654_3_.addEffect(new EffectInstance(effectinstance));
+               p_77654_3_.addEffect(new EffectInstance(effectinstance), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.POTION_DRINK); // CraftBukkit
             }
          }
       }
