--- a/net/minecraft/item/ShearsItem.java
+++ b/net/minecraft/item/ShearsItem.java
@@ -7,6 +_,7 @@
 import net.minecraft.tags.BlockTags;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ShearsItem extends Item {
    public ShearsItem(Item.Properties p_i48471_1_) {
@@ -33,5 +_,32 @@
       } else {
          return 15.0F;
       }
+   }
+
+   @Override
+   public net.minecraft.util.ActionResultType interactLivingEntity(ItemStack stack, net.minecraft.entity.player.PlayerEntity playerIn, LivingEntity entity, net.minecraft.util.Hand hand) {
+      if (entity.level.isClientSide) return net.minecraft.util.ActionResultType.PASS;
+      if (entity instanceof net.minecraftforge.common.IForgeShearable) {
+          net.minecraftforge.common.IForgeShearable target = (net.minecraftforge.common.IForgeShearable)entity;
+         BlockPos pos = new BlockPos(entity.getX(), entity.getY(), entity.getZ());
+         if (target.isShearable(stack, entity.level, pos)) {
+            // CraftBukkit start
+            if (!CraftEventFactory.handlePlayerShearEntityEvent(playerIn, entity, stack, hand)) {
+               return net.minecraft.util.ActionResultType.PASS;
+            }
+            // CraftBukkit end
+            java.util.List<ItemStack> drops = target.onSheared(playerIn, stack, entity.level, pos, net.minecraft.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.enchantment.Enchantments.BLOCK_FORTUNE, stack));
+            java.util.Random rand = new java.util.Random();
+            drops.forEach(d -> {
+               entity.forceDrops = true; // CraftBukkit
+               net.minecraft.entity.item.ItemEntity ent = entity.spawnAtLocation(d, 1.0F);
+               entity.forceDrops = false; // CraftBukkit
+               ent.setDeltaMovement(ent.getDeltaMovement().add((double)((rand.nextFloat() - rand.nextFloat()) * 0.1F), (double)(rand.nextFloat() * 0.05F), (double)((rand.nextFloat() - rand.nextFloat()) * 0.1F)));
+            });
+            stack.hurtAndBreak(1, entity, e -> e.broadcastBreakEvent(hand));
+         }
+         return net.minecraft.util.ActionResultType.SUCCESS;
+      }
+      return net.minecraft.util.ActionResultType.PASS;
    }
 }
