--- a/net/minecraft/item/TridentItem.java
+++ b/net/minecraft/item/TridentItem.java
@@ -1,8 +_,8 @@
 package net.minecraft.item;
 
 import com.google.common.collect.ImmutableMultimap;
-import com.google.common.collect.Multimap;
 import com.google.common.collect.ImmutableMultimap.Builder;
+import com.google.common.collect.Multimap;
 import net.minecraft.block.BlockState;
 import net.minecraft.enchantment.EnchantmentHelper;
 import net.minecraft.enchantment.IVanishable;
@@ -12,6 +_,7 @@
 import net.minecraft.entity.ai.attributes.AttributeModifier;
 import net.minecraft.entity.ai.attributes.Attributes;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.entity.projectile.AbstractArrowEntity;
 import net.minecraft.entity.projectile.TridentEntity;
 import net.minecraft.inventory.EquipmentSlotType;
@@ -67,7 +_,19 @@
                         tridententity.pickup = AbstractArrowEntity.PickupStatus.CREATIVE_ONLY;
                      }
 
-                     p_77615_2_.addFreshEntity(tridententity);
+                     // CraftBukkit start
+                     if (!p_77615_2_.addFreshEntity(tridententity)) {
+                        if (playerentity instanceof ServerPlayerEntity) {
+                           ((ServerPlayerEntity) playerentity).getBukkitEntity().updateInventory();
+                        }
+                        return;
+                     }
+                     p_77615_1_.hurtAndBreak(1, playerentity, (p_220047_1_) -> {
+                        p_220047_1_.broadcastBreakEvent(p_77615_3_.getUsedItemHand());
+                     });
+                     tridententity.tridentItem = p_77615_1_.copy(); // SPIGOT-4511 update since damage call moved
+                     // CraftBukkit end
+
                      p_77615_2_.playSound((PlayerEntity)null, tridententity, SoundEvents.TRIDENT_THROW, SoundCategory.PLAYERS, 1.0F, 1.0F);
                      if (!playerentity.abilities.instabuild) {
                         playerentity.inventory.removeItem(p_77615_1_);
@@ -77,6 +_,10 @@
 
                playerentity.awardStat(Stats.ITEM_USED.get(this));
                if (j > 0) {
+                  // CraftBukkit start
+                  org.bukkit.event.player.PlayerRiptideEvent event = new org.bukkit.event.player.PlayerRiptideEvent((org.bukkit.entity.Player) playerentity.getBukkitEntity(), org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack.asCraftMirror(p_77615_1_));
+                  event.getPlayer().getServer().getPluginManager().callEvent(event);
+                  // CraftBukkit end
                   float f7 = playerentity.yRot;
                   float f = playerentity.xRot;
                   float f1 = -MathHelper.sin(f7 * ((float)Math.PI / 180F)) * MathHelper.cos(f * ((float)Math.PI / 180F));
