--- a/net/minecraft/command/impl/GameModeCommand.java
+++ b/net/minecraft/command/impl/GameModeCommand.java
@@ -11,6 +_,7 @@
 import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.util.Util;
 import net.minecraft.util.text.ITextComponent;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.GameType;
@@ -54,6 +_,12 @@
       for(ServerPlayerEntity serverplayerentity : p_198484_1_) {
          if (serverplayerentity.gameMode.getGameModeForPlayer() != p_198484_2_) {
             serverplayerentity.setGameMode(p_198484_2_);
+            // CraftBukkit start - handle event cancelling the change
+            if (serverplayerentity.gameMode.getGameModeForPlayer() != p_198484_2_) {
+               p_198484_0_.getSource().sendFailure(new StringTextComponent("Failed to set the gamemode of '" + serverplayerentity.getName() + "'"));
+               continue;
+            }
+            // CraftBukkit end
             logGamemodeChange(p_198484_0_.getSource(), serverplayerentity, p_198484_2_);
             ++i;
          }
