--- a/net/minecraft/command/impl/GameRuleCommand.java
+++ b/net/minecraft/command/impl/GameRuleCommand.java
@@ -27,14 +_,15 @@
 
    private static <T extends GameRules.RuleValue<T>> int setRule(CommandContext<CommandSource> p_223485_0_, GameRules.RuleKey<T> p_223485_1_) {
       CommandSource commandsource = p_223485_0_.getSource();
-      T t = commandsource.getServer().getGameRules().getRule(p_223485_1_);
+      T t = commandsource.getLevel().getGameRules().getRule(p_223485_1_); // CraftBukkit
+      t.setPaperRuleKey(p_223485_1_);
       t.setFromArgument(p_223485_0_, "value");
       commandsource.sendSuccess(new TranslationTextComponent("commands.gamerule.set", p_223485_1_.getId(), t.toString()), true);
       return t.getCommandResult();
    }
 
    private static <T extends GameRules.RuleValue<T>> int queryRule(CommandSource p_223486_0_, GameRules.RuleKey<T> p_223486_1_) {
-      T t = p_223486_0_.getServer().getGameRules().getRule(p_223486_1_);
+      T t = p_223486_0_.getLevel().getGameRules().getRule(p_223486_1_); // CraftBukkit
       p_223486_0_.sendSuccess(new TranslationTextComponent("commands.gamerule.query", p_223486_1_.getId(), t.toString()), false);
       return t.getCommandResult();
    }
