--- a/net/minecraft/command/impl/ReloadCommand.java
+++ b/net/minecraft/command/impl/ReloadCommand.java
@@ -3,6 +_,8 @@
 import com.google.common.collect.Lists;
 import com.mojang.brigadier.CommandDispatcher;
 import java.util.Collection;
+
+import io.papermc.paper.event.server.ServerResourcesReloadedEvent;
 import net.minecraft.command.CommandSource;
 import net.minecraft.command.Commands;
 import net.minecraft.resources.ResourcePackList;
@@ -16,7 +_,7 @@
    private static final Logger LOGGER = LogManager.getLogger();
 
    public static void reloadPacks(Collection<String> p_241062_0_, CommandSource p_241062_1_) {
-      p_241062_1_.getServer().reloadResources(p_241062_0_).exceptionally((p_241061_1_) -> {
+      p_241062_1_.getServer().reloadServerResources(p_241062_0_, ServerResourcesReloadedEvent.Cause.COMMAND).exceptionally((p_241061_1_) -> {
          LOGGER.warn("Failed to execute reload", p_241061_1_);
          p_241062_1_.sendFailure(new TranslationTextComponent("commands.reload.failure"));
          return null;
@@ -36,6 +_,17 @@
 
       return collection;
    }
+
+   // CraftBukkit start
+   public static void reload(MinecraftServer minecraftserver) {
+      ResourcePackList resourcepackrepository = minecraftserver.getPackRepository();
+      IServerConfiguration savedata = minecraftserver.getWorldData();
+      Collection<String> collection = resourcepackrepository.getSelectedIds();
+      Collection<String> collection1 = discoverNewPacks(resourcepackrepository, savedata, collection);
+      //minecraftserver.reloadResources(collection1);
+      minecraftserver.reloadServerResources(collection1, ServerResourcesReloadedEvent.Cause.PLUGIN); // Paper
+   }
+   // CraftBukkit end
 
    public static void register(CommandDispatcher<CommandSource> p_198597_0_) {
       p_198597_0_.register(Commands.literal("reload").requires((p_198599_0_) -> {
