--- a/net/minecraft/command/impl/SummonCommand.java
+++ b/net/minecraft/command/impl/SummonCommand.java
@@ -21,6 +_,7 @@
 import net.minecraft.util.text.TranslationTextComponent;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class SummonCommand {
    private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(new TranslationTextComponent("commands.summon.failed"));
@@ -58,7 +_,7 @@
                ((MobEntity)entity).finalizeSpawn(p_198737_0_.getLevel(), p_198737_0_.getLevel().getCurrentDifficultyAt(entity.blockPosition()), SpawnReason.COMMAND, (ILivingEntityData)null, (CompoundNBT)null);
             }
 
-            if (!serverworld.tryAddFreshEntityWithPassengers(entity)) {
+            if (!serverworld.addAllEntitiesSafely(entity, CreatureSpawnEvent.SpawnReason.COMMAND)) { // CraftBukkit - pass a spawn reason of "COMMAND"
                throw ERROR_DUPLICATE_UUID.create();
             } else {
                p_198737_0_.sendSuccess(new TranslationTextComponent("commands.summon.success", entity.getDisplayName()), true);
