--- a/net/minecraft/world/gen/feature/structure/SwampHutPiece.java
+++ b/net/minecraft/world/gen/feature/structure/SwampHutPiece.java
@@ -94,7 +_,7 @@
                witchentity.setPersistenceRequired();
                witchentity.moveTo((double)l + 0.5D, (double)i1, (double)k + 0.5D, 0.0F, 0.0F);
                witchentity.finalizeSpawn(p_230383_1_, p_230383_1_.getCurrentDifficultyAt(new BlockPos(l, i1, k)), SpawnReason.STRUCTURE, (ILivingEntityData)null, (CompoundNBT)null);
-               p_230383_1_.addFreshEntityWithPassengers(witchentity);
+               p_230383_1_.addAllEntities(witchentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
             }
          }
 
@@ -114,7 +_,7 @@
             catentity.setPersistenceRequired();
             catentity.moveTo((double)i + 0.5D, (double)j, (double)k + 0.5D, 0.0F, 0.0F);
             catentity.finalizeSpawn(p_214821_1_, p_214821_1_.getCurrentDifficultyAt(new BlockPos(i, j, k)), SpawnReason.STRUCTURE, (ILivingEntityData)null, (CompoundNBT)null);
-            p_214821_1_.addFreshEntityWithPassengers(catentity);
+            p_214821_1_.addAllEntities(catentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN); // CraftBukkit - add SpawnReason
          }
       }
 
