--- a/net/minecraft/entity/ai/goal/DefendVillageTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/DefendVillageTargetGoal.java
@@ -8,6 +_,7 @@
 import net.minecraft.entity.passive.IronGolemEntity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.util.math.AxisAlignedBB;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class DefendVillageTargetGoal extends TargetGoal {
    private final IronGolemEntity golem;
@@ -44,7 +_,7 @@
    }
 
    public void start() {
-      this.golem.setTarget(this.potentialTarget);
+      this.golem.setGoalTarget(this.potentialTarget, TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
       super.start();
    }
 }
