--- a/net/minecraft/entity/ai/goal/EatGrassGoal.java
+++ b/net/minecraft/entity/ai/goal/EatGrassGoal.java
@@ -10,6 +_,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class EatGrassGoal extends Goal {
    private static final Predicate<BlockState> IS_TALL_GRASS = BlockStateMatcher.forBlock(Blocks.GRASS);
@@ -59,7 +_,8 @@
       if (this.eatAnimationTick == 4) {
          BlockPos blockpos = this.mob.blockPosition();
          if (IS_TALL_GRASS.test(this.level.getBlockState(blockpos))) {
-            if (this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+            // CraftBukkit
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level, this.mob) || !CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockpos, Blocks.AIR.defaultBlockState(), !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) {
                this.level.destroyBlock(blockpos, false);
             }
 
@@ -67,7 +_,8 @@
          } else {
             BlockPos blockpos1 = blockpos.below();
             if (this.level.getBlockState(blockpos1).is(Blocks.GRASS_BLOCK)) {
-               if (this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+               // CraftBukkit
+               if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level, this.mob) || !CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockpos, Blocks.AIR.defaultBlockState(), !this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)).isCancelled()) {
                   this.level.levelEvent(2001, blockpos1, Block.getId(Blocks.GRASS_BLOCK.defaultBlockState()));
                   this.level.setBlock(blockpos1, Blocks.DIRT.defaultBlockState(), 2);
                }
