--- a/net/minecraft/entity/ai/goal/FollowOwnerGoal.java
+++ b/net/minecraft/entity/ai/goal/FollowOwnerGoal.java
@@ -12,6 +_,9 @@
 import net.minecraft.pathfinding.WalkNodeProcessor;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IWorldReader;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftEntity;
+import org.bukkit.event.entity.EntityTeleportEvent;
 
 public class FollowOwnerGoal extends Goal {
    private final TameableEntity tamable;
@@ -113,7 +_,17 @@
       } else if (!this.canTeleportTo(new BlockPos(p_226328_1_, p_226328_2_, p_226328_3_))) {
          return false;
       } else {
-         this.tamable.moveTo((double)p_226328_1_ + 0.5D, (double)p_226328_2_, (double)p_226328_3_ + 0.5D, this.tamable.yRot, this.tamable.xRot);
+         // CraftBukkit start
+         CraftEntity entity = this.tamable.getBukkitEntity();
+         Location to = new Location(entity.getWorld(), (double) ((float) p_226328_1_ + 0.5f), (double) p_226328_2_, (double) ((float) p_226328_3_ + 0.5f), this.tamable.yRot, this.tamable.xRot);
+         EntityTeleportEvent event = new EntityTeleportEvent(entity, entity.getLocation(), to);
+         this.tamable.level.getCBServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return false;
+         }
+         to = event.getTo();
+         this.tamable.moveTo(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
+         // CraftBukkit end
          this.navigation.stop();
          return true;
       }
