--- a/net/minecraft/entity/ai/goal/HurtByTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/HurtByTargetGoal.java
@@ -52,7 +_,7 @@
    }
 
    public void start() {
-      this.mob.setTarget(this.mob.getLastHurtByMob());
+      this.mob.setGoalTarget(this.mob.getLastHurtByMob(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit - reason
       this.targetMob = this.mob.getTarget();
       this.timestamp = this.mob.getLastHurtByMobTimestamp();
       this.unseenMemoryTicks = 300;
@@ -102,6 +_,6 @@
    }
 
    protected void alertOther(MobEntity p_220793_1_, LivingEntity p_220793_2_) {
-      p_220793_1_.setTarget(p_220793_2_);
+      p_220793_1_.setGoalTarget(p_220793_2_, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit - reason
    }
 }
