--- a/net/minecraft/entity/ai/goal/NearestAttackableTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/NearestAttackableTargetGoal.java
@@ -55,7 +_,7 @@
    }
 
    public void start() {
-      this.mob.setTarget(this.target);
+     this.mob.setGoalTarget(this.target, target instanceof ServerPlayerEntity ? org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_PLAYER : org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit - reason
       super.start();
    }
 
