--- a/net/minecraft/entity/ai/goal/OwnerHurtTargetGoal.java
+++ b/net/minecraft/entity/ai/goal/OwnerHurtTargetGoal.java
@@ -32,7 +_,7 @@
    }
 
    public void start() {
-      this.mob.setTarget(this.ownerLastHurt);
+      this.mob.setGoalTarget(this.ownerLastHurt, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit - reason
       LivingEntity livingentity = this.tameAnimal.getOwner();
       if (livingentity != null) {
          this.timestamp = livingentity.getLastHurtMobTimestamp();
