--- a/net/minecraft/entity/ai/goal/TemptGoal.java
+++ b/net/minecraft/entity/ai/goal/TemptGoal.java
@@ -1,13 +_,18 @@
 package net.minecraft.entity.ai.goal;
 
 import java.util.EnumSet;
+
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent;
+import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
+
 import net.minecraft.entity.CreatureEntity;
 import net.minecraft.entity.EntityPredicate;
+import net.minecraft.entity.LivingEntity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.crafting.Ingredient;
-import net.minecraft.pathfinding.FlyingPathNavigator;
-import net.minecraft.pathfinding.GroundPathNavigator;
 
 public class TemptGoal extends Goal {
    private static final EntityPredicate TEMP_TARGETING = (new EntityPredicate()).range(10.0D).allowInvulnerable().allowSameTeam().allowNonAttackable().allowUnseeable();
@@ -34,9 +_,9 @@
       this.items = p_i47823_5_;
       this.canScare = p_i47823_4_;
       this.setFlags(EnumSet.of(Goal.Flag.MOVE, Goal.Flag.LOOK));
-      if (!(p_i47823_1_.getNavigation() instanceof GroundPathNavigator) && !(p_i47823_1_.getNavigation() instanceof FlyingPathNavigator)) {
+      /* if (!(p_i47823_1_.getNavigation() instanceof GroundPathNavigator) && !(p_i47823_1_.getNavigation() instanceof FlyingPathNavigator)) {
          throw new IllegalArgumentException("Unsupported mob type for TemptGoal");
-      }
+      } */
    }
 
    public boolean canUse() {
@@ -45,12 +_,24 @@
          return false;
       } else {
          this.player = this.mob.level.getNearestPlayer(TEMP_TARGETING, this.mob);
-         if (this.player == null) {
-            return false;
-         } else {
-            return this.shouldFollowItem(this.player.getMainHandItem()) || this.shouldFollowItem(this.player.getOffhandItem());
+         // CraftBukkit start
+         boolean tempt = this.player != null && (this.shouldFollowItem(this.player.getMainHandItem()) || this.shouldFollowItem(this.player.getOffhandItem()));
+         if (player instanceof LivingEntity) {
+            if (tempt) {
+               EntityTargetLivingEntityEvent event = CraftEventFactory.callEntityTargetLivingEvent(this.mob, this.player, EntityTargetEvent.TargetReason.TEMPT);
+               if (event.isCancelled()) {
+                  return false;
+               }
+               quarkFix(((event.getTarget() == null) ? null : ((CraftPlayer) event.getTarget()).getHandle()));
+            }
          }
+         return tempt;
+         // CraftBukkit end
       }
+   }
+
+   public void quarkFix(PlayerEntity player) {
+      this.player = player;
    }
 
    protected boolean shouldFollowItem(ItemStack p_188508_1_) {
