--- a/net/minecraft/entity/ai/goal/TriggerSkeletonTrapGoal.java
+++ b/net/minecraft/entity/ai/goal/TriggerSkeletonTrapGoal.java
@@ -28,6 +_,10 @@
 
    public void tick() {
       ServerWorld serverworld = (ServerWorld)this.horse.level;
+      // Forge: Trigger the trap in a tick task to avoid crashes when mods add goals to skeleton horses
+      // (MC-206338/Forge PR #7509)
+      serverworld.getServer().tell(new net.minecraft.util.concurrent.TickDelayedTask(serverworld.getServer().getTickCount(), () -> {
+         if (!this.horse.isAlive()) return;
       DifficultyInstance difficultyinstance = serverworld.getCurrentDifficultyAt(this.horse.blockPosition());
       this.horse.setTrap(false);
       this.horse.setTamed(true);
@@ -35,19 +_,20 @@
       LightningBoltEntity lightningboltentity = EntityType.LIGHTNING_BOLT.create(serverworld);
       lightningboltentity.moveTo(this.horse.getX(), this.horse.getY(), this.horse.getZ());
       lightningboltentity.setVisualOnly(true);
-      serverworld.addFreshEntity(lightningboltentity);
+      ((ServerWorld)this.horse.level).strikeLightning(lightningboltentity, org.bukkit.event.weather.LightningStrikeEvent.Cause.TRAP); // CraftBukkit
       SkeletonEntity skeletonentity = this.createSkeleton(difficultyinstance, this.horse);
-      skeletonentity.startRiding(this.horse);
-      serverworld.addFreshEntityWithPassengers(skeletonentity);
+      if (skeletonentity != null) skeletonentity.startRiding(this.horse); // CraftBukkit
+      serverworld.addAllEntities(skeletonentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.TRAP); // CraftBukkit
 
       for(int i = 0; i < 3; ++i) {
          AbstractHorseEntity abstracthorseentity = this.createHorse(difficultyinstance);
+         if (abstracthorseentity == null) continue; // CraftBukkit
          SkeletonEntity skeletonentity1 = this.createSkeleton(difficultyinstance, abstracthorseentity);
-         skeletonentity1.startRiding(abstracthorseentity);
+         if (skeletonentity1 != null) skeletonentity1.startRiding(abstracthorseentity); // CraftBukkit
          abstracthorseentity.push(this.horse.getRandom().nextGaussian() * 0.5D, 0.0D, this.horse.getRandom().nextGaussian() * 0.5D);
-         serverworld.addFreshEntityWithPassengers(abstracthorseentity);
+         serverworld.addAllEntities(abstracthorseentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.JOCKEY); // CraftBukkit
       }
-
+      }));
    }
 
    private AbstractHorseEntity createHorse(DifficultyInstance p_188515_1_) {
