--- a/net/minecraft/util/registry/Bootstrap.java
+++ b/net/minecraft/util/registry/Bootstrap.java
@@ -22,6 +_,9 @@
 import net.minecraft.tags.TagRegistryManager;
 import net.minecraft.util.LoggingPrintStream;
 import net.minecraft.util.SharedConstants;
+import net.minecraft.util.datafix.fixes.BlockStateFlatteningMap;
+import net.minecraft.util.datafix.fixes.ItemIntIDToString;
+import net.minecraft.util.datafix.fixes.SpawnEggNames;
 import net.minecraft.util.text.LanguageMap;
 import net.minecraft.world.GameRules;
 import org.apache.logging.log4j.LogManager;
@@ -48,7 +_,68 @@
                IDispenseItemBehavior.bootStrap();
                ArgumentTypes.bootStrap();
                TagRegistryManager.bootStrap();
+               net.minecraftforge.registries.GameData.vanillaSnapshot();
+               if (false) // skip redirectOutputToLog, Forge already redirects stdout and stderr output to log so that they print with more context
                wrapStreams();
+               // CraftBukkit start - easier than fixing the decompile
+               BlockStateFlatteningMap.register(1008, "{Name:'minecraft:oak_sign',Properties:{rotation:'0'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'0'}}");
+               BlockStateFlatteningMap.register(1009, "{Name:'minecraft:oak_sign',Properties:{rotation:'1'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'1'}}");
+               BlockStateFlatteningMap.register(1010, "{Name:'minecraft:oak_sign',Properties:{rotation:'2'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'2'}}");
+               BlockStateFlatteningMap.register(1011, "{Name:'minecraft:oak_sign',Properties:{rotation:'3'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'3'}}");
+               BlockStateFlatteningMap.register(1012, "{Name:'minecraft:oak_sign',Properties:{rotation:'4'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'4'}}");
+               BlockStateFlatteningMap.register(1013, "{Name:'minecraft:oak_sign',Properties:{rotation:'5'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'5'}}");
+               BlockStateFlatteningMap.register(1014, "{Name:'minecraft:oak_sign',Properties:{rotation:'6'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'6'}}");
+               BlockStateFlatteningMap.register(1015, "{Name:'minecraft:oak_sign',Properties:{rotation:'7'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'7'}}");
+               BlockStateFlatteningMap.register(1016, "{Name:'minecraft:oak_sign',Properties:{rotation:'8'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'8'}}");
+               BlockStateFlatteningMap.register(1017, "{Name:'minecraft:oak_sign',Properties:{rotation:'9'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'9'}}");
+               BlockStateFlatteningMap.register(1018, "{Name:'minecraft:oak_sign',Properties:{rotation:'10'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'10'}}");
+               BlockStateFlatteningMap.register(1019, "{Name:'minecraft:oak_sign',Properties:{rotation:'11'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'11'}}");
+               BlockStateFlatteningMap.register(1020, "{Name:'minecraft:oak_sign',Properties:{rotation:'12'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'12'}}");
+               BlockStateFlatteningMap.register(1021, "{Name:'minecraft:oak_sign',Properties:{rotation:'13'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'13'}}");
+               BlockStateFlatteningMap.register(1022, "{Name:'minecraft:oak_sign',Properties:{rotation:'14'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'14'}}");
+               BlockStateFlatteningMap.register(1023, "{Name:'minecraft:oak_sign',Properties:{rotation:'15'}}", "{Name:'minecraft:standing_sign',Properties:{rotation:'15'}}");
+               ItemIntIDToString.ITEM_NAMES.put(323, "minecraft:oak_sign");
+               BlockStateFlatteningMap.register(1440, "{Name:'minecraft:portal',Properties:{axis:'x'}}", "{Name:'minecraft:portal',Properties:{axis:'x'}}");
+               ItemIntIDToString.ITEM_NAMES.put(409, "minecraft:prismarine_shard");
+               ItemIntIDToString.ITEM_NAMES.put(410, "minecraft:prismarine_crystals");
+               ItemIntIDToString.ITEM_NAMES.put(411, "minecraft:rabbit");
+               ItemIntIDToString.ITEM_NAMES.put(412, "minecraft:cooked_rabbit");
+               ItemIntIDToString.ITEM_NAMES.put(413, "minecraft:rabbit_stew");
+               ItemIntIDToString.ITEM_NAMES.put(414, "minecraft:rabbit_foot");
+               ItemIntIDToString.ITEM_NAMES.put(415, "minecraft:rabbit_hide");
+               ItemIntIDToString.ITEM_NAMES.put(416, "minecraft:armor_stand");
+               ItemIntIDToString.ITEM_NAMES.put(423, "minecraft:mutton");
+               ItemIntIDToString.ITEM_NAMES.put(424, "minecraft:cooked_mutton");
+               ItemIntIDToString.ITEM_NAMES.put(425, "minecraft:banner");
+               ItemIntIDToString.ITEM_NAMES.put(426, "minecraft:end_crystal");
+               ItemIntIDToString.ITEM_NAMES.put(427, "minecraft:spruce_door");
+               ItemIntIDToString.ITEM_NAMES.put(428, "minecraft:birch_door");
+               ItemIntIDToString.ITEM_NAMES.put(429, "minecraft:jungle_door");
+               ItemIntIDToString.ITEM_NAMES.put(430, "minecraft:acacia_door");
+               ItemIntIDToString.ITEM_NAMES.put(431, "minecraft:dark_oak_door");
+               ItemIntIDToString.ITEM_NAMES.put(432, "minecraft:chorus_fruit");
+               ItemIntIDToString.ITEM_NAMES.put(433, "minecraft:chorus_fruit_popped");
+               ItemIntIDToString.ITEM_NAMES.put(434, "minecraft:beetroot");
+               ItemIntIDToString.ITEM_NAMES.put(435, "minecraft:beetroot_seeds");
+               ItemIntIDToString.ITEM_NAMES.put(436, "minecraft:beetroot_soup");
+               ItemIntIDToString.ITEM_NAMES.put(437, "minecraft:dragon_breath");
+               ItemIntIDToString.ITEM_NAMES.put(438, "minecraft:splash_potion");
+               ItemIntIDToString.ITEM_NAMES.put(439, "minecraft:spectral_arrow");
+               ItemIntIDToString.ITEM_NAMES.put(440, "minecraft:tipped_arrow");
+               ItemIntIDToString.ITEM_NAMES.put(441, "minecraft:lingering_potion");
+               ItemIntIDToString.ITEM_NAMES.put(442, "minecraft:shield");
+               ItemIntIDToString.ITEM_NAMES.put(443, "minecraft:elytra");
+               ItemIntIDToString.ITEM_NAMES.put(444, "minecraft:spruce_boat");
+               ItemIntIDToString.ITEM_NAMES.put(445, "minecraft:birch_boat");
+               ItemIntIDToString.ITEM_NAMES.put(446, "minecraft:jungle_boat");
+               ItemIntIDToString.ITEM_NAMES.put(447, "minecraft:acacia_boat");
+               ItemIntIDToString.ITEM_NAMES.put(448, "minecraft:dark_oak_boat");
+               ItemIntIDToString.ITEM_NAMES.put(449, "minecraft:totem_of_undying");
+               ItemIntIDToString.ITEM_NAMES.put(450, "minecraft:shulker_shell");
+               ItemIntIDToString.ITEM_NAMES.put(452, "minecraft:iron_nugget");
+               ItemIntIDToString.ITEM_NAMES.put(453, "minecraft:knowledge_book");
+               SpawnEggNames.ID_TO_ENTITY[23] = "Arrow";
+               // CraftBukkit end
             }
          }
       }
@@ -103,7 +_,6 @@
             Commands.validate();
          }
 
-         GlobalEntityTypeAttributes.validate();
       }
    }
 
