--- a/net/minecraft/pathfinding/PathNavigator.java
+++ b/net/minecraft/pathfinding/PathNavigator.java
@@ -11,6 +_,7 @@
 import net.minecraft.entity.MobEntity;
 import net.minecraft.entity.ai.attributes.Attributes;
 import net.minecraft.network.DebugPacketSender;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.Util;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
@@ -142,9 +_,30 @@
       return this.moveTo(this.createPath(p_75492_1_, p_75492_3_, p_75492_5_, 1), p_75492_7_);
    }
 
+   // Paper start - Optimize Pathfinding
+   private int lastFailure = 0;
+   private int pathfindFailures = 0;
+   // Paper end
+
    public boolean moveTo(Entity p_75497_1_, double p_75497_2_) {
+      // Paper start - Optimize Pathfinding
+      if (this.pathfindFailures > 10 && this.path == null && MinecraftServer.currentTick < this.lastFailure + 40) {
+          return false;
+      }
+      // Paper end
       Path path = this.createPath(p_75497_1_, 1);
-      return path != null && this.moveTo(path, p_75497_2_);
+
+      // Paper start - Optimize Pathfinding
+      if (path != null && this.moveTo(path, p_75497_2_)) {
+          this.lastFailure = 0;
+          this.pathfindFailures = 0;
+          return true;
+      } else {
+          this.pathfindFailures++;
+          this.lastFailure = MinecraftServer.currentTick;
+          return false;
+      }
+      // Paper end
    }
 
    public boolean moveTo(@Nullable Path p_75484_1_, double p_75484_2_) {
@@ -208,10 +_,10 @@
       Vector3d vector3d = this.getTempMobPos();
       this.maxDistanceToWaypoint = this.mob.getBbWidth() > 0.75F ? this.mob.getBbWidth() / 2.0F : 0.75F - this.mob.getBbWidth() / 2.0F;
       Vector3i vector3i = this.path.getNextNodePos();
-      double d0 = Math.abs(this.mob.getX() - ((double)vector3i.getX() + 0.5D));
+      double d0 = Math.abs(this.mob.getX() - ((double)vector3i.getX() + (this.mob.getBbWidth() + 1) / 2D)); //Forge: Fix MC-94054
       double d1 = Math.abs(this.mob.getY() - (double)vector3i.getY());
-      double d2 = Math.abs(this.mob.getZ() - ((double)vector3i.getZ() + 0.5D));
-      boolean flag = d0 < (double)this.maxDistanceToWaypoint && d2 < (double)this.maxDistanceToWaypoint && d1 < 1.0D;
+      double d2 = Math.abs(this.mob.getZ() - ((double)vector3i.getZ() + (this.mob.getBbWidth() + 1) / 2D)); //Forge: Fix MC-94054
+      boolean flag = d0 <= (double)this.maxDistanceToWaypoint && d2 <= (double)this.maxDistanceToWaypoint && d1 < 1.0D; //Forge: Fix MC-94054
       if (flag || this.mob.canCutCorner(this.path.getNextNode().type) && this.shouldTargetNextNodeInDirection(vector3d)) {
          this.path.advance();
       }
