--- a/net/minecraft/server/integrated/IntegratedPlayerList.java
+++ b/net/minecraft/server/integrated/IntegratedPlayerList.java
@@ -30,7 +_,7 @@
    }
 
    public ITextComponent canPlayerLogin(SocketAddress p_206258_1_, GameProfile p_206258_2_) {
-      return (ITextComponent)(p_206258_2_.getName().equalsIgnoreCase(this.getServer().getSingleplayerName()) && this.getPlayerByName(p_206258_2_.getName()) != null ? new TranslationTextComponent("multiplayer.disconnect.name_taken") : super.canPlayerLogin(p_206258_1_, p_206258_2_));
+      return (ITextComponent)(p_206258_2_.getName().equalsIgnoreCase(this.getServer().getSingleplayerName()) && this.getPlayerByName(p_206258_2_.getName()) != null ? new TranslationTextComponent("multiplayer.disconnect.name_taken") : null);
    }
 
    public IntegratedServer getServer() {
