--- a/net/minecraft/entity/boss/dragon/phase/PhaseManager.java
+++ b/net/minecraft/entity/boss/dragon/phase/PhaseManager.java
@@ -3,6 +_,8 @@
 import net.minecraft.entity.boss.dragon.EnderDragonEntity;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftEnderDragon;
+import org.bukkit.event.entity.EnderDragonChangePhaseEvent;
 
 public class PhaseManager {
    private static final Logger LOGGER = LogManager.getLogger();
@@ -20,6 +_,16 @@
          if (this.currentPhase != null) {
             this.currentPhase.end();
          }
+
+         // CraftBukkit start - Call EnderDragonChangePhaseEvent
+         EnderDragonChangePhaseEvent event = new EnderDragonChangePhaseEvent((CraftEnderDragon) this.dragon.getBukkitEntity(),
+             (this.currentPhase == null) ? null : CraftEnderDragon.getBukkitPhase(this.currentPhase.getPhase()), CraftEnderDragon.getBukkitPhase(p_188758_1_));
+         this.dragon.level.getCBServer().getPluginManager().callEvent(event);
+         if (event.isCancelled()) {
+            return;
+         }
+         if (event.getNewPhase() != org.bukkit.entity.EnderDragon.Phase.CUSTOM) p_188758_1_ = CraftEnderDragon.getMinecraftPhase(event.getNewPhase()); // LoliServer
+         // CraftBukkit end
 
          this.currentPhase = this.getPhase(p_188758_1_);
          if (!this.dragon.level.isClientSide) {
