--- a/net/minecraft/entity/passive/horse/HorseEntity.java
+++ b/net/minecraft/entity/passive/horse/HorseEntity.java
@@ -89,7 +_,7 @@
       return this.entityData.get(DATA_ID_TYPE_VARIANT);
    }
 
-   private void setVariantAndMarkings(CoatColors p_234238_1_, CoatTypes p_234238_2_) {
+   public void setVariantAndMarkings(CoatColors p_234238_1_, CoatTypes p_234238_2_) {
       this.setTypeVariant(p_234238_1_.getId() & 255 | p_234238_2_.getId() << 8 & '\uff00');
    }
 
@@ -139,6 +_,8 @@
          this.playSound(SoundEvents.HORSE_BREATHE, p_190680_1_.getVolume() * 0.6F, p_190680_1_.getPitch());
       }
 
+      ItemStack stack = this.inventory.getItem(1);
+      if (isArmor(stack)) stack.onHorseArmorTick(level, this);
    }
 
    protected SoundEvent getAmbientSound() {
