--- a/net/minecraft/entity/passive/horse/TraderLlamaEntity.java
+++ b/net/minecraft/entity/passive/horse/TraderLlamaEntity.java
@@ -20,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class TraderLlamaEntity extends LlamaEntity {
    private int despawnDelay = 47999;
@@ -135,7 +_,7 @@
       }
 
       public void start() {
-         this.mob.setTarget(this.ownerLastHurtBy);
+         this.mob.setGoalTarget(this.ownerLastHurtBy, TargetReason.TARGET_ATTACKED_OWNER, true); // CraftBukkit
          Entity entity = this.llama.getLeashHolder();
          if (entity instanceof WanderingTraderEntity) {
             this.timestamp = ((WanderingTraderEntity)entity).getLastHurtByMobTimestamp();
