--- a/net/minecraft/world/BossInfo.java
+++ b/net/minecraft/world/BossInfo.java
@@ -6,10 +_,10 @@
 
 public abstract class BossInfo {
    private final UUID id;
-   protected ITextComponent name;
+   public ITextComponent name;
    protected float percent;
-   protected BossInfo.Color color;
-   protected BossInfo.Overlay overlay;
+   public BossInfo.Color color;
+   public BossInfo.Overlay overlay;
    protected boolean darkenScreen;
    protected boolean playBossMusic;
    protected boolean createWorldFog;
