--- a/net/minecraft/world/WorldSettings.java
+++ b/net/minecraft/world/WorldSettings.java
@@ -4,9 +_,9 @@
 import net.minecraft.util.datafix.codec.DatapackCodec;
 
 public final class WorldSettings {
-   private final String levelName;
+   public String levelName;
    private final GameType gameType;
-   private final boolean hardcore;
+   public boolean hardcore;
    private final Difficulty difficulty;
    private final boolean allowCommands;
    private final GameRules gameRules;
@@ -62,6 +_,7 @@
    }
 
    public WorldSettings withDifficulty(Difficulty p_234948_1_) {
+      net.minecraftforge.common.ForgeHooks.onDifficultyChange(p_234948_1_, this.difficulty);
       return new WorldSettings(this.levelName, this.gameType, this.hardcore, p_234948_1_, this.allowCommands, this.gameRules, this.dataPackConfig);
    }
 
