--- a/net/minecraft/world/chunk/listener/LoggingChunkStatusListener.java
+++ b/net/minecraft/world/chunk/listener/LoggingChunkStatusListener.java
@@ -26,15 +_,21 @@
       this.startTime = this.nextTickTime;
    }
 
+   private static int lastPercentage = -1;
+
    public void onStatusChange(ChunkPos p_219508_1_, @Nullable ChunkStatus p_219508_2_) {
       if (p_219508_2_ == ChunkStatus.FULL) {
          ++this.count;
       }
 
       int i = this.getProgress();
+      int calc = MathHelper.clamp(i, 0, 100);
       if (Util.getMillis() > this.nextTickTime) {
          this.nextTickTime += 500L;
-         LOGGER.info((new TranslationTextComponent("menu.preparingSpawn", MathHelper.clamp(i, 0, 100))).getString());
+         if(lastPercentage != calc){
+            LOGGER.info((new TranslationTextComponent("menu.preparingSpawn", MathHelper.clamp(i, 0, 100))).getString());
+            lastPercentage = calc;
+         }
       }
 
    }
