--- a/net/minecraft/block/AbstractButtonBlock.java
+++ b/net/minecraft/block/AbstractButtonBlock.java
@@ -23,6 +_,10 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+// CraftBukkit start
+import org.bukkit.event.block.BlockRedstoneEvent;
+import org.bukkit.event.entity.EntityInteractEvent;
+// CraftBukkit end
 
 public abstract class AbstractButtonBlock extends HorizontalFaceBlock {
    public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
@@ -90,6 +_,19 @@
       if (p_225533_1_.getValue(POWERED)) {
          return ActionResultType.CONSUME;
       } else {
+         // CraftBukkit start
+         boolean powered = ((Boolean) p_225533_1_.getValue(POWERED));
+         org.bukkit.block.Block block = p_225533_2_.getWorld().getBlockAt(p_225533_3_.getX(), p_225533_3_.getY(), p_225533_3_.getZ());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_225533_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((eventRedstone.getNewCurrent() > 0) != (!powered)) {
+            return ActionResultType.SUCCESS;
+         }
+         // CraftBukkit end
          this.press(p_225533_1_, p_225533_2_, p_225533_3_);
          this.playSound(p_225533_4_, p_225533_2_, p_225533_3_, true);
          return ActionResultType.sidedSuccess(p_225533_2_.isClientSide);
@@ -135,6 +_,16 @@
          if (this.sensitive) {
             this.checkPressed(p_225534_1_, p_225534_2_, p_225534_3_);
          } else {
+            // CraftBukkit start
+            org.bukkit.block.Block block = p_225534_2_.getWorld().getBlockAt(p_225534_3_.getX(), p_225534_3_.getY(), p_225534_3_.getZ());
+
+            BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, 15, 0);
+            p_225534_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+
+            if (eventRedstone.getNewCurrent() > 0) {
+               return;
+            }
+            // CraftBukkit end
             p_225534_2_.setBlock(p_225534_3_, p_225534_1_.setValue(POWERED, Boolean.valueOf(false)), 3);
             this.updateNeighbours(p_225534_1_, p_225534_2_, p_225534_3_);
             this.playSound((PlayerEntity)null, p_225534_2_, p_225534_3_, false);
@@ -153,7 +_,43 @@
       List<? extends Entity> list = p_185616_2_.getEntitiesOfClass(AbstractArrowEntity.class, p_185616_1_.getShape(p_185616_2_, p_185616_3_).bounds().move(p_185616_3_));
       boolean flag = !list.isEmpty();
       boolean flag1 = p_185616_1_.getValue(POWERED);
+      // CraftBukkit start - Call interact event when arrows turn on wooden buttons
+      if (flag1 != flag && flag) {
+         org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.getX(), p_185616_3_.getY(), p_185616_3_.getZ());
+         boolean allowed = false;
+
+         // If all of the events are cancelled block the button press, else allow
+         for (Object object : list) {
+            if (object != null) {
+               EntityInteractEvent event = new EntityInteractEvent(((Entity) object).getBukkitEntity(), block);
+               p_185616_2_.getCBServer().getPluginManager().callEvent(event);
+
+               if (!event.isCancelled()) {
+                  allowed = true;
+                  break;
+               }
+            }
+         }
+
+         if (!allowed) {
+            return;
+         }
+      }
+         // CraftBukkit end
       if (flag != flag1) {
+         // CraftBukkit start
+         boolean powered = flag1;
+         org.bukkit.block.Block block = p_185616_2_.getWorld().getBlockAt(p_185616_3_.getX(), p_185616_3_.getY(), p_185616_3_.getZ());
+         int old = (powered) ? 15 : 0;
+         int current = (!powered) ? 15 : 0;
+
+         BlockRedstoneEvent eventRedstone = new BlockRedstoneEvent(block, old, current);
+         p_185616_2_.getCBServer().getPluginManager().callEvent(eventRedstone);
+
+         if ((flag && eventRedstone.getNewCurrent() <= 0) || (!flag && eventRedstone.getNewCurrent() > 0)) {
+            return;
+         }
+         // CraftBukkit end
          p_185616_2_.setBlock(p_185616_3_, p_185616_1_.setValue(POWERED, Boolean.valueOf(flag)), 3);
          this.updateNeighbours(p_185616_1_, p_185616_2_, p_185616_3_);
          this.playSound((PlayerEntity)null, p_185616_2_, p_185616_3_, flag);
