--- a/net/minecraft/block/BambooSaplingBlock.java
+++ b/net/minecraft/block/BambooSaplingBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class BambooSaplingBlock extends Block implements IGrowable {
    protected static final VoxelShape SAPLING_SHAPE = Block.box(4.0D, 0.0D, 4.0D, 12.0D, 12.0D, 12.0D);
@@ -78,6 +_,6 @@
    }
 
    protected void growBamboo(World p_220087_1_, BlockPos p_220087_2_) {
-      p_220087_1_.setBlock(p_220087_2_.above(), Blocks.BAMBOO.defaultBlockState().setValue(BambooBlock.LEAVES, BambooLeaves.SMALL), 3);
+      CraftEventFactory.handleBlockSpreadEvent(p_220087_1_, p_220087_2_, p_220087_2_.above(), Blocks.BAMBOO.defaultBlockState().setValue(BambooBlock.LEAVES, BambooLeaves.SMALL), 3); // CraftBukkit - BlockSpreadEvent
    }
 }
