--- a/net/minecraft/block/DaylightDetectorBlock.java
+++ b/net/minecraft/block/DaylightDetectorBlock.java
@@ -18,6 +_,7 @@
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.LightType;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class DaylightDetectorBlock extends ContainerBlock {
    public static final IntegerProperty POWER = BlockStateProperties.POWER;
@@ -56,6 +_,7 @@
 
          i = MathHelper.clamp(i, 0, 15);
          if (p_196319_0_.getValue(POWER) != i) {
+            i = CraftEventFactory.callRedstoneChange(p_196319_1_, p_196319_2_, p_196319_0_.getValue(POWER), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
             p_196319_1_.setBlock(p_196319_2_, p_196319_0_.setValue(POWER, Integer.valueOf(i)), 3);
          }
 
