--- a/net/minecraft/block/DispenserBlock.java
+++ b/net/minecraft/block/DispenserBlock.java
@@ -40,9 +_,10 @@
 public class DispenserBlock extends ContainerBlock {
    public static final DirectionProperty FACING = DirectionalBlock.FACING;
    public static final BooleanProperty TRIGGERED = BlockStateProperties.TRIGGERED;
-   private static final Map<Item, IDispenseItemBehavior> DISPENSER_REGISTRY = Util.make(new Object2ObjectOpenHashMap<>(), (p_212564_0_) -> {
+   public static final Map<Item, IDispenseItemBehavior> DISPENSER_REGISTRY = Util.make(new Object2ObjectOpenHashMap<>(), (p_212564_0_) -> {
       p_212564_0_.defaultReturnValue(new DefaultDispenseItemBehavior());
    });
+   public static boolean eventFired = false; // CraftBukkit
 
    public static void registerBehavior(IItemProvider p_199774_0_, IDispenseItemBehavior p_199774_1_) {
       DISPENSER_REGISTRY.put(p_199774_0_.asItem(), p_199774_1_);
@@ -71,7 +_,7 @@
       }
    }
 
-   protected void dispenseFrom(ServerWorld p_176439_1_, BlockPos p_176439_2_) {
+   public void dispenseFrom(ServerWorld p_176439_1_, BlockPos p_176439_2_) {
       ProxyBlockSource proxyblocksource = new ProxyBlockSource(p_176439_1_, p_176439_2_);
       DispenserTileEntity dispensertileentity = proxyblocksource.getEntity();
       int i = dispensertileentity.getRandomSlot();
@@ -81,6 +_,7 @@
          ItemStack itemstack = dispensertileentity.getItem(i);
          IDispenseItemBehavior idispenseitembehavior = this.getDispenseMethod(itemstack);
          if (idispenseitembehavior != IDispenseItemBehavior.NOOP) {
+            eventFired = false; // CraftBukkit - reset event status
             dispensertileentity.setItem(i, idispenseitembehavior.dispense(proxyblocksource, itemstack));
          }
 
