--- a/net/minecraft/block/DoorBlock.java
+++ b/net/minecraft/block/DoorBlock.java
@@ -32,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class DoorBlock extends Block {
    public static final DirectionProperty FACING = HorizontalBlock.FACING;
@@ -177,8 +_,21 @@
    }
 
    public void neighborChanged(BlockState p_220069_1_, World p_220069_2_, BlockPos p_220069_3_, Block p_220069_4_, BlockPos p_220069_5_, boolean p_220069_6_) {
-      boolean flag = p_220069_2_.hasNeighborSignal(p_220069_3_) || p_220069_2_.hasNeighborSignal(p_220069_3_.relative(p_220069_1_.getValue(HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN));
-      if (p_220069_4_ != this && flag != p_220069_1_.getValue(POWERED)) {
+      // CraftBukkit start
+      BlockPos otherHalf = p_220069_3_.relative(p_220069_1_.getValue(HALF) == DoubleBlockHalf.LOWER ? Direction.UP : Direction.DOWN);
+      org.bukkit.World bworld = p_220069_2_.getWorld();
+      org.bukkit.block.Block bukkitBlock = bworld.getBlockAt(p_220069_3_.getX(), p_220069_3_.getY(), p_220069_3_.getZ());
+      org.bukkit.block.Block blockTop = bworld.getBlockAt(otherHalf.getX(), otherHalf.getY(), otherHalf.getZ());
+      int power = bukkitBlock.getBlockPower();
+      int powerTop = blockTop.getBlockPower();
+      if (powerTop > power) power = powerTop;
+      int oldPower = (Boolean) p_220069_1_.getValue(DoorBlock.POWERED) ? 15 : 0;
+      if(oldPower == 0 ^ power == 0){
+         BlockRedstoneEvent blockRedstoneEvent = new BlockRedstoneEvent(bukkitBlock, oldPower, power);
+         p_220069_2_.getCBServer().getPluginManager().callEvent(blockRedstoneEvent);
+
+         boolean flag = blockRedstoneEvent.getNewCurrent() > 0;
+         // CraftBukkit end
          if (flag != p_220069_1_.getValue(OPEN)) {
             this.playSound(p_220069_2_, p_220069_3_, flag);
          }
