--- a/net/minecraft/block/EndPortalBlock.java
+++ b/net/minecraft/block/EndPortalBlock.java
@@ -2,9 +_,11 @@
 
 import java.util.Random;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.fluid.Fluid;
 import net.minecraft.item.ItemStack;
 import net.minecraft.particles.ParticleTypes;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.tileentity.EndPortalTileEntity;
 import net.minecraft.tileentity.TileEntity;
 import net.minecraft.util.RegistryKey;
@@ -13,11 +_,14 @@
 import net.minecraft.util.math.shapes.ISelectionContext;
 import net.minecraft.util.math.shapes.VoxelShape;
 import net.minecraft.util.math.shapes.VoxelShapes;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPortalEnterEvent;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class EndPortalBlock extends ContainerBlock {
    protected static final VoxelShape SHAPE = Block.box(0.0D, 0.0D, 0.0D, 16.0D, 12.0D, 16.0D);
@@ -36,11 +_,36 @@
 
    public void entityInside(BlockState p_196262_1_, World p_196262_2_, BlockPos p_196262_3_, Entity p_196262_4_) {
       if (p_196262_2_ instanceof ServerWorld && !p_196262_4_.isPassenger() && !p_196262_4_.isVehicle() && p_196262_4_.canChangeDimensions() && VoxelShapes.joinIsNotEmpty(VoxelShapes.create(p_196262_4_.getBoundingBox().move((double)(-p_196262_3_.getX()), (double)(-p_196262_3_.getY()), (double)(-p_196262_3_.getZ()))), p_196262_1_.getShape(p_196262_2_, p_196262_3_), IBooleanFunction.AND)) {
-         RegistryKey<World> registrykey = p_196262_2_.dimension() == World.END ? World.OVERWORLD : World.END;
+         //LoliServer start
+         RegistryKey<World> originRegistrykey = p_196262_2_.dimension() == World.END ? World.OVERWORLD : World.END;
+         RegistryKey<World> originRegistrykey1 = p_196262_2_.dimension1() == World.END ? World.OVERWORLD : World.END;
+         //LoliServer end
+         //CraftBukkit start
+         RegistryKey<World> registrykey = p_196262_2_.dimensionType() == MinecraftServer.getServer().registryHolder.dimensionTypes().get(DimensionType.END_LOCATION) ? World.OVERWORLD : World.END; // CraftBukkit - SPIGOT-6152: send back to main overworld in custom ends
+         //CraftBukkit end
+         //LoliServer start
+         if (originRegistrykey != originRegistrykey1) {
+            registrykey = originRegistrykey;
+         }
+         //LoliServer end
          ServerWorld serverworld = ((ServerWorld)p_196262_2_).getServer().getLevel(registrykey);
          if (serverworld == null) {
-            return;
-         }
+            //return; // CraftBukkit - always fire event in case plugins wish to change it
+         }
+
+         // CraftBukkit start - Entity in portal
+         EntityPortalEnterEvent event = new EntityPortalEnterEvent(p_196262_4_.getBukkitEntity(), new org.bukkit.Location(p_196262_2_.getWorld(), p_196262_3_.getX(), p_196262_3_.getY(), p_196262_3_.getZ()));
+         p_196262_2_.getCBServer().getPluginManager().callEvent(event);
+         // LoliServer start - don't teleport when serverworld == null
+         if (serverworld == null) {
+            return;
+         }
+         // LoliServer end
+         if (p_196262_4_ instanceof ServerPlayerEntity) {
+            ((ServerPlayerEntity) p_196262_4_).changeDimension(serverworld, serverworld.getPortalForcer(), PlayerTeleportEvent.TeleportCause.END_PORTAL);
+            return;
+         }
+         // CraftBukkit end
 
          p_196262_4_.changeDimension(serverworld);
       }
