--- a/net/minecraft/block/FungusBlock.java
+++ b/net/minecraft/block/FungusBlock.java
@@ -40,6 +_,13 @@
    }
 
    public void performBonemeal(ServerWorld p_225535_1_, Random p_225535_2_, BlockPos p_225535_3_, BlockState p_225535_4_) {
+      // CraftBukkit start
+      if (this == Blocks.WARPED_FUNGUS) { // WARPED_FUNGUS
+         SaplingBlock.treeType = org.bukkit.TreeType.WARPED_FUNGUS;
+      } else if (this == Blocks.CRIMSON_FUNGUS) { // CRIMSON_FUNGUS
+         SaplingBlock.treeType = org.bukkit.TreeType.CRIMSON_FUNGUS;
+      }
+      // CraftBukkit end
       this.feature.get().place(p_225535_1_, p_225535_1_.getChunkSource().getGenerator(), p_225535_2_, p_225535_3_);
    }
 }
