--- a/net/minecraft/block/GrassPathBlock.java
+++ b/net/minecraft/block/GrassPathBlock.java
@@ -36,6 +_,11 @@
    }
 
    public void tick(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
+      // CraftBukkit start - do not fade if the block is valid here
+      if (p_225534_1_.canSurvive(p_225534_2_, p_225534_3_)) {
+         return;
+      }
+      // CraftBukkit end
       FarmlandBlock.turnToDirt(p_225534_1_, p_225534_2_, p_225534_3_);
    }
 
