--- a/net/minecraft/block/LeavesBlock.java
+++ b/net/minecraft/block/LeavesBlock.java
@@ -18,8 +_,9 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.block.LeavesDecayEvent;
 
-public class LeavesBlock extends Block {
+public class LeavesBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final IntegerProperty DISTANCE = BlockStateProperties.DISTANCE;
    public static final BooleanProperty PERSISTENT = BlockStateProperties.PERSISTENT;
 
@@ -38,6 +_,13 @@
 
    public void randomTick(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       if (!p_225542_1_.getValue(PERSISTENT) && p_225542_1_.getValue(DISTANCE) == 7) {
+         // CraftBukkit start
+         LeavesDecayEvent event = new LeavesDecayEvent(p_225542_2_.getWorld().getBlockAt(p_225542_3_.getX(), p_225542_3_.getY(), p_225542_3_.getZ()));
+         p_225542_2_.getCBServer().getPluginManager().callEvent(event);
+         if (event.isCancelled() || p_225542_2_.getBlockState(p_225542_3_).getBlock() != this) {
+            return;
+         }
+         // CraftBukkit end
          dropResources(p_225542_1_, p_225542_2_, p_225542_3_);
          p_225542_2_.removeBlock(p_225542_3_, false);
       }
