--- a/net/minecraft/block/LecternBlock.java
+++ b/net/minecraft/block/LecternBlock.java
@@ -178,11 +_,12 @@
    }
 
    private void popBook(BlockState p_220150_1_, World p_220150_2_, BlockPos p_220150_3_) {
-      TileEntity tileentity = p_220150_2_.getBlockEntity(p_220150_3_);
+      TileEntity tileentity = p_220150_2_.getTileEntity(p_220150_3_, false); // CraftBukkit - don't validate, type may be changed already
       if (tileentity instanceof LecternTileEntity) {
          LecternTileEntity lecterntileentity = (LecternTileEntity)tileentity;
          Direction direction = p_220150_1_.getValue(FACING);
          ItemStack itemstack = lecterntileentity.getBook().copy();
+         if(itemstack.isEmpty()) return;  // CraftBukkit - SPIGOT-5500
          float f = 0.25F * (float)direction.getStepX();
          float f1 = 0.25F * (float)direction.getStepZ();
          ItemEntity itementity = new ItemEntity(p_220150_2_, (double)p_220150_3_.getX() + 0.5D + (double)f, (double)(p_220150_3_.getY() + 1), (double)p_220150_3_.getZ() + 0.5D + (double)f1, itemstack);
