--- a/net/minecraft/block/ObserverBlock.java
+++ b/net/minecraft/block/ObserverBlock.java
@@ -13,6 +_,7 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ObserverBlock extends DirectionalBlock {
    public static final BooleanProperty POWERED = BlockStateProperties.POWERED;
@@ -36,8 +_,18 @@
 
    public void tick(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
       if (p_225534_1_.getValue(POWERED)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_225534_2_, p_225534_3_, 15, 0).getNewCurrent() != 0) {
+            return;
+         }
+         // CraftBukkit end
          p_225534_2_.setBlock(p_225534_3_, p_225534_1_.setValue(POWERED, Boolean.valueOf(false)), 2);
       } else {
+         // CraftBukkit start
+         if (CraftEventFactory.callRedstoneChange(p_225534_2_, p_225534_3_, 0, 15).getNewCurrent() != 15) {
+            return;
+         }
+         // CraftBukkit end
          p_225534_2_.setBlock(p_225534_3_, p_225534_1_.setValue(POWERED, Boolean.valueOf(true)), 2);
          p_225534_2_.getBlockTicks().scheduleTick(p_225534_3_, this, 2);
       }
