--- a/net/minecraft/block/PortalSize.java
+++ b/net/minecraft/block/PortalSize.java
@@ -13,10 +_,14 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlockState;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftPortalEvent;
+import org.bukkit.event.world.PortalCreateEvent;
 
 public class PortalSize {
    private static final AbstractBlock.IPositionPredicate FRAME = (p_242966_0_, p_242966_1_, p_242966_2_) -> {
-      return p_242966_0_.is(Blocks.OBSIDIAN);
+      return p_242966_0_.isPortalFrame(p_242966_1_, p_242966_2_);
    };
    private final IWorld level;
    private final Direction.Axis axis;
@@ -26,6 +_,7 @@
    private BlockPos bottomLeft;
    private int height;
    private int width;
+   org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator blocks; // CraftBukkit - add field
 
    public static Optional<PortalSize> findEmptyPortalShape(IWorld p_242964_0_, BlockPos p_242964_1_, Direction.Axis p_242964_2_) {
       return findPortalShape(p_242964_0_, p_242964_1_, (p_242968_0_) -> {
@@ -44,6 +_,7 @@
    }
 
    public PortalSize(IWorld p_i48740_1_, BlockPos p_i48740_2_, Direction.Axis p_i48740_3_) {
+      blocks = new org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator(p_i48740_1_.getMinecraftWorld()); // CraftBukkit
       this.level = p_i48740_1_;
       this.axis = p_i48740_3_;
       this.rightDir = p_i48740_3_ == Direction.Axis.X ? Direction.WEST : Direction.SOUTH;
@@ -84,6 +_,7 @@
          BlockState blockstate = this.level.getBlockState(blockpos$mutable);
          if (!isEmpty(blockstate)) {
             if (FRAME.test(blockstate, this.level, blockpos$mutable)) {
+               blocks.setBlock(blockpos$mutable, blockstate, 18); // CraftBukkit - lower left / right
                return i;
             }
             break;
@@ -93,6 +_,7 @@
          if (!FRAME.test(blockstate1, this.level, blockpos$mutable)) {
             break;
          }
+         blocks.setBlock(blockpos$mutable, blockstate1, 18); // CraftBukkit - bottom row
       }
 
       return 0;
@@ -138,6 +_,10 @@
                ++this.numPortalBlocks;
             }
          }
+         // CraftBukkit start - left and right
+         blocks.setBlock(p_242969_1_.set(this.bottomLeft).move(Direction.UP, i).move(this.rightDir, -1), this.level.getBlockState(p_242969_1_), 18);
+         blocks.setBlock(p_242969_1_.set(this.bottomLeft).move(Direction.UP, i).move(this.rightDir, this.width), this.level.getBlockState(p_242969_1_), 18);
+         // CraftBukkit end
       }
 
       return 21;
@@ -151,11 +_,27 @@
       return this.bottomLeft != null && this.width >= 2 && this.width <= 21 && this.height >= 3 && this.height <= 21;
    }
 
-   public void createPortalBlocks() {
+   // CraftBukkit start - return boolean
+   public boolean placePortalBlocks() {
+      org.bukkit.World bworld = this.level.getMinecraftWorld().getWorld();
+
+      // Copy below for loop
       BlockState blockstate = Blocks.NETHER_PORTAL.defaultBlockState().setValue(NetherPortalBlock.AXIS, this.axis);
       BlockPos.betweenClosed(this.bottomLeft, this.bottomLeft.relative(Direction.UP, this.height - 1).relative(this.rightDir, this.width - 1)).forEach((p_242967_2_) -> {
+         blocks.setBlock(p_242967_2_, blockstate, 18);
+      });
+
+      PortalCreateEvent event = new PortalCreateEvent((java.util.List<org.bukkit.block.BlockState>) (java.util.List) blocks.getList(), bworld, null, PortalCreateEvent.CreateReason.FIRE);
+      this.level.getMinecraftWorld().getServer().server.getPluginManager().callEvent(event);
+
+      if (event.isCancelled()) {
+         return false;
+      }
+      // CraftBukkit end
+      BlockPos.betweenClosed(this.bottomLeft, this.bottomLeft.relative(Direction.UP, this.height - 1).relative(this.rightDir, this.width - 1)).forEach((p_242967_2_) -> {
          this.level.setBlock(p_242967_2_, blockstate, 18);
       });
+      return true; // CraftBukkit
    }
 
    public boolean isComplete() {
@@ -188,18 +_,22 @@
    }
 
    public static PortalInfo createPortalInfo(ServerWorld p_242963_0_, TeleportationRepositioner.Result p_242963_1_, Direction.Axis p_242963_2_, Vector3d p_242963_3_, EntitySize p_242963_4_, Vector3d p_242963_5_, float p_242963_6_, float p_242963_7_) {
-      BlockPos blockpos = p_242963_1_.minCorner;
-      BlockState blockstate = p_242963_0_.getBlockState(blockpos);
+       return createPortalInfo(p_242963_0_, p_242963_1_, p_242963_2_, p_242963_3_, p_242963_4_, p_242963_5_, p_242963_6_, p_242963_7_, null);
+   }
+
+   public static PortalInfo createPortalInfo(ServerWorld world, TeleportationRepositioner.Result result, Direction.Axis axis, Vector3d offsetVector, EntitySize size, Vector3d motion, float rotationYaw, float rotationPitch, CraftPortalEvent portalEventInfo) {
+      BlockPos blockpos = result.minCorner;
+      BlockState blockstate = world.getBlockState(blockpos);
       Direction.Axis direction$axis = blockstate.getValue(BlockStateProperties.HORIZONTAL_AXIS);
-      double d0 = (double)p_242963_1_.axis1Size;
-      double d1 = (double)p_242963_1_.axis2Size;
-      int i = p_242963_2_ == direction$axis ? 0 : 90;
-      Vector3d vector3d = p_242963_2_ == direction$axis ? p_242963_5_ : new Vector3d(p_242963_5_.z, p_242963_5_.y, -p_242963_5_.x);
-      double d2 = (double)p_242963_4_.width / 2.0D + (d0 - (double)p_242963_4_.width) * p_242963_3_.x();
-      double d3 = (d1 - (double)p_242963_4_.height) * p_242963_3_.y();
-      double d4 = 0.5D + p_242963_3_.z();
+      double d0 = (double)result.axis1Size;
+      double d1 = (double)result.axis2Size;
+      int i = axis == direction$axis ? 0 : 90;
+      Vector3d vector3d = axis == direction$axis ? motion : new Vector3d(motion.z, motion.y, -motion.x);
+      double d2 = (double)size.width / 2.0D + (d0 - (double)size.width) * offsetVector.x();
+      double d3 = (d1 - (double)size.height) * offsetVector.y();
+      double d4 = 0.5D + offsetVector.z();
       boolean flag = direction$axis == Direction.Axis.X;
       Vector3d vector3d1 = new Vector3d((double)blockpos.getX() + (flag ? d2 : d4), (double)blockpos.getY() + d3, (double)blockpos.getZ() + (flag ? d4 : d2));
-      return new PortalInfo(vector3d1, vector3d, p_242963_6_ + (float)i, p_242963_7_);
+      return new PortalInfo(vector3d1, vector3d, rotationYaw + (float)i, rotationPitch).putCB(world, portalEventInfo); // CatServer
    }
 }
