--- a/net/minecraft/block/RedstoneOreBlock.java
+++ b/net/minecraft/block/RedstoneOreBlock.java
@@ -1,8 +_,7 @@
 package net.minecraft.block;
 
 import java.util.Random;
-import net.minecraft.enchantment.EnchantmentHelper;
-import net.minecraft.enchantment.Enchantments;
+
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.BlockItem;
@@ -20,6 +_,9 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityInteractEvent;
+import org.bukkit.event.player.PlayerInteractEvent;
 
 public class RedstoneOreBlock extends Block {
    public static final BooleanProperty LIT = RedstoneTorchBlock.LIT;
@@ -30,30 +_,50 @@
    }
 
    public void attack(BlockState p_196270_1_, World p_196270_2_, BlockPos p_196270_3_, PlayerEntity p_196270_4_) {
-      interact(p_196270_1_, p_196270_2_, p_196270_3_);
+      interact(p_196270_1_, p_196270_2_, p_196270_3_, p_196270_4_);  // CraftBukkit - add entityhuman
       super.attack(p_196270_1_, p_196270_2_, p_196270_3_, p_196270_4_);
    }
 
    public void stepOn(World p_176199_1_, BlockPos p_176199_2_, Entity p_176199_3_) {
-      interact(p_176199_1_.getBlockState(p_176199_2_), p_176199_1_, p_176199_2_);
-      super.stepOn(p_176199_1_, p_176199_2_, p_176199_3_);
+
+      // CraftBukkit start
+      if (p_176199_3_ instanceof PlayerEntity) {
+         PlayerInteractEvent event = CraftEventFactory.callPlayerInteractEvent((PlayerEntity) p_176199_3_, org.bukkit.event.block.Action.PHYSICAL, p_176199_2_, null, null, null);
+         if (!event.isCancelled()) {
+            interact(p_176199_1_.getBlockState(p_176199_2_), p_176199_1_, p_176199_2_, p_176199_3_); // add entity
+            super.stepOn(p_176199_1_, p_176199_2_, p_176199_3_);
+         }
+      } else {
+         EntityInteractEvent event = new EntityInteractEvent(p_176199_3_.getBukkitEntity(), p_176199_1_.getWorld().getBlockAt(p_176199_2_.getX(), p_176199_2_.getY(), p_176199_2_.getZ()));
+         p_176199_1_.getCBServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
+            interact(p_176199_1_.getBlockState(p_176199_2_), p_176199_1_, p_176199_2_, p_176199_3_); // add entity
+            super.stepOn(p_176199_1_, p_176199_2_, p_176199_3_);
+         }
+      }
+      // CraftBukkit end
    }
 
    public ActionResultType use(BlockState p_225533_1_, World p_225533_2_, BlockPos p_225533_3_, PlayerEntity p_225533_4_, Hand p_225533_5_, BlockRayTraceResult p_225533_6_) {
       if (p_225533_2_.isClientSide) {
          spawnParticles(p_225533_2_, p_225533_3_);
       } else {
-         interact(p_225533_1_, p_225533_2_, p_225533_3_);
+         interact(p_225533_1_, p_225533_2_, p_225533_3_, p_225533_4_); // CraftBukkit - add entityhuman
       }
 
       ItemStack itemstack = p_225533_4_.getItemInHand(p_225533_5_);
       return itemstack.getItem() instanceof BlockItem && (new BlockItemUseContext(p_225533_4_, p_225533_5_, itemstack, p_225533_6_)).canPlace() ? ActionResultType.PASS : ActionResultType.SUCCESS;
    }
 
-   private static void interact(BlockState p_196500_0_, World p_196500_1_, BlockPos p_196500_2_) {
-      spawnParticles(p_196500_1_, p_196500_2_);
-      if (!p_196500_0_.getValue(LIT)) {
-         p_196500_1_.setBlock(p_196500_2_, p_196500_0_.setValue(LIT, Boolean.valueOf(true)), 3);
+   private static void interact(BlockState state, World world, BlockPos pos, Entity entity) { // CraftBukkit - add Entity
+      spawnParticles(world, pos);
+      if (!state.getValue(LIT)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callEntityChangeBlockEvent(entity, pos, state.setValue(LIT, true)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
+         world.setBlock(pos, state.setValue(LIT, Boolean.valueOf(true)), 3);
       }
 
    }
@@ -64,6 +_,11 @@
 
    public void randomTick(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       if (p_225542_1_.getValue(LIT)) {
+         // CraftBukkit start
+         if (CraftEventFactory.callBlockFadeEvent(p_225542_2_, p_225542_3_, p_225542_1_.setValue(LIT, true)).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          p_225542_2_.setBlock(p_225542_3_, p_225542_1_.setValue(LIT, Boolean.valueOf(false)), 3);
       }
 
@@ -71,11 +_,11 @@
 
    public void spawnAfterBreak(BlockState p_220062_1_, ServerWorld p_220062_2_, BlockPos p_220062_3_, ItemStack p_220062_4_) {
       super.spawnAfterBreak(p_220062_1_, p_220062_2_, p_220062_3_, p_220062_4_);
-      if (EnchantmentHelper.getItemEnchantmentLevel(Enchantments.SILK_TOUCH, p_220062_4_) == 0) {
-         int i = 1 + p_220062_2_.random.nextInt(5);
-         this.popExperience(p_220062_2_, p_220062_3_, i);
-      }
+   }
 
+   @Override
+   public int getExpDrop(BlockState state, net.minecraft.world.IWorldReader world, BlockPos pos, int fortune, int silktouch) {
+      return silktouch == 0 ? 1 + RANDOM.nextInt(5) : 0;
    }
 
    @OnlyIn(Dist.CLIENT)
