--- a/net/minecraft/block/RedstoneTorchBlock.java
+++ b/net/minecraft/block/RedstoneTorchBlock.java
@@ -16,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.block.BlockRedstoneEvent;
 
 public class RedstoneTorchBlock extends TorchBlock {
    public static final BooleanProperty LIT = BlockStateProperties.LIT;
@@ -58,8 +_,24 @@
          list.remove(0);
       }
 
+      // CraftBukkit start
+      org.bukkit.plugin.PluginManager manager = p_225534_2_.getCBServer().getPluginManager();
+      org.bukkit.block.Block block = p_225534_2_.getWorld().getBlockAt(p_225534_3_.getX(), p_225534_3_.getY(), p_225534_3_.getZ());
+      int oldCurrent = ((Boolean) p_225534_1_.getValue(LIT)).booleanValue() ? 15 : 0;
+      BlockRedstoneEvent event = new BlockRedstoneEvent(block, oldCurrent, oldCurrent);
+      // CraftBukkit end
+
       if (p_225534_1_.getValue(LIT)) {
          if (flag) {
+            // CraftBukkit start
+            if (oldCurrent != 0) {
+               event.setNewCurrent(0);
+               manager.callEvent(event);
+               if (event.getNewCurrent() != 0) {
+                  return;
+               }
+            }
+            // CraftBukkit end
             p_225534_2_.setBlock(p_225534_3_, p_225534_1_.setValue(LIT, Boolean.valueOf(false)), 3);
             if (isToggledTooFrequently(p_225534_2_, p_225534_3_, true)) {
                p_225534_2_.levelEvent(1502, p_225534_3_, 0);
@@ -67,6 +_,15 @@
             }
          }
       } else if (!flag && !isToggledTooFrequently(p_225534_2_, p_225534_3_, false)) {
+         // CraftBukkit start
+         if (oldCurrent != 15) {
+            event.setNewCurrent(15);
+            manager.callEvent(event);
+            if (event.getNewCurrent() != 15) {
+               return;
+            }
+         }
+         // CraftBukkit end
          p_225534_2_.setBlock(p_225534_3_, p_225534_1_.setValue(LIT, Boolean.valueOf(true)), 3);
       }
 
