--- a/net/minecraft/block/ScaffoldingBlock.java
+++ b/net/minecraft/block/ScaffoldingBlock.java
@@ -19,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class ScaffoldingBlock extends Block implements IWaterLoggable {
    private static final VoxelShape STABLE_SHAPE;
@@ -83,7 +_,7 @@
    public void tick(BlockState p_225534_1_, ServerWorld p_225534_2_, BlockPos p_225534_3_, Random p_225534_4_) {
       int i = getDistance(p_225534_2_, p_225534_3_);
       BlockState blockstate = p_225534_1_.setValue(DISTANCE, Integer.valueOf(i)).setValue(BOTTOM, Boolean.valueOf(this.isBottom(p_225534_2_, p_225534_3_, i)));
-      if (blockstate.getValue(DISTANCE) == 7) {
+      if (blockstate.getValue(DISTANCE) == 7 && !CraftEventFactory.callBlockFadeEvent(p_225534_2_, p_225534_3_, Blocks.AIR.defaultBlockState()).isCancelled()) { // CraftBukkit - BlockFadeEvent
          if (p_225534_1_.getValue(DISTANCE) == 7) {
             p_225534_2_.addFreshEntity(new FallingBlockEntity(p_225534_2_, (double)p_225534_3_.getX() + 0.5D, (double)p_225534_3_.getY(), (double)p_225534_3_.getZ() + 0.5D, blockstate.setValue(WATERLOGGED, Boolean.valueOf(false))));
          } else {
