--- a/net/minecraft/block/ShulkerBoxBlock.java
+++ b/net/minecraft/block/ShulkerBoxBlock.java
@@ -50,7 +_,7 @@
    public static final EnumProperty<Direction> FACING = DirectionalBlock.FACING;
    public static final ResourceLocation CONTENTS = new ResourceLocation("contents");
    @Nullable
-   private final DyeColor color;
+   public final DyeColor color;
 
    public ShulkerBoxBlock(@Nullable DyeColor p_i48334_1_, AbstractBlock.Properties p_i48334_2_) {
       super(p_i48334_2_);
