--- a/net/minecraft/block/SilverfishBlock.java
+++ b/net/minecraft/block/SilverfishBlock.java
@@ -12,6 +_,7 @@
 import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 
 public class SilverfishBlock extends Block {
    private final Block hostBlock;
@@ -34,7 +_,7 @@
    private void spawnInfestation(ServerWorld p_235505_1_, BlockPos p_235505_2_) {
       SilverfishEntity silverfishentity = EntityType.SILVERFISH.create(p_235505_1_);
       silverfishentity.moveTo((double)p_235505_2_.getX() + 0.5D, (double)p_235505_2_.getY(), (double)p_235505_2_.getZ() + 0.5D, 0.0F, 0.0F);
-      p_235505_1_.addFreshEntity(silverfishentity);
+      p_235505_1_.addEntity(silverfishentity, SpawnReason.SILVERFISH_BLOCK); // CraftBukkit - add SpawnReason
       silverfishentity.spawnAnim();
    }
 
