--- a/net/minecraft/block/SnowBlock.java
+++ b/net/minecraft/block/SnowBlock.java
@@ -17,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.LightType;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class SnowBlock extends Block {
    public static final IntegerProperty LAYERS = BlockStateProperties.LAYERS;
@@ -79,6 +_,11 @@
 
    public void randomTick(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
       if (p_225542_2_.getBrightness(LightType.BLOCK, p_225542_3_) > 11) {
+         // CraftBukkit start
+         if (CraftEventFactory.callBlockFadeEvent(p_225542_2_, p_225542_3_, Blocks.AIR.defaultBlockState()).isCancelled()) {
+            return;
+         }
+         // CraftBukkit end
          dropResources(p_225542_1_, p_225542_2_, p_225542_3_);
          p_225542_2_.removeBlock(p_225542_3_, false);
       }
