--- a/net/minecraft/block/VineBlock.java
+++ b/net/minecraft/block/VineBlock.java
@@ -21,8 +_,9 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
-public class VineBlock extends Block {
+public class VineBlock extends Block implements net.minecraftforge.common.IForgeShearable {
    public static final BooleanProperty UP = SixWayBlock.UP;
    public static final BooleanProperty NORTH = SixWayBlock.NORTH;
    public static final BooleanProperty EAST = SixWayBlock.EAST;
@@ -152,32 +_,36 @@
    }
 
    public void randomTick(BlockState p_225542_1_, ServerWorld p_225542_2_, BlockPos p_225542_3_, Random p_225542_4_) {
-      if (p_225542_2_.random.nextInt(4) == 0) {
+      if (p_225542_2_.random.nextInt(Math.max(1, (int) (100.0F / p_225542_2_.spigotConfig.vineModifier) * 4)) == 0 && p_225542_2_.isAreaLoaded(p_225542_3_, 4)) { // Forge: check area to prevent loading unloaded chunks
          Direction direction = Direction.getRandom(p_225542_4_);
          BlockPos blockpos = p_225542_3_.above();
          if (direction.getAxis().isHorizontal() && !p_225542_1_.getValue(getPropertyForFace(direction))) {
             if (this.canSpread(p_225542_2_, p_225542_3_)) {
                BlockPos blockpos4 = p_225542_3_.relative(direction);
                BlockState blockstate4 = p_225542_2_.getBlockState(blockpos4);
-               if (blockstate4.isAir()) {
+               if (blockstate4.isAir(p_225542_2_, blockpos4)) {
                   Direction direction3 = direction.getClockWise();
                   Direction direction4 = direction.getCounterClockWise();
                   boolean flag = p_225542_1_.getValue(getPropertyForFace(direction3));
                   boolean flag1 = p_225542_1_.getValue(getPropertyForFace(direction4));
                   BlockPos blockpos2 = blockpos4.relative(direction3);
                   BlockPos blockpos3 = blockpos4.relative(direction4);
+
+                  // CraftBukkit start - Call BlockSpreadEvent
+                  BlockPos source = p_225542_3_;
+
                   if (flag && isAcceptableNeighbour(p_225542_2_, blockpos2, direction3)) {
-                     p_225542_2_.setBlock(blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction3), Boolean.valueOf(true)), 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, source, blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction3), Boolean.valueOf(true)), 2);
                   } else if (flag1 && isAcceptableNeighbour(p_225542_2_, blockpos3, direction4)) {
-                     p_225542_2_.setBlock(blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction4), Boolean.valueOf(true)), 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, source, blockpos4, this.defaultBlockState().setValue(getPropertyForFace(direction4), Boolean.valueOf(true)), 2);
                   } else {
                      Direction direction1 = direction.getOpposite();
                      if (flag && p_225542_2_.isEmptyBlock(blockpos2) && isAcceptableNeighbour(p_225542_2_, p_225542_3_.relative(direction3), direction1)) {
-                        p_225542_2_.setBlock(blockpos2, this.defaultBlockState().setValue(getPropertyForFace(direction1), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, source, blockpos2, this.defaultBlockState().setValue(getPropertyForFace(direction1), Boolean.valueOf(true)), 2);
                      } else if (flag1 && p_225542_2_.isEmptyBlock(blockpos3) && isAcceptableNeighbour(p_225542_2_, p_225542_3_.relative(direction4), direction1)) {
-                        p_225542_2_.setBlock(blockpos3, this.defaultBlockState().setValue(getPropertyForFace(direction1), Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, source, blockpos3, this.defaultBlockState().setValue(getPropertyForFace(direction1), Boolean.valueOf(true)), 2);
                      } else if ((double)p_225542_2_.random.nextFloat() < 0.05D && isAcceptableNeighbour(p_225542_2_, blockpos4.above(), Direction.UP)) {
-                        p_225542_2_.setBlock(blockpos4, this.defaultBlockState().setValue(UP, Boolean.valueOf(true)), 2);
+                        CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, source, blockpos4, this.defaultBlockState().setValue(UP, Boolean.valueOf(true)), 2);
                      }
                   }
                } else if (isAcceptableNeighbour(p_225542_2_, blockpos4, direction)) {
@@ -206,7 +_,7 @@
                   }
 
                   if (this.hasHorizontalConnection(blockstate3)) {
-                     p_225542_2_.setBlock(blockpos, blockstate3, 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, p_225542_3_, blockpos, blockstate3, 2); // CraftBukkit
                   }
 
                   return;
@@ -216,11 +_,12 @@
             if (p_225542_3_.getY() > 0) {
                BlockPos blockpos1 = p_225542_3_.below();
                BlockState blockstate = p_225542_2_.getBlockState(blockpos1);
-               if (blockstate.isAir() || blockstate.is(this)) {
-                  BlockState blockstate1 = blockstate.isAir() ? this.defaultBlockState() : blockstate;
+               boolean isAir = blockstate.isAir(p_225542_2_, blockpos1);
+               if (isAir || blockstate.is(this)) {
+                  BlockState blockstate1 = isAir ? this.defaultBlockState() : blockstate;
                   BlockState blockstate2 = this.copyRandomFaces(p_225542_1_, blockstate1, p_225542_4_);
                   if (blockstate1 != blockstate2 && this.hasHorizontalConnection(blockstate2)) {
-                     p_225542_2_.setBlock(blockpos1, blockstate2, 2);
+                     CraftEventFactory.handleBlockSpreadEvent(p_225542_2_, p_225542_3_, blockpos1, blockstate2, 2); // CraftBukkit
                   }
                }
             }
