--- a/net/minecraft/block/WitherSkeletonSkullBlock.java
+++ b/net/minecraft/block/WitherSkeletonSkullBlock.java
@@ -21,6 +_,8 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.Difficulty;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.util.BlockStateListPopulator;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
 
 public class WitherSkeletonSkullBlock extends SkullBlock {
    @Nullable
@@ -42,6 +_,7 @@
    }
 
    public static void checkSpawn(World p_196298_0_, BlockPos p_196298_1_, SkullTileEntity p_196298_2_) {
+      if (p_196298_0_.captureBlockStates) return; // CraftBukkit
       if (!p_196298_0_.isClientSide) {
          BlockState blockstate = p_196298_2_.getBlockState();
          boolean flag = blockstate.is(Blocks.WITHER_SKELETON_SKULL) || blockstate.is(Blocks.WITHER_SKELETON_WALL_SKULL);
@@ -67,7 +_,7 @@
                   CriteriaTriggers.SUMMONED_ENTITY.trigger(serverplayerentity, witherentity);
                }
 
-               p_196298_0_.addFreshEntity(witherentity);
+               p_196298_0_.addEntity(witherentity, SpawnReason.BUILD_WITHER);
 
                for(int k = 0; k < blockpattern.getWidth(); ++k) {
                   for(int l = 0; l < blockpattern.getHeight(); ++l) {
