--- a/net/minecraft/nbt/JsonToNBT.java
+++ b/net/minecraft/nbt/JsonToNBT.java
@@ -77,7 +_,7 @@
       }
    }
 
-   private INBT type(String p_193596_1_) {
+   public INBT type(String p_193596_1_) {
       try {
          if (FLOAT_PATTERN.matcher(p_193596_1_).matches()) {
             return FloatNBT.valueOf(Float.parseFloat(p_193596_1_.substring(0, p_193596_1_.length() - 1)));
@@ -201,7 +_,7 @@
       }
    }
 
-   private INBT readArrayTag() throws CommandSyntaxException {
+   public INBT readArrayTag() throws CommandSyntaxException {
       this.expect('[');
       int i = this.reader.getCursor();
       char c0 = this.reader.read();
