--- a/net/minecraft/world/chunk/ChunkSection.java
+++ b/net/minecraft/world/chunk/ChunkSection.java
@@ -15,7 +_,7 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class ChunkSection {
-   private static final IPalette<BlockState> GLOBAL_BLOCKSTATE_PALETTE = new IdentityPalette<>(Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState());
+   public static final IPalette<BlockState> GLOBAL_BLOCKSTATE_PALETTE = new IdentityPalette<>(Block.BLOCK_STATE_REGISTRY, Blocks.AIR.defaultBlockState());
    private final int bottomBlockY;
    private short nonEmptyBlockCount;
    private short tickingBlockCount;
