--- a/net/minecraft/world/chunk/NibbleArray.java
+++ b/net/minecraft/world/chunk/NibbleArray.java
@@ -33,7 +_,7 @@
       return p_177483_2_ << 8 | p_177483_3_ << 4 | p_177483_1_;
    }
 
-   private int get(int p_177480_1_) {
+   public int get(int p_177480_1_) {
       if (this.data == null) {
          return 0;
       } else {
@@ -42,7 +_,7 @@
       }
    }
 
-   private void set(int p_177482_1_, int p_177482_2_) {
+   public void set(int p_177482_1_, int p_177482_2_) {
       if (this.data == null) {
          this.data = new byte[2048];
       }
