--- a/net/minecraft/world/storage/DerivedWorldInfo.java
+++ b/net/minecraft/world/storage/DerivedWorldInfo.java
@@ -13,7 +_,15 @@
 public class DerivedWorldInfo implements IServerWorldInfo {
    private final IServerConfiguration worldData;
    private final IServerWorldInfo wrapped;
-
+   private String properName;
+
+   public DerivedWorldInfo(IServerConfiguration configuration, IServerWorldInfo delegate, String properName) {
+       this.worldData = configuration;
+       this.wrapped = delegate;
+       this.properName = properName;
+   }
+
+   @Deprecated
    public DerivedWorldInfo(IServerConfiguration p_i232150_1_, IServerWorldInfo p_i232150_2_) {
       this.worldData = p_i232150_1_;
       this.wrapped = p_i232150_2_;
@@ -44,7 +_,7 @@
    }
 
    public String getLevelName() {
-      return this.worldData.getLevelName();
+      return properName != null ? properName : this.worldData.getLevelName();
    }
 
    public int getClearWeatherTime() {
