--- a/net/minecraft/world/storage/ServerWorldInfo.java
+++ b/net/minecraft/world/storage/ServerWorldInfo.java
@@ -13,12 +_,14 @@
 import net.minecraft.command.TimerCallbackManager;
 import net.minecraft.command.TimerCallbackSerializers;
 import net.minecraft.crash.CrashReportCategory;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.INBT;
 import net.minecraft.nbt.ListNBT;
 import net.minecraft.nbt.NBTDynamicOps;
 import net.minecraft.nbt.NBTUtil;
 import net.minecraft.nbt.StringNBT;
+import net.minecraft.network.play.server.SServerDifficultyPacket;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.SharedConstants;
 import net.minecraft.util.UUIDCodec;
@@ -34,14 +_,18 @@
 import net.minecraft.world.WorldSettings;
 import net.minecraft.world.border.WorldBorder;
 import net.minecraft.world.gen.settings.DimensionGeneratorSettings;
+import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.event.weather.ThunderChangeEvent;
+import org.bukkit.event.weather.WeatherChangeEvent;
 
 public class ServerWorldInfo implements IServerWorldInfo, IServerConfiguration {
    private static final Logger LOGGER = LogManager.getLogger();
-   private WorldSettings settings;
+   public WorldSettings settings;
    private final DimensionGeneratorSettings worldGenSettings;
    private final Lifecycle worldGenSettingsLifecycle;
    private int xSpawn;
@@ -75,6 +_,7 @@
    private final Set<String> knownServerBrands;
    private boolean wasModded;
    private final TimerCallbackManager<MinecraftServer> scheduledEvents;
+   public ServerWorld world; // CraftBukkit
 
    private ServerWorldInfo(@Nullable DataFixer p_i242043_1_, int p_i242043_2_, @Nullable CompoundNBT p_i242043_3_, boolean p_i242043_4_, int p_i242043_5_, int p_i242043_6_, int p_i242043_7_, float p_i242043_8_, long p_i242043_9_, long p_i242043_11_, int p_i242043_13_, int p_i242043_14_, int p_i242043_15_, boolean p_i242043_16_, int p_i242043_17_, boolean p_i242043_18_, boolean p_i242043_19_, boolean p_i242043_20_, WorldBorder.Serializer p_i242043_21_, int p_i242043_22_, int p_i242043_23_, @Nullable UUID p_i242043_24_, LinkedHashSet<String> p_i242043_25_, TimerCallbackManager<MinecraftServer> p_i242043_26_, @Nullable CompoundNBT p_i242043_27_, CompoundNBT p_i242043_28_, WorldSettings p_i242043_29_, DimensionGeneratorSettings p_i242043_30_, Lifecycle p_i242043_31_) {
       this.fixerUpper = p_i242043_1_;
@@ -189,6 +_,7 @@
          p_237370_2_.putUUID("WanderingTraderId", this.wanderingTraderId);
       }
 
+      p_237370_2_.putString("Bukkit.Version", "CatServer/1.16.5/" + org.bukkit.craftbukkit.v1_16_R3.util.Versioning.getBukkitVersion()); // CraftBukkit
    }
 
    public int getXSpawn() {
@@ -286,6 +_,19 @@
    }
 
    public void setThundering(boolean p_76069_1_) {
+      // CraftBukkit start
+      if (this.thundering == p_76069_1_) {
+         return;
+      }
+      org.bukkit.World world = Bukkit.getWorld(getLevelName());
+      if (world != null) {
+         ThunderChangeEvent thunder = new ThunderChangeEvent(world, p_76069_1_);
+         Bukkit.getServer().getPluginManager().callEvent(thunder);
+         if (thunder.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.thundering = p_76069_1_;
    }
 
@@ -302,6 +_,19 @@
    }
 
    public void setRaining(boolean p_76084_1_) {
+      // CraftBukkit start
+      if (this.raining == p_76084_1_) {
+         return;
+      }
+      org.bukkit.World world = Bukkit.getWorld(getLevelName());
+      if (world != null) {
+         WeatherChangeEvent weather = new WeatherChangeEvent(world, p_76084_1_);
+         Bukkit.getServer().getPluginManager().callEvent(weather);
+         if (weather.isCancelled()) {
+            return;
+         }
+      }
+      // CraftBukkit end
       this.raining = p_76084_1_;
    }
 
@@ -355,6 +_,12 @@
 
    public void setDifficulty(Difficulty p_230409_1_) {
       this.settings = this.settings.withDifficulty(p_230409_1_);
+      /*/ CraftBukkit start
+      SServerDifficultyPacket packet = new SServerDifficultyPacket(this.getDifficulty(), this.isDifficultyLocked());
+      for (ServerPlayerEntity player : world.getPlayers()) {
+         player.connection.sendPacket(packet);
+      }
+      // CraftBukkit end*/
    }
 
    public boolean isDifficultyLocked() {
@@ -449,4 +_,12 @@
    public WorldSettings getLevelSettings() {
       return this.settings.copy();
    }
+
+   // CraftBukkit start - Check if the name stored in NBT is the correct one
+   public void checkName(String name) {
+      if (!this.settings.levelName.equals(name)) {
+         this.settings.levelName = name;
+      }
+   }
+   // CraftBukkit end
 }
