--- a/net/minecraft/block/trees/BigTree.java
+++ b/net/minecraft/block/trees/BigTree.java
@@ -34,6 +_,7 @@
          return false;
       } else {
          configuredfeature.config.setFromSapling();
+         setTreeType(configuredfeature); // CraftBukkit
          BlockState blockstate = Blocks.AIR.defaultBlockState();
          p_235678_1_.setBlock(p_235678_3_.offset(p_235678_6_, 0, p_235678_7_), blockstate, 4);
          p_235678_1_.setBlock(p_235678_3_.offset(p_235678_6_ + 1, 0, p_235678_7_), blockstate, 4);
