--- a/net/minecraft/inventory/container/CartographyContainer.java
+++ b/net/minecraft/inventory/container/CartographyContainer.java
@@ -14,8 +_,28 @@
 import net.minecraft.util.SoundCategory;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.storage.MapData;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryCartography;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
+import org.bukkit.entity.Player;
 
 public class CartographyContainer extends Container {
+
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private Player player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryCartography inventory = new CraftInventoryCartography(this.container, this.resultContainer);
+      bukkitEntity = new CraftInventoryView(this.player, inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    private final IWorldPosCallable access;
    private long lastSoundTime;
    public final IInventory container = new Inventory(2) {
@@ -23,12 +_,26 @@
          CartographyContainer.this.slotsChanged(this);
          super.setChanged();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return access.getLocation();
+      }
+      // CraftBukkit end
    };
    private final CraftResultInventory resultContainer = new CraftResultInventory() {
       public void setChanged() {
          CartographyContainer.this.slotsChanged(this);
          super.setChanged();
       }
+
+      // CraftBukkit start
+      @Override
+      public Location getLocation() {
+         return access.getLocation();
+      }
+      // CraftBukkit end
    };
 
    public CartographyContainer(int p_i50093_1_, PlayerInventory p_i50093_2_) {
@@ -80,9 +_,12 @@
          this.addSlot(new Slot(p_i50094_2_, k, 8 + k * 18, 142));
       }
 
+      player = (Player) p_i50094_2_.player.getBukkitEntity(); // CraftBukkit
+
    }
 
    public boolean stillValid(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return stillValid(this.access, p_75145_1_, Blocks.CARTOGRAPHY_TABLE);
    }
 
