--- a/net/minecraft/inventory/container/ChestContainer.java
+++ b/net/minecraft/inventory/container/ChestContainer.java
@@ -1,5 +_,6 @@
 package net.minecraft.inventory.container;
 
+import net.minecraft.block.ChestBlock.DoubleInventory;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.entity.player.PlayerInventory;
 import net.minecraft.inventory.IInventory;
@@ -7,11 +_,35 @@
 import net.minecraft.item.ItemStack;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class ChestContainer extends Container {
    private final IInventory container;
    private final int containerRows;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventory inventory;
+      if (this.container instanceof PlayerInventory) {
+         inventory = new org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryPlayer((PlayerInventory) this.container);
+      } else if (this.container instanceof DoubleInventory) {
+         inventory = new org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryDoubleChest((DoubleInventory) this.container);
+      } else {
+         inventory = new CraftInventory(this.container);
+      }
+      bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
+
    private ChestContainer(ContainerType<?> p_i50091_1_, int p_i50091_2_, PlayerInventory p_i50091_3_, int p_i50091_4_) {
       this(p_i50091_1_, p_i50091_2_, p_i50091_3_, new Inventory(9 * p_i50091_4_), p_i50091_4_);
    }
@@ -53,6 +_,9 @@
       checkContainerSize(p_i50092_4_, p_i50092_5_ * 9);
       this.container = p_i50092_4_;
       this.containerRows = p_i50092_5_;
+      // CraftBukkit start - Save player
+      this.player = p_i50092_3_;
+      // CraftBukkit end
       p_i50092_4_.startOpen(p_i50092_3_.player);
       int i = (this.containerRows - 4) * 18;
 
@@ -75,6 +_,7 @@
    }
 
    public boolean stillValid(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.container.stillValid(p_75145_1_);
    }
 
