--- a/net/minecraft/inventory/container/DispenserContainer.java
+++ b/net/minecraft/inventory/container/DispenserContainer.java
@@ -5,16 +_,26 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class DispenserContainer extends Container {
    private final IInventory dispenser;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+   // CraftBukkit end
+
    public DispenserContainer(int p_i50087_1_, PlayerInventory p_i50087_2_) {
       this(p_i50087_1_, p_i50087_2_, new Inventory(9));
    }
 
    public DispenserContainer(int p_i50088_1_, PlayerInventory p_i50088_2_, IInventory p_i50088_3_) {
       super(ContainerType.GENERIC_3x3, p_i50088_1_);
+      // CraftBukkit start - Save player
+      this.player = p_i50088_2_;
+      // CraftBukkit end
       checkContainerSize(p_i50088_3_, 9);
       this.dispenser = p_i50088_3_;
       p_i50088_3_.startOpen(p_i50088_2_.player);
@@ -38,6 +_,7 @@
    }
 
    public boolean stillValid(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.dispenser.stillValid(p_75145_1_);
    }
 
@@ -75,4 +_,16 @@
       super.removed(p_75134_1_);
       this.dispenser.stopOpen(p_75134_1_);
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventory inventory = new CraftInventory(this.dispenser);
+      bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
