--- a/net/minecraft/inventory/container/FurnaceResultSlot.java
+++ b/net/minecraft/inventory/container/FurnaceResultSlot.java
@@ -40,9 +_,10 @@
    protected void checkTakeAchievements(ItemStack p_75208_1_) {
       p_75208_1_.onCraftedBy(this.player.level, this.player, this.removeCount);
       if (!this.player.level.isClientSide && this.container instanceof AbstractFurnaceTileEntity) {
-         ((AbstractFurnaceTileEntity)this.container).awardUsedRecipesAndPopExperience(this.player);
+         ((AbstractFurnaceTileEntity)this.container).unlockRecipes(this.player, p_75208_1_, this.removeCount); // CraftBukkit
       }
 
       this.removeCount = 0;
+      net.minecraftforge.fml.hooks.BasicEventHooks.firePlayerSmeltedEvent(this.player, p_75208_1_);
    }
 }
