--- a/net/minecraft/inventory/container/HopperContainer.java
+++ b/net/minecraft/inventory/container/HopperContainer.java
@@ -5,10 +_,28 @@
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
 import net.minecraft.item.ItemStack;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class HopperContainer extends Container {
    private final IInventory hopper;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventory inventory = new CraftInventory(this.hopper);
+      bukkitEntity = new CraftInventoryView(this.player.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+
+   // CraftBukkit stop
+
    public HopperContainer(int p_i50078_1_, PlayerInventory p_i50078_2_) {
       this(p_i50078_1_, p_i50078_2_, new Inventory(5));
    }
@@ -16,6 +_,7 @@
    public HopperContainer(int p_i50079_1_, PlayerInventory p_i50079_2_, IInventory p_i50079_3_) {
       super(ContainerType.HOPPER, p_i50079_1_);
       this.hopper = p_i50079_3_;
+      this.player = p_i50079_2_; // CraftBukkit - save player
       checkContainerSize(p_i50079_3_, 5);
       p_i50079_3_.startOpen(p_i50079_2_.player);
       int i = 51;
@@ -37,6 +_,7 @@
    }
 
    public boolean stillValid(PlayerEntity p_75145_1_) {
+      if (!this.checkReachable) return true; // CraftBukkit
       return this.hopper.stillValid(p_75145_1_);
    }
 
