--- a/net/minecraft/inventory/container/HorseInventoryContainer.java
+++ b/net/minecraft/inventory/container/HorseInventoryContainer.java
@@ -9,15 +_,31 @@
 import net.minecraft.item.Items;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class HorseInventoryContainer extends Container {
    private final IInventory horseContainer;
    private final AbstractHorseEntity horse;
 
+   // CraftBukkit start
+   CraftInventoryView bukkitEntity;
+   PlayerInventory player;
+
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      return bukkitEntity = new CraftInventoryView(player.player.getBukkitEntity(), horseContainer.getOwner().getInventory(), this);
+   }
+   // CraftBukkit stop
+
+
    public HorseInventoryContainer(int p_i50077_1_, PlayerInventory p_i50077_2_, IInventory p_i50077_3_, final AbstractHorseEntity p_i50077_4_) {
       super((ContainerType<?>)null, p_i50077_1_);
       this.horseContainer = p_i50077_3_;
       this.horse = p_i50077_4_;
+      player = p_i50077_2_; // Craftbukkit
       int i = 3;
       p_i50077_3_.startOpen(p_i50077_2_.player);
       int j = -18;
@@ -66,7 +_,7 @@
    }
 
    public boolean stillValid(PlayerEntity p_75145_1_) {
-      return this.horseContainer.stillValid(p_75145_1_) && this.horse.isAlive() && this.horse.distanceTo(p_75145_1_) < 8.0F;
+      return this.horseContainer.stillValid(p_75145_1_) && (this.horse.isAlive() && this.horse.valid) && this.horse.distanceTo(p_75145_1_) < 8.0F; // Paper - Fix MC-161754
    }
 
    public ItemStack quickMoveStack(PlayerEntity p_82846_1_, int p_82846_2_) {
