--- a/net/minecraft/inventory/container/PlayerContainer.java
+++ b/net/minecraft/inventory/container/PlayerContainer.java
@@ -14,8 +_,11 @@
 import net.minecraft.item.crafting.RecipeBookCategory;
 import net.minecraft.item.crafting.RecipeItemHelper;
 import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.text.StringTextComponent;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryCrafting;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class PlayerContainer extends RecipeBookContainer<CraftingInventory> {
    public static final ResourceLocation BLOCK_ATLAS = new ResourceLocation("textures/atlas/blocks.png");
@@ -26,15 +_,27 @@
    public static final ResourceLocation EMPTY_ARMOR_SLOT_SHIELD = new ResourceLocation("item/empty_armor_slot_shield");
    private static final ResourceLocation[] TEXTURE_EMPTY_SLOTS = new ResourceLocation[]{EMPTY_ARMOR_SLOT_BOOTS, EMPTY_ARMOR_SLOT_LEGGINGS, EMPTY_ARMOR_SLOT_CHESTPLATE, EMPTY_ARMOR_SLOT_HELMET};
    private static final EquipmentSlotType[] SLOT_IDS = new EquipmentSlotType[]{EquipmentSlotType.HEAD, EquipmentSlotType.CHEST, EquipmentSlotType.LEGS, EquipmentSlotType.FEET};
-   private final CraftingInventory craftSlots = new CraftingInventory(this, 2, 2);
+   // CraftBukkit start
+   private final CraftingInventory craftSlots = new CraftingInventory(this, 2, 2); // LoliServer
    private final CraftResultInventory resultSlots = new CraftResultInventory();
+   // CraftBukkit end
    public final boolean active;
    private final PlayerEntity owner;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity = null;
+   private PlayerInventory playerCB;
+   // CraftBukkit end
 
    public PlayerContainer(PlayerInventory p_i1819_1_, boolean p_i1819_2_, PlayerEntity p_i1819_3_) {
       super((ContainerType<?>)null, 0);
       this.active = p_i1819_2_;
       this.owner = p_i1819_3_;
+      // CraftBukkit start
+      this.craftSlots.setOwner(p_i1819_1_.player); // LoliServer
+      this.craftSlots.resultInventory = this.resultSlots; // CraftBukkit - let InventoryCrafting know about its result slo
+      this.playerCB = p_i1819_1_; // CraftBukkit - save player
+      setTitle(new StringTextComponent("container.crafting")); // SPIGOT-4722: Allocate title for player inventory
+      // CraftBukkit end
       this.addSlot(new CraftingResultSlot(p_i1819_1_.player, this.craftSlots, this.resultSlots, 0, 154, 28));
 
       for(int i = 0; i < 2; ++i) {
@@ -51,7 +_,7 @@
             }
 
             public boolean mayPlace(ItemStack p_75214_1_) {
-               return equipmentslottype == MobEntity.getEquipmentSlotForItem(p_75214_1_);
+               return p_75214_1_.canEquip(equipmentslottype, owner);
             }
 
             public boolean mayPickup(PlayerEntity p_82869_1_) {
@@ -98,7 +_,7 @@
    }
 
    public void slotsChanged(IInventory p_75130_1_) {
-      WorkbenchContainer.slotChangedCraftingGrid(this.containerId, this.owner.level, this.owner, this.craftSlots, this.resultSlots);
+      WorkbenchContainer.slotChangedCraftingGrid(this.containerId, this.owner.level, this.owner, this.craftSlots, this.resultSlots, this);  // CraftBukkit
    }
 
    public void removed(PlayerEntity p_75134_1_) {
@@ -203,4 +_,16 @@
    public RecipeBookCategory getRecipeBookType() {
       return RecipeBookCategory.CRAFTING;
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      CraftInventoryCrafting inventory = new CraftInventoryCrafting(this.craftSlots, this.resultSlots);
+      bukkitEntity = new CraftInventoryView(this.playerCB.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
