--- a/net/minecraft/inventory/container/SmithingTableContainer.java
+++ b/net/minecraft/inventory/container/SmithingTableContainer.java
@@ -11,6 +_,7 @@
 import net.minecraft.item.crafting.SmithingRecipe;
 import net.minecraft.util.IWorldPosCallable;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventoryView;
 
 public class SmithingTableContainer extends AbstractRepairContainer {
    private final World level;
@@ -18,6 +_,10 @@
    private SmithingRecipe selectedRecipe;
    private final List<SmithingRecipe> recipes;
 
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity;
+   // CraftBukkit end
+
    public SmithingTableContainer(int p_i231590_1_, PlayerInventory p_i231590_2_) {
       this(p_i231590_1_, p_i231590_2_, IWorldPosCallable.NULL);
    }
@@ -56,12 +_,15 @@
    public void createResult() {
       List<SmithingRecipe> list = this.level.getRecipeManager().getRecipesFor(IRecipeType.SMITHING, this.inputSlots, this.level);
       if (list.isEmpty()) {
-         this.resultSlots.setItem(0, ItemStack.EMPTY);
+         org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), ItemStack.EMPTY); // CraftBukkit
       } else {
          this.selectedRecipe = list.get(0);
          ItemStack itemstack = this.selectedRecipe.assemble(this.inputSlots);
          this.resultSlots.setRecipeUsed(this.selectedRecipe);
-         this.resultSlots.setItem(0, itemstack);
+         // CraftBukkit start
+         this.inputSlots.setCurrentRecipe(this.selectedRecipe);
+         org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callPrepareSmithingEvent(getBukkitView(), itemstack);
+         // CraftBukkit end
       }
 
    }
@@ -75,4 +_,16 @@
    public boolean canTakeItemForPickAll(ItemStack p_94530_1_, Slot p_94530_2_) {
       return p_94530_2_.container != this.resultSlots && super.canTakeItemForPickAll(p_94530_1_, p_94530_2_);
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.v1_16_R3.inventory.CraftInventorySmithing(this.access.getLocation(), this.inputSlots, this.resultSlots);
+      bukkitEntity = new CraftInventoryView(this.player.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
