--- a/net/minecraft/tileentity/CampfireTileEntity.java
+++ b/net/minecraft/tileentity/CampfireTileEntity.java
@@ -5,6 +_,7 @@
 import javax.annotation.Nullable;
 import net.minecraft.block.BlockState;
 import net.minecraft.block.CampfireBlock;
+import net.minecraft.entity.item.ItemEntity;
 import net.minecraft.inventory.IClearable;
 import net.minecraft.inventory.IInventory;
 import net.minecraft.inventory.Inventory;
@@ -21,11 +_,14 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockCookEvent;
 
 public class CampfireTileEntity extends TileEntity implements IClearable, ITickableTileEntity {
    private final NonNullList<ItemStack> items = NonNullList.withSize(4, ItemStack.EMPTY);
-   private final int[] cookingProgress = new int[4];
-   private final int[] cookingTime = new int[4];
+   public final int[] cookingProgress = new int[4];
+   public final int[] cookingTime = new int[4];
 
    public CampfireTileEntity() {
       super(TileEntityType.CAMPFIRE);
@@ -64,7 +_,23 @@
                   return p_213979_1_.assemble(iinventory);
                }).orElse(itemstack);
                BlockPos blockpos = this.getBlockPos();
-               InventoryHelper.dropItemStack(this.level, (double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ(), itemstack1);
+
+               // CraftBukkit start - fire BlockCookEvent
+               CraftItemStack source = CraftItemStack.asCraftMirror(itemstack);
+               org.bukkit.inventory.ItemStack result = CraftItemStack.asBukkitCopy(itemstack1);
+               BlockCookEvent blockCookEvent = new BlockCookEvent(CraftBlock.at(this.level, this.worldPosition), source, result);
+               this.level.getCBServer().getPluginManager().callEvent(blockCookEvent);
+               if (blockCookEvent.isCancelled()) {
+                  return;
+               }
+               result = blockCookEvent.getResult();
+               itemstack1 = CraftItemStack.asNMSCopy(result);
+               // CraftBukkit end
+               // Paper start
+               ItemEntity droppedItem = new ItemEntity(this.level, blockpos.getX() + 0.5D, blockpos.getY() + 0.5D, blockpos.getZ() + 0.5D, itemstack1.split(this.level.random.nextInt(21) + 10));
+               droppedItem.setDeltaMovement(this.level.random.nextGaussian() * 0.05D, this.level.random.nextGaussian() * 0.05D + 0.2D, this.level.random.nextGaussian() * 0.05D);
+               this.level.addFreshEntity(droppedItem);
+               // Paper end
                this.items.set(i, ItemStack.EMPTY);
                this.markUpdated();
             }
