--- a/net/minecraft/tileentity/ConduitTileEntity.java
+++ b/net/minecraft/tileentity/ConduitTileEntity.java
@@ -26,6 +_,9 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.block.CraftBlock;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ConduitTileEntity extends TileEntity implements ITickableTileEntity {
    private static final Block[] VALID_BLOCKS = new Block[]{Blocks.PRISMARINE, Blocks.PRISMARINE_BRICKS, Blocks.SEA_LANTERN, Blocks.DARK_PRISMARINE};
@@ -131,7 +_,7 @@
                   BlockState blockstate = this.level.getBlockState(blockpos1);
 
                   for(Block block : VALID_BLOCKS) {
-                     if (blockstate.is(block)) {
+                     if (blockstate.isConduitFrame(this.level, blockpos1, getBlockPos())) {
                         this.effectBlocks.add(blockpos1);
                      }
                   }
@@ -155,7 +_,7 @@
       if (!list.isEmpty()) {
          for(PlayerEntity playerentity : list) {
             if (this.worldPosition.closerThan(playerentity.blockPosition(), (double)j) && playerentity.isInWaterOrRain()) {
-               playerentity.addEffect(new EffectInstance(Effects.CONDUIT_POWER, 260, 0, true, true));
+               playerentity.addEffect(new EffectInstance(Effects.CONDUIT_POWER, 260, 0, true, true), Cause.CONDUIT); // CraftBukkit
             }
          }
 
@@ -182,8 +_,13 @@
       }
 
       if (this.destroyTarget != null) {
+         // CraftBukkit start
+         CraftEventFactory.blockDamage = CraftBlock.at(this.level, this.worldPosition);
+         if (this.destroyTarget.hurt(DamageSource.MAGIC, 4.0F)) {
          this.level.playSound((PlayerEntity)null, this.destroyTarget.getX(), this.destroyTarget.getY(), this.destroyTarget.getZ(), SoundEvents.CONDUIT_ATTACK_TARGET, SoundCategory.BLOCKS, 1.0F, 1.0F);
-         this.destroyTarget.hurt(DamageSource.MAGIC, 4.0F);
+         }
+         CraftEventFactory.blockDamage = null;
+         // CraftBukkit end
       }
 
       if (livingentity != this.destroyTarget) {
