--- a/net/minecraft/tileentity/EndGatewayTileEntity.java
+++ b/net/minecraft/tileentity/EndGatewayTileEntity.java
@@ -20,6 +_,7 @@
 import net.minecraft.util.math.ChunkPos;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.vector.Vector3d;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.World;
 import net.minecraft.world.chunk.Chunk;
@@ -31,14 +_,18 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer;
+import org.bukkit.event.player.PlayerTeleportEvent;
 
 public class EndGatewayTileEntity extends EndPortalTileEntity implements ITickableTileEntity {
    private static final Logger LOGGER = LogManager.getLogger();
-   private long age;
+   public long age;
    private int teleportCooldown;
    @Nullable
-   private BlockPos exitPortal;
-   private boolean exactTeleport;
+   public BlockPos exitPortal;
+   public boolean exactTeleport;
 
    public EndGatewayTileEntity() {
       super(TileEntityType.END_GATEWAY);
@@ -148,7 +_,7 @@
    public void teleportEntity(Entity p_195496_1_) {
       if (this.level instanceof ServerWorld && !this.isCoolingDown()) {
          this.teleportCooldown = 100;
-         if (this.exitPortal == null && this.level.dimension() == World.END) {
+         if (this.exitPortal == null && this.level.getTypeKey() == DimensionType.END_LOCATION) { // CraftBukkit - work in alternate worlds
             this.findExitPortal((ServerWorld)this.level);
          }
 
@@ -170,6 +_,24 @@
             } else {
                entity = p_195496_1_.getRootVehicle();
             }
+
+
+            // CraftBukkit start - Fire PlayerTeleportEvent
+            if (p_195496_1_ instanceof ServerPlayerEntity) {
+               org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer player = (CraftPlayer) p_195496_1_.getBukkitEntity();
+               org.bukkit.Location location = new Location(level.getWorld(), (double) blockpos.getX() + 0.5D, (double) blockpos.getY() + 0.5D, (double) blockpos.getZ() + 0.5D);
+               location.setPitch(player.getLocation().getPitch());
+               location.setYaw(player.getLocation().getYaw());
+               PlayerTeleportEvent teleEvent = new PlayerTeleportEvent(player, player.getLocation(), location, PlayerTeleportEvent.TeleportCause.END_GATEWAY);
+               Bukkit.getPluginManager().callEvent(teleEvent);
+               if (teleEvent.isCancelled()) {
+                  return;
+               }
+               ((ServerPlayerEntity) p_195496_1_).connection.teleport(teleEvent.getTo());
+               this.isCoolingDown(); // CraftBukkit - call at end of method
+               return;
+            }
+            // CraftBukkit end
 
             entity.setPortalCooldown();
             entity.teleportToWithTicket((double)blockpos.getX() + 0.5D, (double)blockpos.getY(), (double)blockpos.getZ() + 0.5D);
