--- a/net/minecraft/tileentity/StructureBlockTileEntity.java
+++ b/net/minecraft/tileentity/StructureBlockTileEntity.java
@@ -35,19 +_,19 @@
 
 public class StructureBlockTileEntity extends TileEntity {
    private ResourceLocation structureName;
-   private String author = "";
-   private String metaData = "";
-   private BlockPos structurePos = new BlockPos(0, 1, 0);
-   private BlockPos structureSize = BlockPos.ZERO;
-   private Mirror mirror = Mirror.NONE;
-   private Rotation rotation = Rotation.NONE;
-   private StructureMode mode = StructureMode.DATA;
-   private boolean ignoreEntities = true;
+   public String author = "";
+   public String metaData = "";
+   public BlockPos structurePos = new BlockPos(0, 1, 0);
+   public BlockPos structureSize = BlockPos.ZERO;
+   public Mirror mirror = Mirror.NONE;
+   public Rotation rotation = Rotation.NONE;
+   public StructureMode mode = StructureMode.DATA;
+   public boolean ignoreEntities = true;
    private boolean powered;
-   private boolean showAir;
-   private boolean showBoundingBox = true;
-   private float integrity = 1.0F;
-   private long seed;
+   public boolean showAir;
+   public boolean showBoundingBox = true;
+   public float integrity = 1.0F;
+   public long seed;
 
    public StructureBlockTileEntity() {
       super(TileEntityType.STRUCTURE_BLOCK);
