--- a/net/minecraft/client/renderer/entity/model/SkeletonModel.java
+++ b/net/minecraft/client/renderer/entity/model/SkeletonModel.java
@@ -58,7 +_,7 @@
    public void setupAnim(T p_225597_1_, float p_225597_2_, float p_225597_3_, float p_225597_4_, float p_225597_5_, float p_225597_6_) {
       super.setupAnim(p_225597_1_, p_225597_2_, p_225597_3_, p_225597_4_, p_225597_5_, p_225597_6_);
       ItemStack itemstack = p_225597_1_.getMainHandItem();
-      if (p_225597_1_.isAggressive() && (itemstack.isEmpty() || itemstack.getItem() != Items.BOW)) {
+      if (p_225597_1_.isAggressive() && (itemstack.isEmpty() || !(itemstack.getItem() == Items.BOW))) {
          float f = MathHelper.sin(this.attackTime * (float)Math.PI);
          float f1 = MathHelper.sin((1.0F - (1.0F - this.attackTime) * (1.0F - this.attackTime)) * (float)Math.PI);
          this.rightArm.zRot = 0.0F;
