--- a/net/minecraft/world/gen/surfacebuilders/NetherSurfaceBuilder.java
+++ b/net/minecraft/world/gen/surfacebuilders/NetherSurfaceBuilder.java
@@ -35,7 +_,7 @@
       BlockState blockstate = p_205610_14_.getTopMaterial();
       BlockState blockstate1 = p_205610_14_.getUnderMaterial();
 
-      for(int j1 = 127; j1 >= 0; --j1) {
+      for(int j1 = p_205610_6_; j1 >= 0; --j1) {
          blockpos$mutable.set(j, j1, k);
          BlockState blockstate2 = p_205610_2_.getBlockState(blockpos$mutable);
          if (blockstate2.isAir()) {
