--- a/net/minecraft/entity/passive/fish/AbstractFishEntity.java
+++ b/net/minecraft/entity/passive/fish/AbstractFishEntity.java
@@ -64,7 +_,7 @@
    }
 
    public boolean removeWhenFarAway(double p_213397_1_) {
-      return !this.fromBucket() && !this.hasCustomName();
+      return true; // CraftBukkit
    }
 
    public int getMaxSpawnClusterSize() {
@@ -82,6 +_,7 @@
 
    public void setFromBucket(boolean p_203706_1_) {
       this.entityData.set(FROM_BUCKET, p_203706_1_);
+      this.persistenceRequired = this.isPersistenceRequired();  // CraftBukkit - SPIGOT-4106 update persistence
    }
 
    public void addAdditionalSaveData(CompoundNBT p_213281_1_) {
