--- a/net/minecraft/entity/passive/fish/PufferfishEntity.java
+++ b/net/minecraft/entity/passive/fish/PufferfishEntity.java
@@ -24,6 +_,7 @@
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class PufferfishEntity extends AbstractFishEntity {
    private static final DataParameter<Integer> PUFF_STATE = EntityDataManager.defineId(PufferfishEntity.class, DataSerializers.INT);
@@ -126,7 +_,7 @@
    private void touch(MobEntity p_205719_1_) {
       int i = this.getPuffState();
       if (p_205719_1_.hurt(DamageSource.mobAttack(this), (float)(1 + i))) {
-         p_205719_1_.addEffect(new EffectInstance(Effects.POISON, 60 * i, 0));
+         p_205719_1_.addEffect(new EffectInstance(Effects.POISON, 60 * i, 0), Cause.ATTACK); // CraftBukkit
          this.playSound(SoundEvents.PUFFER_FISH_STING, 1.0F, 1.0F);
       }
 
@@ -139,7 +_,7 @@
             ((ServerPlayerEntity)p_70100_1_).connection.send(new SChangeGameStatePacket(SChangeGameStatePacket.PUFFER_FISH_STING, 0.0F));
          }
 
-         p_70100_1_.addEffect(new EffectInstance(Effects.POISON, 60 * i, 0));
+         p_70100_1_.addEffect(new EffectInstance(Effects.POISON, 60 * i, 0), Cause.ATTACK); // CraftBukkit
       }
 
    }
