--- a/net/minecraft/network/rcon/RConConsoleSource.java
+++ b/net/minecraft/network/rcon/RConConsoleSource.java
@@ -10,6 +_,7 @@
 import net.minecraft.util.text.ITextComponent;
 import net.minecraft.util.text.StringTextComponent;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.command.CommandSender;
 
 public class RConConsoleSource implements ICommandSource {
    private static final StringTextComponent RCON_COMPONENT = new StringTextComponent("Rcon");
@@ -33,8 +_,19 @@
       return new CommandSource(this, Vector3d.atLowerCornerOf(serverworld.getSharedSpawnPos()), Vector2f.ZERO, serverworld, 4, "Rcon", RCON_COMPONENT, this.server, (Entity)null);
    }
 
+   // CraftBukkit start - Send a String
+   public void sendMessage(String message) {
+      this.buffer.append(message);
+   }
+
+   @Override
+   public CommandSender getBukkitSender(CommandSource p0) {
+      return server.remoteConsole;
+   }
+   // CraftBukkit end
+
    public void sendMessage(ITextComponent p_145747_1_, UUID p_145747_2_) {
-      this.buffer.append(p_145747_1_.getString());
+      this.buffer.append(p_145747_1_.getString()).append("\n"); // FIX MC-7569 - RCON has no newlines in multiline output
    }
 
    public boolean acceptsSuccess() {
