--- a/net/minecraft/entity/monster/AbstractRaiderEntity.java
+++ b/net/minecraft/entity/monster/AbstractRaiderEntity.java
@@ -43,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public abstract class AbstractRaiderEntity extends PatrollerEntity {
    protected static final DataParameter<Boolean> IS_CELEBRATING = EntityDataManager.defineId(AbstractRaiderEntity.class, DataSerializers.BOOLEAN);
@@ -151,7 +_,7 @@
                i = MathHelper.clamp(i, 0, 4);
                EffectInstance effectinstance = new EffectInstance(Effects.BAD_OMEN, 120000, i, false, false, true);
                if (!this.level.getGameRules().getBoolean(GameRules.RULE_DISABLE_RAIDS)) {
-                  playerentity.addEffect(effectinstance);
+                  playerentity.addEffect(effectinstance, Cause.PATROL_CAPTAIN); // CraftBukkit
                }
             }
          }
@@ -335,7 +_,7 @@
          this.mob.getNavigation().stop();
 
          for(AbstractRaiderEntity abstractraiderentity : this.mob.level.getNearbyEntities(AbstractRaiderEntity.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0D, 8.0D, 8.0D))) {
-            abstractraiderentity.setTarget(this.mob.getTarget());
+            abstractraiderentity.setGoalTarget(this.mob.getTarget(), TargetReason.FOLLOW_LEADER, true); // CraftBukkit
          }
 
       }
@@ -345,7 +_,7 @@
          LivingEntity livingentity = this.mob.getTarget();
          if (livingentity != null) {
             for(AbstractRaiderEntity abstractraiderentity : this.mob.level.getNearbyEntities(AbstractRaiderEntity.class, this.shoutTargeting, this.mob, this.mob.getBoundingBox().inflate(8.0D, 8.0D, 8.0D))) {
-               abstractraiderentity.setTarget(livingentity);
+               abstractraiderentity.setGoalTarget(livingentity, TargetReason.FOLLOW_LEADER, true); // CraftBukkit
                abstractraiderentity.setAggressive(true);
             }
 
