--- a/net/minecraft/entity/monster/CreeperEntity.java
+++ b/net/minecraft/entity/monster/CreeperEntity.java
@@ -34,11 +_,14 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.Explosion;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason;
+import org.bukkit.event.entity.CreeperPowerEvent;
+import org.bukkit.event.entity.ExplosionPrimeEvent;
 
 @OnlyIn(
    value = Dist.CLIENT,
@@ -49,9 +_,9 @@
    private static final DataParameter<Boolean> DATA_IS_POWERED = EntityDataManager.defineId(CreeperEntity.class, DataSerializers.BOOLEAN);
    private static final DataParameter<Boolean> DATA_IS_IGNITED = EntityDataManager.defineId(CreeperEntity.class, DataSerializers.BOOLEAN);
    private int oldSwell;
-   private int swell;
-   private int maxSwell = 30;
-   private int explosionRadius = 3;
+   public int swell;
+   public int maxSwell = 30;
+   public int explosionRadius = 3;
    private int droppedSkulls;
 
    public CreeperEntity(EntityType<? extends CreeperEntity> p_i50213_1_, World p_i50213_2_) {
@@ -194,8 +_,17 @@
 
    public void thunderHit(ServerWorld p_241841_1_, LightningBoltEntity p_241841_2_) {
       super.thunderHit(p_241841_1_, p_241841_2_);
-      this.entityData.set(DATA_IS_POWERED, true);
-   }
+      // CraftBukkit start
+      if (CraftEventFactory.callCreeperPowerEvent(this, p_241841_2_, CreeperPowerEvent.PowerCause.LIGHTNING).isCancelled()) {
+         return;
+      }
+      this.setPowered(true);
+   }
+
+   public void setPowered(boolean powered) {
+      this.entityData.set(DATA_IS_POWERED, powered);
+   }
+   // CraftBukkit end
 
    protected ActionResultType mobInteract(PlayerEntity p_230254_1_, Hand p_230254_2_) {
       ItemStack itemstack = p_230254_1_.getItemInHand(p_230254_2_);
@@ -214,14 +_,22 @@
       }
    }
 
-   private void explodeCreeper() {
+   public void explodeCreeper() {
       if (!this.level.isClientSide) {
-         Explosion.Mode explosion$mode = this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) ? Explosion.Mode.DESTROY : Explosion.Mode.NONE;
+         Explosion.Mode explosion$mode = net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.level, this) ? Explosion.Mode.DESTROY : Explosion.Mode.NONE;
          float f = this.isPowered() ? 2.0F : 1.0F;
+         // CraftBukkit start
+         ExplosionPrimeEvent event = new ExplosionPrimeEvent(this.getBukkitEntity(), this.explosionRadius * f, false);
+         this.level.getCBServer().getPluginManager().callEvent(event);
+         if (!event.isCancelled()) {
          this.dead = true;
-         this.level.explode(this, this.getX(), this.getY(), this.getZ(), (float)this.explosionRadius * f, explosion$mode);
+            this.level.explode(this, this.getX(), this.getY(), this.getZ(), event.getRadius(), event.getFire(), explosion$mode);
          this.remove();
          this.spawnLingeringCloud();
+         } else {
+            swell = 0;
+         }
+         // CraftBukkit end
       }
 
    }
@@ -230,6 +_,7 @@
       Collection<EffectInstance> collection = this.getActiveEffects();
       if (!collection.isEmpty()) {
          AreaEffectCloudEntity areaeffectcloudentity = new AreaEffectCloudEntity(this.level, this.getX(), this.getY(), this.getZ());
+         areaeffectcloudentity.setOwner(this); // CraftBukkit
          areaeffectcloudentity.setRadius(2.5F);
          areaeffectcloudentity.setRadiusOnUse(-0.5F);
          areaeffectcloudentity.setWaitTime(10);
@@ -240,7 +_,7 @@
             areaeffectcloudentity.addEffect(new EffectInstance(effectinstance));
          }
 
-         this.level.addFreshEntity(areaeffectcloudentity);
+         this.level.addEntity(areaeffectcloudentity, SpawnReason.EXPLOSION); // CraftBukkit - reason
       }
 
    }
