--- a/net/minecraft/entity/monster/ElderGuardianEntity.java
+++ b/net/minecraft/entity/monster/ElderGuardianEntity.java
@@ -14,6 +_,7 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ElderGuardianEntity extends GuardianEntity {
    public static final float ELDER_SIZE_SCALE = EntityType.ELDER_GUARDIAN.getWidth() / EntityType.GUARDIAN.getWidth();
@@ -66,7 +_,7 @@
          for(ServerPlayerEntity serverplayerentity : list) {
             if (!serverplayerentity.hasEffect(effect) || serverplayerentity.getEffect(effect).getAmplifier() < 2 || serverplayerentity.getEffect(effect).getDuration() < 1200) {
                serverplayerentity.connection.send(new SChangeGameStatePacket(SChangeGameStatePacket.GUARDIAN_ELDER_EFFECT, this.isSilent() ? 0.0F : 1.0F));
-               serverplayerentity.addEffect(new EffectInstance(effect, 6000, 2));
+               serverplayerentity.addEffect(new EffectInstance(effect, 6000, 2), Cause.ATTACK); // CraftBukkit
             }
          }
       }
