--- a/net/minecraft/entity/monster/GhastEntity.java
+++ b/net/minecraft/entity/monster/GhastEntity.java
@@ -176,7 +_,8 @@
                }
 
                FireballEntity fireballentity = new FireballEntity(world, this.ghast, d2, d3, d4);
-               fireballentity.explosionPower = this.ghast.getExplosionPower();
+               // CraftBukkit - set bukkitYield when setting explosionPower
+               fireballentity.bukkitYield = fireballentity.explosionPower = this.ghast.getExplosionPower();
                fireballentity.setPos(this.ghast.getX() + vector3d.x * 4.0D, this.ghast.getY(0.5D) + 0.5D, fireballentity.getZ() + vector3d.z * 4.0D);
                world.addFreshEntity(fireballentity);
                this.chargeTime = -40;
