--- a/net/minecraft/entity/monster/HoglinEntity.java
+++ b/net/minecraft/entity/monster/HoglinEntity.java
@@ -48,8 +_,8 @@
 public class HoglinEntity extends AnimalEntity implements IMob, IFlinging {
    private static final DataParameter<Boolean> DATA_IMMUNE_TO_ZOMBIFICATION = EntityDataManager.defineId(HoglinEntity.class, DataSerializers.BOOLEAN);
    private int attackAnimationRemainingTicks;
-   private int timeInOverworld = 0;
-   private boolean cannotBeHunted = false;
+   public int timeInOverworld = 0;
+   public boolean cannotBeHunted = false;
    protected static final ImmutableList<? extends SensorType<? extends Sensor<? super HoglinEntity>>> SENSOR_TYPES = ImmutableList.of(SensorType.NEAREST_LIVING_ENTITIES, SensorType.NEAREST_PLAYERS, SensorType.NEAREST_ADULT, SensorType.HOGLIN_SPECIFIC_SENSOR);
    protected static final ImmutableList<? extends MemoryModuleType<?>> MEMORY_TYPES = ImmutableList.of(MemoryModuleType.BREED_TARGET, MemoryModuleType.LIVING_ENTITIES, MemoryModuleType.VISIBLE_LIVING_ENTITIES, MemoryModuleType.NEAREST_VISIBLE_PLAYER, MemoryModuleType.NEAREST_VISIBLE_TARGETABLE_PLAYER, MemoryModuleType.LOOK_TARGET, MemoryModuleType.WALK_TARGET, MemoryModuleType.CANT_REACH_WALK_TARGET_SINCE, MemoryModuleType.PATH, MemoryModuleType.ATTACK_TARGET, MemoryModuleType.ATTACK_COOLING_DOWN, MemoryModuleType.NEAREST_VISIBLE_ADULT_PIGLIN, MemoryModuleType.AVOID_TARGET, MemoryModuleType.VISIBLE_ADULT_PIGLIN_COUNT, MemoryModuleType.VISIBLE_ADULT_HOGLIN_COUNT, MemoryModuleType.NEAREST_VISIBLE_ADULT_HOGLINS, MemoryModuleType.NEAREST_VISIBLE_ADULT, MemoryModuleType.NEAREST_REPELLENT, MemoryModuleType.PACIFIED);
 
@@ -117,7 +_,7 @@
       HoglinTasks.updateActivity(this);
       if (this.isConverting()) {
          ++this.timeInOverworld;
-         if (this.timeInOverworld > 300) {
+         if (this.timeInOverworld > 300 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.ZOGLIN, (timer) -> this.timeInOverworld = timer)) {
             this.playSound(SoundEvents.HOGLIN_CONVERTED_TO_ZOMBIFIED);
             this.finishConversion((ServerWorld)this.level);
          }
@@ -212,6 +_,7 @@
       ZoglinEntity zoglinentity = this.convertTo(EntityType.ZOGLIN, true);
       if (zoglinentity != null) {
          zoglinentity.addEffect(new EffectInstance(Effects.CONFUSION, 200, 0));
+         net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, zoglinentity);
       }
 
    }
@@ -253,7 +_,7 @@
       this.getEntityData().set(DATA_IMMUNE_TO_ZOMBIFICATION, p_234370_1_);
    }
 
-   private boolean isImmuneToZombification() {
+   public boolean isImmuneToZombification() {
       return this.getEntityData().get(DATA_IMMUNE_TO_ZOMBIFICATION);
    }
 
