--- a/net/minecraft/entity/monster/HuskEntity.java
+++ b/net/minecraft/entity/monster/HuskEntity.java
@@ -15,6 +_,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class HuskEntity extends ZombieEntity {
    public HuskEntity(EntityType<? extends HuskEntity> p_i50204_1_, World p_i50204_2_) {
@@ -49,7 +_,7 @@
       boolean flag = super.doHurtTarget(p_70652_1_);
       if (flag && this.getMainHandItem().isEmpty() && p_70652_1_ instanceof LivingEntity) {
          float f = this.level.getCurrentDifficultyAt(this.blockPosition()).getEffectiveDifficulty();
-         ((LivingEntity)p_70652_1_).addEffect(new EffectInstance(Effects.HUNGER, 140 * (int)f));
+         ((LivingEntity)p_70652_1_).addEffect(new EffectInstance(Effects.HUNGER, 140 * (int)f), Cause.ATTACK); // CraftBukkit
       }
 
       return flag;
