--- a/net/minecraft/entity/monster/IllusionerEntity.java
+++ b/net/minecraft/entity/monster/IllusionerEntity.java
@@ -23,6 +_,7 @@
 import net.minecraft.entity.projectile.AbstractArrowEntity;
 import net.minecraft.entity.projectile.ProjectileHelper;
 import net.minecraft.inventory.EquipmentSlotType;
+import net.minecraft.item.BowItem;
 import net.minecraft.item.ItemStack;
 import net.minecraft.item.Items;
 import net.minecraft.nbt.CompoundNBT;
@@ -41,6 +_,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class IllusionerEntity extends SpellcastingIllagerEntity implements IRangedAttackMob {
    private int clientSideIllusionTicks;
@@ -180,8 +_,10 @@
    }
 
    public void performRangedAttack(LivingEntity p_82196_1_, float p_82196_2_) {
-      ItemStack itemstack = this.getProjectile(this.getItemInHand(ProjectileHelper.getWeaponHoldingHand(this, Items.BOW)));
+      ItemStack itemstack = this.getProjectile(this.getItemInHand(ProjectileHelper.getWeaponHoldingHand(this, item -> item instanceof net.minecraft.item.BowItem)));
       AbstractArrowEntity abstractarrowentity = ProjectileHelper.getMobArrow(this, itemstack, p_82196_2_);
+      if (this.getMainHandItem().getItem() instanceof net.minecraft.item.BowItem)
+         abstractarrowentity = ((net.minecraft.item.BowItem)this.getMainHandItem().getItem()).customArrow(abstractarrowentity);
       double d0 = p_82196_1_.getX() - this.getX();
       double d1 = p_82196_1_.getY(0.3333333333333333D) - abstractarrowentity.getY();
       double d2 = p_82196_1_.getZ() - this.getZ();
@@ -232,7 +_,7 @@
       }
 
       protected void performSpellCasting() {
-         IllusionerEntity.this.getTarget().addEffect(new EffectInstance(Effects.BLINDNESS, 400));
+         IllusionerEntity.this.getTarget().addEffect(new EffectInstance(Effects.BLINDNESS, 400), Cause.ATTACK); // CraftBukkit
       }
 
       protected SoundEvent getSpellPrepareSound() {
@@ -265,7 +_,7 @@
       }
 
       protected void performSpellCasting() {
-         IllusionerEntity.this.addEffect(new EffectInstance(Effects.INVISIBILITY, 1200));
+         IllusionerEntity.this.addEffect(new EffectInstance(Effects.INVISIBILITY, 1200), Cause.ILLUSION); // Craftbukkit
       }
 
       @Nullable
