--- a/net/minecraft/entity/monster/MagmaCubeEntity.java
+++ b/net/minecraft/entity/monster/MagmaCubeEntity.java
@@ -39,7 +_,7 @@
       return p_205019_1_.isUnobstructed(this) && !p_205019_1_.containsAnyLiquid(this.getBoundingBox());
    }
 
-   protected void setSize(int p_70799_1_, boolean p_70799_2_) {
+   public void setSize(int p_70799_1_, boolean p_70799_2_) {
       super.setSize(p_70799_1_, p_70799_2_);
       this.getAttribute(Attributes.ARMOR).setBaseValue((double)(p_70799_1_ * 3));
    }
@@ -52,7 +_,7 @@
       return ParticleTypes.FLAME;
    }
 
-   protected ResourceLocation getDefaultLootTable() {
+   public ResourceLocation getDefaultLootTable() {
       return this.isTiny() ? LootTables.EMPTY : this.getType().getDefaultLootTable();
    }
 
@@ -72,6 +_,7 @@
       Vector3d vector3d = this.getDeltaMovement();
       this.setDeltaMovement(vector3d.x, (double)(this.getJumpPower() + (float)this.getSize() * 0.1F), vector3d.z);
       this.hasImpulse = true;
+      net.minecraftforge.common.ForgeHooks.onLivingJump(this);
    }
 
    protected void jumpInLiquid(ITag<Fluid> p_180466_1_) {
