--- a/net/minecraft/entity/monster/PhantomEntity.java
+++ b/net/minecraft/entity/monster/PhantomEntity.java
@@ -41,6 +_,7 @@
 import net.minecraft.world.gen.Heightmap;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class PhantomEntity extends FlyingEntity implements IMob {
    private static final DataParameter<Integer> ID_SIZE = EntityDataManager.defineId(PhantomEntity.class, DataSerializers.INT);
@@ -218,7 +_,7 @@
 
                for(PlayerEntity playerentity : list) {
                   if (PhantomEntity.this.canAttack(playerentity, EntityPredicate.DEFAULT)) {
-                     PhantomEntity.this.setTarget(playerentity);
+                     PhantomEntity.this.setGoalTarget(playerentity, TargetReason.CLOSEST_PLAYER, true); // CraftBukkit - reason
                      return true;
                   }
                }
