--- a/net/minecraft/entity/monster/PillagerEntity.java
+++ b/net/minecraft/entity/monster/PillagerEntity.java
@@ -56,7 +_,7 @@
 
 public class PillagerEntity extends AbstractIllagerEntity implements ICrossbowUser {
    private static final DataParameter<Boolean> IS_CHARGING_CROSSBOW = EntityDataManager.defineId(PillagerEntity.class, DataSerializers.BOOLEAN);
-   private final Inventory inventory = new Inventory(5);
+   public final Inventory inventory = new Inventory(5);
 
    public PillagerEntity(EntityType<? extends PillagerEntity> p_i50198_1_, World p_i50198_2_) {
       super(p_i50198_1_, p_i50198_2_);
@@ -120,7 +_,7 @@
    public AbstractIllagerEntity.ArmPose getArmPose() {
       if (this.isChargingCrossbow()) {
          return AbstractIllagerEntity.ArmPose.CROSSBOW_CHARGE;
-      } else if (this.isHolding(Items.CROSSBOW)) {
+      } else if (this.isHolding(item -> item instanceof net.minecraft.item.CrossbowItem)) {
          return AbstractIllagerEntity.ArmPose.CROSSBOW_HOLD;
       } else {
          return this.isAggressive() ? AbstractIllagerEntity.ArmPose.ATTACKING : AbstractIllagerEntity.ArmPose.NEUTRAL;
