--- a/net/minecraft/entity/monster/SilverfishEntity.java
+++ b/net/minecraft/entity/monster/SilverfishEntity.java
@@ -4,6 +_,7 @@
 import java.util.Random;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
 import net.minecraft.block.SilverfishBlock;
 import net.minecraft.entity.CreatureAttribute;
 import net.minecraft.entity.EntitySize;
@@ -25,10 +_,10 @@
 import net.minecraft.util.SoundEvent;
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class SilverfishEntity extends MonsterEntity {
    private SilverfishEntity.SummonSilverfishGoal friendsGoal;
@@ -134,7 +_,7 @@
             return false;
          } else {
             Random random = this.mob.getRandom();
-            if (this.mob.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING) && random.nextInt(10) == 0) {
+            if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(this.mob.level, this.mob) && random.nextInt(10) == 0) {
                this.selectedDirection = Direction.getRandom(random);
                BlockPos blockpos = (new BlockPos(this.mob.getX(), this.mob.getY() + 0.5D, this.mob.getZ())).relative(this.selectedDirection);
                BlockState blockstate = this.mob.level.getBlockState(blockpos);
@@ -161,6 +_,11 @@
             BlockPos blockpos = (new BlockPos(this.mob.getX(), this.mob.getY() + 0.5D, this.mob.getZ())).relative(this.selectedDirection);
             BlockState blockstate = iworld.getBlockState(blockpos);
             if (SilverfishBlock.isCompatibleHostBlock(blockstate)) {
+               // CraftBukkit start
+               if (CraftEventFactory.callEntityChangeBlockEvent(this.mob, blockpos, SilverfishBlock.stateByHostBlock(blockstate.getBlock())).isCancelled()) {
+                  return;
+               }
+               // CraftBukkit end
                iworld.setBlock(blockpos, SilverfishBlock.stateByHostBlock(blockstate.getBlock()), 3);
                this.mob.spawnAnim();
                this.mob.remove();
@@ -203,7 +_,12 @@
                      BlockState blockstate = world.getBlockState(blockpos1);
                      Block block = blockstate.getBlock();
                      if (block instanceof SilverfishBlock) {
-                        if (world.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+                        // CraftBukkit start
+                        if (CraftEventFactory.callEntityChangeBlockEvent(this.silverfish, blockpos1, Blocks.AIR.defaultBlockState()).isCancelled()) {
+                           continue;
+                        }
+                        // CraftBukkit end
+                        if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(world, this.silverfish)) {
                            world.destroyBlock(blockpos1, true, this.silverfish);
                         } else {
                            world.setBlock(blockpos1, ((SilverfishBlock)block).getHostBlock().defaultBlockState(), 3);
