--- a/net/minecraft/entity/monster/SpellcastingIllagerEntity.java
+++ b/net/minecraft/entity/monster/SpellcastingIllagerEntity.java
@@ -62,7 +_,7 @@
       this.entityData.set(DATA_SPELL_CASTING_ID, (byte)p_193081_1_.id);
    }
 
-   protected SpellcastingIllagerEntity.SpellType getCurrentSpell() {
+   public SpellcastingIllagerEntity.SpellType getCurrentSpell() {
       return !this.level.isClientSide ? this.currentSpell : SpellcastingIllagerEntity.SpellType.byId(this.entityData.get(DATA_SPELL_CASTING_ID));
    }
 
@@ -190,6 +_,11 @@
       public void tick() {
          --this.attackWarmupDelay;
          if (this.attackWarmupDelay == 0) {
+            // CraftBukkit start
+            if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.handleEntitySpellCastEvent(SpellcastingIllagerEntity.this, getSpell())) { // PAIL rename getSpell
+               return;
+            }
+            // CraftBukkit end
             this.performSpellCasting();
             SpellcastingIllagerEntity.this.playSound(SpellcastingIllagerEntity.this.getCastingSoundEvent(), 1.0F, 1.0F);
          }
