--- a/net/minecraft/entity/monster/SpiderEntity.java
+++ b/net/minecraft/entity/monster/SpiderEntity.java
@@ -41,6 +_,7 @@
 import net.minecraft.world.DifficultyInstance;
 import net.minecraft.world.IServerWorld;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class SpiderEntity extends MonsterEntity {
    private static final DataParameter<Byte> DATA_FLAGS_ID = EntityDataManager.defineId(SpiderEntity.class, DataSerializers.BYTE);
@@ -118,7 +_,12 @@
    }
 
    public boolean canBeAffected(EffectInstance p_70687_1_) {
-      return p_70687_1_.getEffect() == Effects.POISON ? false : super.canBeAffected(p_70687_1_);
+      if (p_70687_1_.getEffect() == Effects.POISON) {
+         net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent event = new net.minecraftforge.event.entity.living.PotionEvent.PotionApplicableEvent(this, p_70687_1_);
+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         return event.getResult() == net.minecraftforge.eventbus.api.Event.Result.ALLOW;
+      }
+      return super.canBeAffected(p_70687_1_);
    }
 
    public boolean isClimbing() {
@@ -156,7 +_,7 @@
       if (p_213386_4_ instanceof SpiderEntity.GroupData) {
          Effect effect = ((SpiderEntity.GroupData)p_213386_4_).effect;
          if (effect != null) {
-            this.addEffect(new EffectInstance(effect, Integer.MAX_VALUE));
+            this.addEffect(new EffectInstance(effect, Integer.MAX_VALUE), Cause.SPIDER_SPAWN); // CraftBukkit
          }
       }
 
