--- a/net/minecraft/entity/monster/ZombifiedPiglinEntity.java
+++ b/net/minecraft/entity/monster/ZombifiedPiglinEntity.java
@@ -4,6 +_,7 @@
 import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.block.Blocks;
+import net.minecraft.entity.Entity;
 import net.minecraft.entity.EntityType;
 import net.minecraft.entity.IAngerable;
 import net.minecraft.entity.LivingEntity;
@@ -36,6 +_,7 @@
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.event.entity.EntityTargetEvent;
 
 public class ZombifiedPiglinEntity extends ZombieEntity implements IAngerable {
    private static final UUID SPEED_MODIFIER_ATTACKING_UUID = UUID.fromString("49455A49-7EC5-45BA-B886-3B90B23A1718");
@@ -133,7 +_,7 @@
       }).filter((p_241406_1_) -> {
          return !p_241406_1_.isAlliedTo(this.getTarget());
       }).forEach((p_241405_1_) -> {
-         p_241405_1_.setTarget(this.getTarget());
+         p_241405_1_.setAttackTarget(this.getTarget(), org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_ATTACKED_NEARBY_ENTITY, true); // CraftBukkit
       });
    }
 
@@ -141,21 +_,30 @@
       this.playSound(SoundEvents.ZOMBIFIED_PIGLIN_ANGRY, this.getSoundVolume() * 2.0F, this.getVoicePitch() * 1.8F);
    }
 
-   public void setTarget(@Nullable LivingEntity p_70624_1_) {
-      if (this.getTarget() == null && p_70624_1_ != null) {
+   public boolean setAttackTarget(@Nullable LivingEntity entitylivingbaseIn, EntityTargetEvent.TargetReason reason, final boolean fireEvent) {
+      if (this.getTarget() == null && entitylivingbaseIn != null) {
          this.playFirstAngerSoundIn = FIRST_ANGER_SOUND_DELAY.randomValue(this.random);
          this.ticksUntilNextAlert = ALERT_INTERVAL.randomValue(this.random);
       }
 
-      if (p_70624_1_ instanceof PlayerEntity) {
-         this.setLastHurtByPlayer((PlayerEntity)p_70624_1_);
+      if (entitylivingbaseIn instanceof PlayerEntity) {
+         this.setLastHurtByPlayer((PlayerEntity)entitylivingbaseIn);
       }
 
-      super.setTarget(p_70624_1_);
+      return super.setGoalTarget(entitylivingbaseIn, reason, fireEvent);
    }
 
    public void startPersistentAngerTimer() {
-      this.setRemainingPersistentAngerTime(PERSISTENT_ANGER_TIME.randomValue(this.random));
+      // CraftBukkit start
+      Entity entity = ((ServerWorld) this.level).getEntity(getPersistentAngerTarget());
+      org.bukkit.event.entity.PigZombieAngerEvent event = new org.bukkit.event.entity.PigZombieAngerEvent((org.bukkit.entity.PigZombie) this.getBukkitEntity(), (entity == null) ? null : entity.getBukkitEntity(), PERSISTENT_ANGER_TIME.randomValue(this.random));
+      this.level.getCBServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) {
+         this.setPersistentAngerTarget(null);
+         return;
+      }
+      this.setRemainingPersistentAngerTime(event.getNewAnger());
+      // CraftBukkit end
    }
 
    public static boolean checkZombifiedPiglinSpawnRules(EntityType<ZombifiedPiglinEntity> p_234351_0_, IWorld p_234351_1_, SpawnReason p_234351_2_, BlockPos p_234351_3_, Random p_234351_4_) {
@@ -173,6 +_,7 @@
 
    public void readAdditionalSaveData(CompoundNBT p_70037_1_) {
       super.readAdditionalSaveData(p_70037_1_);
+      if(!level.isClientSide) //FORGE: allow this entity to be read from nbt on client. (Fixes MC-189565)
       this.readPersistentAngerSaveData((ServerWorld)this.level, p_70037_1_);
    }
 
