--- a/net/minecraft/entity/monster/ZombieVillagerEntity.java
+++ b/net/minecraft/entity/monster/ZombieVillagerEntity.java
@@ -32,6 +_,7 @@
 import net.minecraft.network.datasync.EntityDataManager;
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.potion.Effects;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.util.ActionResultType;
 import net.minecraft.util.DamageSource;
 import net.minecraft.util.Hand;
@@ -45,15 +_,19 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.entity.ZombieVillager;
+import org.bukkit.event.entity.CreatureSpawnEvent;
+import org.bukkit.event.entity.EntityTransformEvent;
 
 public class ZombieVillagerEntity extends ZombieEntity implements IVillagerDataHolder {
-   private static final DataParameter<Boolean> DATA_CONVERTING_ID = EntityDataManager.defineId(ZombieVillagerEntity.class, DataSerializers.BOOLEAN);
+   public static final DataParameter<Boolean> DATA_CONVERTING_ID = EntityDataManager.defineId(ZombieVillagerEntity.class, DataSerializers.BOOLEAN);
    private static final DataParameter<VillagerData> DATA_VILLAGER_DATA = EntityDataManager.defineId(ZombieVillagerEntity.class, DataSerializers.VILLAGER_DATA);
-   private int villagerConversionTime;
-   private UUID conversionStarter;
+   public int villagerConversionTime;
+   public UUID conversionStarter;
    private INBT gossips;
    private CompoundNBT tradeOffers;
    private int villagerXp;
+   private int lastTick = MinecraftServer.currentTick;
 
    public ZombieVillagerEntity(EntityType<? extends ZombieVillagerEntity> p_i50186_1_, World p_i50186_2_) {
       super(p_i50186_1_, p_i50186_2_);
@@ -115,13 +_,19 @@
    public void tick() {
       if (!this.level.isClientSide && this.isAlive() && this.isConverting()) {
          int i = this.getConversionProgress();
+         // CraftBukkit start - Use wall time instead of ticks for villager conversion
+         int elapsedTicks = MinecraftServer.currentTick - this.lastTick;
+         i *= elapsedTicks;
+         // CraftBukkit end
+
          this.villagerConversionTime -= i;
-         if (this.villagerConversionTime <= 0) {
+         if (this.villagerConversionTime <= 0 && net.minecraftforge.event.ForgeEventFactory.canLivingConvert(this, EntityType.VILLAGER, (timer) -> this.villagerConversionTime = timer)) {
             this.finishConversion((ServerWorld)this.level);
          }
       }
 
       super.tick();
+      this.lastTick = MinecraftServer.currentTick;
    }
 
    public ActionResultType mobInteract(PlayerEntity p_230254_1_, Hand p_230254_2_) {
@@ -157,12 +_,15 @@
       return this.getEntityData().get(DATA_CONVERTING_ID);
    }
 
-   private void startConverting(@Nullable UUID p_191991_1_, int p_191991_2_) {
+   public void startConverting(@Nullable UUID p_191991_1_, int p_191991_2_) {
       this.conversionStarter = p_191991_1_;
       this.villagerConversionTime = p_191991_2_;
       this.getEntityData().set(DATA_CONVERTING_ID, true);
-      this.removeEffect(Effects.WEAKNESS);
-      this.addEffect(new EffectInstance(Effects.DAMAGE_BOOST, p_191991_2_, Math.min(this.level.getDifficulty().getId() - 1, 0)));
+      // CraftBukkit start
+      this.persistenceRequired = true; // CraftBukkit - SPIGOT-4684 update persistence
+      this.removeEffect(Effects.WEAKNESS, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+      this.addEffect(new EffectInstance(Effects.DAMAGE_BOOST, p_191991_2_, Math.min(this.level.getDifficulty().getId() - 1, 0)), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION);
+      // CraftBukkit end
       this.level.broadcastEntityEvent(this, (byte)16);
    }
 
@@ -179,7 +_,13 @@
    }
 
    private void finishConversion(ServerWorld p_213791_1_) {
-      VillagerEntity villagerentity = this.convertTo(EntityType.VILLAGER, false);
+      // CraftBukkit start
+      VillagerEntity villagerentity = this.convertTo(EntityType.VILLAGER, false,  EntityTransformEvent.TransformReason.CURED, CreatureSpawnEvent.SpawnReason.CURED);
+      if (villagerentity == null) {
+         ((ZombieVillager) getBukkitEntity()).setConversionTime(-1); // SPIGOT-5208: End conversion to stop event spam
+         return;
+      }
+      // CraftBukkit end
 
       for(EquipmentSlotType equipmentslottype : EquipmentSlotType.values()) {
          ItemStack itemstack = this.getItemBySlot(equipmentslottype);
@@ -189,7 +_,9 @@
             } else {
                double d0 = (double)this.getEquipmentDropChance(equipmentslottype);
                if (d0 > 1.0D) {
+                  this.forceDrops = true; // CraftBukkit
                   this.spawnAtLocation(itemstack);
+                  this.forceDrops = false; // CraftBukkit
                }
             }
          }
@@ -214,11 +_,11 @@
          }
       }
 
-      villagerentity.addEffect(new EffectInstance(Effects.CONFUSION, 200, 0));
+      villagerentity.addEffect(new EffectInstance(Effects.CONFUSION, 200, 0), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.CONVERSION); // CraftBukkit
       if (!this.isSilent()) {
          p_213791_1_.levelEvent((PlayerEntity)null, 1027, this.blockPosition(), 0);
       }
-
+      net.minecraftforge.event.ForgeEventFactory.onLivingConvert(this, villagerentity);
    }
 
    private int getConversionProgress() {
