--- a/net/minecraft/entity/ai/brain/task/FarmTask.java
+++ b/net/minecraft/entity/ai/brain/task/FarmTask.java
@@ -22,7 +_,6 @@
 import net.minecraft.util.SoundEvents;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.BlockPosWrapper;
-import net.minecraft.world.GameRules;
 import net.minecraft.world.server.ServerWorld;
 
 public class FarmTask extends Task<VillagerEntity> {
@@ -37,7 +_,7 @@
    }
 
    protected boolean checkExtraStartConditions(ServerWorld p_212832_1_, VillagerEntity p_212832_2_) {
-      if (!p_212832_1_.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+      if (!net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(p_212832_1_, p_212832_2_)) {
          return false;
       } else if (p_212832_2_.getVillagerData().getProfession() != VillagerProfession.FARMER) {
          return false;
@@ -95,7 +_,11 @@
             Block block = blockstate.getBlock();
             Block block1 = p_212833_1_.getBlockState(this.aboveFarmlandPos.below()).getBlock();
             if (block instanceof CropsBlock && ((CropsBlock)block).isMaxAge(blockstate)) {
-               p_212833_1_.destroyBlock(this.aboveFarmlandPos, true, p_212833_2_);
+               // CraftBukkit start
+               if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityChangeBlockEvent(p_212833_2_, this.aboveFarmlandPos, Blocks.AIR.defaultBlockState()).isCancelled()) {
+                  p_212833_1_.destroyBlock(this.aboveFarmlandPos, true, p_212833_2_);
+               }
+               // CraftBukkit end
             }
 
             if (blockstate.isAir() && block1 instanceof FarmlandBlock && p_212833_2_.hasFarmSeeds()) {
@@ -105,19 +_,33 @@
                   ItemStack itemstack = inventory.getItem(i);
                   boolean flag = false;
                   if (!itemstack.isEmpty()) {
+                     // CraftBukkit start
+                     Block planted = null;
                      if (itemstack.getItem() == Items.WHEAT_SEEDS) {
-                        p_212833_1_.setBlock(this.aboveFarmlandPos, Blocks.WHEAT.defaultBlockState(), 3);
+                        planted = Blocks.WHEAT;
                         flag = true;
                      } else if (itemstack.getItem() == Items.POTATO) {
-                        p_212833_1_.setBlock(this.aboveFarmlandPos, Blocks.POTATOES.defaultBlockState(), 3);
+                        planted = Blocks.POTATOES;
                         flag = true;
                      } else if (itemstack.getItem() == Items.CARROT) {
-                        p_212833_1_.setBlock(this.aboveFarmlandPos, Blocks.CARROTS.defaultBlockState(), 3);
+                        planted = Blocks.CARROTS;
                         flag = true;
                      } else if (itemstack.getItem() == Items.BEETROOT_SEEDS) {
-                        p_212833_1_.setBlock(this.aboveFarmlandPos, Blocks.BEETROOTS.defaultBlockState(), 3);
+                        planted = Blocks.BEETROOTS;
                         flag = true;
-                     }
+                     } else if (itemstack.getItem() instanceof net.minecraftforge.common.IPlantable) {
+                        if (((net.minecraftforge.common.IPlantable)itemstack.getItem()).getPlantType(p_212833_1_, aboveFarmlandPos) == net.minecraftforge.common.PlantType.CROP) {
+                           p_212833_1_.setBlock(aboveFarmlandPos, ((net.minecraftforge.common.IPlantable)itemstack.getItem()).getPlant(p_212833_1_, aboveFarmlandPos), 3);
+                           flag = true;
+                        }
+                     }
+
+                     if (planted != null && !org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityChangeBlockEvent(p_212833_2_, this.aboveFarmlandPos, planted.defaultBlockState()).isCancelled()) {
+                        p_212833_1_.setBlock(this.aboveFarmlandPos, planted.defaultBlockState(), 3);
+                     } else {
+                        flag = false;
+                     }
+                     // CraftBukkit end
                   }
 
                   if (flag) {
