--- a/net/minecraft/world/biome/BiomeContainer.java
+++ b/net/minecraft/world/biome/BiomeContainer.java
@@ -17,7 +_,7 @@
    public static final int BIOMES_SIZE = 1 << WIDTH_BITS + WIDTH_BITS + HEIGHT_BITS;
    public static final int HORIZONTAL_MASK = (1 << WIDTH_BITS) - 1;
    public static final int VERTICAL_MASK = (1 << HEIGHT_BITS) - 1;
-   private final IObjectIntIterable<Biome> biomeRegistry;
+   public final IObjectIntIterable<Biome> biomeRegistry; // PAIL
    private final Biome[] biomes;
 
    public BiomeContainer(IObjectIntIterable<Biome> p_i241971_1_, Biome[] p_i241971_2_) {
@@ -101,4 +_,13 @@
       int k = p_225526_3_ & HORIZONTAL_MASK;
       return this.biomes[j << WIDTH_BITS + WIDTH_BITS | k << WIDTH_BITS | i];
    }
+
+   // CraftBukkit start
+   public void setBiome(int x, int y, int z, Biome biome) {
+      int i = x & HORIZONTAL_MASK;
+      int j = MathHelper.clamp(y, 0, VERTICAL_MASK);
+      int k = z & HORIZONTAL_MASK;
+      this.biomes[j << WIDTH_BITS + WIDTH_BITS | k << WIDTH_BITS | i] = biome;
+   }
+   // CraftBukkit end
 }
