--- a/net/minecraft/world/biome/MobSpawnInfo.java
+++ b/net/minecraft/world/biome/MobSpawnInfo.java
@@ -42,23 +_,35 @@
    private final Map<EntityClassification, List<MobSpawnInfo.Spawners>> spawners;
    private final Map<EntityType<?>, MobSpawnInfo.SpawnCosts> mobSpawnCosts;
    private final boolean playerSpawnFriendly;
+   private final java.util.Set<EntityClassification> typesView;
+   private final java.util.Set<EntityType<?>> costView;
 
    private MobSpawnInfo(float p_i241946_1_, Map<EntityClassification, List<MobSpawnInfo.Spawners>> p_i241946_2_, Map<EntityType<?>, MobSpawnInfo.SpawnCosts> p_i241946_3_, boolean p_i241946_4_) {
       this.creatureGenerationProbability = p_i241946_1_;
       this.spawners = p_i241946_2_;
       this.mobSpawnCosts = p_i241946_3_;
       this.playerSpawnFriendly = p_i241946_4_;
+      this.typesView = java.util.Collections.unmodifiableSet(this.spawners.keySet());
+      this.costView = java.util.Collections.unmodifiableSet(this.mobSpawnCosts.keySet());
    }
 
    public List<MobSpawnInfo.Spawners> getMobs(EntityClassification p_242559_1_) {
       return this.spawners.getOrDefault(p_242559_1_, ImmutableList.of());
    }
 
+   public java.util.Set<EntityClassification> getSpawnerTypes() {
+       return this.typesView;
+   }
+
    @Nullable
    public MobSpawnInfo.SpawnCosts getMobSpawnCost(EntityType<?> p_242558_1_) {
       return this.mobSpawnCosts.get(p_242558_1_);
    }
 
+   public java.util.Set<EntityType<?>> getEntityTypes() {
+       return this.costView;
+   }
+
    public float getCreatureProbability() {
       return this.creatureGenerationProbability;
    }
@@ -148,8 +_,15 @@
       public Spawners(EntityType<?> p_i241950_1_, int p_i241950_2_, int p_i241950_3_, int p_i241950_4_) {
          super(p_i241950_2_);
          this.type = p_i241950_1_.getCategory() == EntityClassification.MISC ? EntityType.PIG : p_i241950_1_;
-         this.minCount = p_i241950_3_;
-         this.maxCount = p_i241950_4_;
+         // Some mods confused minCount and maxCount (like MDE's MC Dungeons mod), so make a check here should be fine
+         if (p_i241950_3_ > p_i241950_4_) {
+            LOGGER.catching(new IllegalArgumentException("Some mod registered a MobSpawner with minCount=" + p_i241950_3_ + " and maxCount=" + p_i241950_4_ + ", auto swapped"));
+            this.maxCount = p_i241950_3_;
+            this.minCount = p_i241950_4_;
+         } else {
+            this.minCount = p_i241950_3_;
+            this.maxCount = p_i241950_4_;
+         }
       }
 
       public String toString() {
