--- a/net/minecraft/network/datasync/EntityDataManager.java
+++ b/net/minecraft/network/datasync/EntityDataManager.java
@@ -15,12 +_,15 @@
 import net.minecraft.crash.CrashReportCategory;
 import net.minecraft.crash.ReportedException;
 import net.minecraft.entity.Entity;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.network.PacketBuffer;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.commons.lang3.ObjectUtils;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer;
 
 public class EntityDataManager {
    private static final Logger LOGGER = LogManager.getLogger();
@@ -36,11 +_,13 @@
    }
 
    public static <T> DataParameter<T> defineId(Class<? extends Entity> p_187226_0_, IDataSerializer<T> p_187226_1_) {
-      if (LOGGER.isDebugEnabled()) {
+      if (true || LOGGER.isDebugEnabled()) { // Forge: This is very useful for mods that register keys on classes that are not their own
          try {
             Class<?> oclass = Class.forName(Thread.currentThread().getStackTrace()[2].getClassName());
             if (!oclass.equals(p_187226_0_)) {
-               LOGGER.debug("defineId called for: {} from {}", p_187226_0_, oclass, new RuntimeException());
+               // Forge: log at warn, mods should not add to classes that they don't own, and only add stacktrace when in debug is enabled as it is mostly not needed and consumes time
+               if (LOGGER.isDebugEnabled()) LOGGER.warn("defineId called for: {} from {}", p_187226_0_, oclass, new RuntimeException());
+               else LOGGER.warn("defineId called for: {} from {}", p_187226_0_, oclass);
             }
          } catch (ClassNotFoundException classnotfoundexception) {
          }
@@ -123,8 +_,18 @@
          dataentry.setDirty(true);
          this.isDirty = true;
       }
+      if (p_187227_1_ == LivingEntity.DATA_HEALTH_ID && entity instanceof ServerPlayerEntity && ((ServerPlayerEntity) entity).initialized()){
+         CraftPlayer player = ((ServerPlayerEntity) this.entity).getBukkitEntity();
+         player.setRealHealth(((Float) p_187227_2_));
+      }
+   }
 
+   // CraftBukkit start - add method from above
+   public <T> void markDirty(DataParameter<T> datawatcherobject) {
+      this.getItem(datawatcherobject).setDirty(true);
+      this.isDirty = true;
    }
+   // CraftBukkit end
 
    public boolean isDirty() {
       return this.isDirty;
