--- a/net/minecraft/server/management/UserList.java
+++ b/net/minecraft/server/management/UserList.java
@@ -50,7 +_,7 @@
    @Nullable
    public V get(K p_152683_1_) {
       this.removeExpired();
-      return this.map.get(this.getKeyForUser(p_152683_1_));
+      return (V)(this.map.get(this.getKeyForUser(p_152683_1_)));
    }
 
    public void remove(K p_152684_1_) {
@@ -72,6 +_,12 @@
       return this.map.keySet().toArray(new String[this.map.size()]);
    }
 
+   // CraftBukkit start
+   public Collection<V> getValues() {
+      return this.map.values();
+   }
+   // CraftBukkit end
+
    public boolean isEmpty() {
       return this.map.size() < 1;
    }
@@ -130,6 +_,13 @@
                   this.map.put(this.getKeyForUser(userlistentry.getUser()), (V)userlistentry);
                }
             }
+            // Spigot Start
+         } catch ( com.google.gson.JsonParseException | NullPointerException ex ) {
+            org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING, "Unable to read file " + this.file + ", backing it up to {0}.backup and creating new copy.", ex);
+            File backup = new File(this.file + ".backup");
+            this.file.renameTo(backup);
+            this.file.delete();
+            // Spigot End
          }
 
       }
