--- a/net/minecraft/entity/AgeableEntity.java
+++ b/net/minecraft/entity/AgeableEntity.java
@@ -16,11 +_,36 @@
    protected int age;
    protected int forcedAge;
    protected int forcedAgeTimer;
+   public boolean ageLocked; // CraftBukkit
 
    protected AgeableEntity(EntityType<? extends AgeableEntity> p_i48581_1_, World p_i48581_2_) {
       super(p_i48581_1_, p_i48581_2_);
    }
 
+   // Spigot start
+   @Override
+   public void inactiveTick()
+   {
+      super.inactiveTick();
+      if ( this.level.isClientSide || this.ageLocked )
+      { // CraftBukkit
+         this.refreshDimensions();
+      } else
+      {
+         int i = this.getAge();
+         if ( i < 0 )
+         {
+            ++i;
+            this.setAge( i );
+         } else if ( i > 0 )
+         {
+            --i;
+            this.setAge( i );
+         }
+      }
+   }
+   // Spigot end
+
    public ILivingEntityData finalizeSpawn(IServerWorld p_213386_1_, DifficultyInstance p_213386_2_, SpawnReason p_213386_3_, @Nullable ILivingEntityData p_213386_4_, @Nullable CompoundNBT p_213386_5_) {
       if (p_213386_4_ == null) {
          p_213386_4_ = new AgeableEntity.AgeableData(true);
@@ -95,12 +_,14 @@
       super.addAdditionalSaveData(p_213281_1_);
       p_213281_1_.putInt("Age", this.getAge());
       p_213281_1_.putInt("ForcedAge", this.forcedAge);
+      p_213281_1_.putBoolean("AgeLocked", this.ageLocked); // CraftBukkit
    }
 
    public void readAdditionalSaveData(CompoundNBT p_70037_1_) {
       super.readAdditionalSaveData(p_70037_1_);
       this.setAge(p_70037_1_.getInt("Age"));
       this.forcedAge = p_70037_1_.getInt("ForcedAge");
+      this.ageLocked = p_70037_1_.getBoolean("AgeLocked"); // CraftBukkit
    }
 
    public void onSyncedDataUpdated(DataParameter<?> p_184206_1_) {
@@ -113,7 +_,7 @@
 
    public void aiStep() {
       super.aiStep();
-      if (this.level.isClientSide) {
+      if (this.level.isClientSide || ageLocked) { // CraftBukkit
          if (this.forcedAgeTimer > 0) {
             if (this.forcedAgeTimer % 4 == 0) {
                this.level.addParticle(ParticleTypes.HAPPY_VILLAGER, this.getRandomX(1.0D), this.getRandomY() + 0.5D, this.getRandomZ(1.0D), 0.0D, 0.0D, 0.0D);
