--- a/net/minecraft/entity/BoostHelper.java
+++ b/net/minecraft/entity/BoostHelper.java
@@ -37,6 +_,15 @@
       }
    }
 
+   // CraftBukkit add setBoostTicks(int)
+   public void setBoostTicks(int ticks) {
+      this.boosting = true;
+      this.boostTime = 0;
+      this.boostTimeTotal = ticks;
+      this.entityData.set(this.boostTimeAccessor, this.boostTimeTotal);
+   }
+   // CraftBukkit end
+
    public void addAdditionalSaveData(CompoundNBT p_233618_1_) {
       p_233618_1_.putBoolean("Saddle", this.hasSaddle());
    }
