--- a/net/minecraft/entity/CreatureEntity.java
+++ b/net/minecraft/entity/CreatureEntity.java
@@ -7,6 +_,8 @@
 import net.minecraft.world.IWorld;
 import net.minecraft.world.IWorldReader;
 import net.minecraft.world.World;
+import org.bukkit.event.entity.EntityUnleashEvent;
+import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
 
 public abstract class CreatureEntity extends MobEntity {
    protected CreatureEntity(EntityType<? extends CreatureEntity> p_i48575_1_, World p_i48575_2_) {
@@ -37,6 +_,7 @@
          float f = this.distanceTo(entity);
          if (this instanceof TameableEntity && ((TameableEntity)this).isInSittingPose()) {
             if (f > 10.0F) {
+               this.level.getCBServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.DISTANCE)); // CraftBukkit
                this.dropLeash(true, true);
             }
 
@@ -45,6 +_,7 @@
 
          this.onLeashDistance(f);
          if (f > 10.0F) {
+            this.level.getCBServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.DISTANCE)); // CraftBukkit
             this.dropLeash(true, true);
             this.goalSelector.disableControlFlag(Goal.Flag.MOVE);
          } else if (f > 6.0F) {
