--- a/net/minecraft/entity/IAngerable.java
+++ b/net/minecraft/entity/IAngerable.java
@@ -104,7 +_,7 @@
    default void stopBeingAngry() {
       this.setLastHurtByMob((LivingEntity)null);
       this.setPersistentAngerTarget((UUID)null);
-      this.setTarget((LivingEntity)null);
+      this.setGoalTarget((LivingEntity)null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FORGOT_TARGET, true); // CraftBukkit
       this.setRemainingPersistentAngerTime(0);
    }
 
@@ -113,6 +_,8 @@
    void setLastHurtByPlayer(@Nullable PlayerEntity p_230246_1_);
 
    void setTarget(@Nullable LivingEntity p_70624_1_);
+
+   boolean setGoalTarget(LivingEntity entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent); // CraftBukkit
 
    @Nullable
    LivingEntity getTarget();
