--- a/net/minecraft/network/play/client/CClientSettingsPacket.java
+++ b/net/minecraft/network/play/client/CClientSettingsPacket.java
@@ -10,8 +_,8 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class CClientSettingsPacket implements IPacket<IServerPlayNetHandler> {
-   private String language;
-   private int viewDistance;
+   public String language;
+   public int viewDistance;
    private ChatVisibility chatVisibility;
    private boolean chatColors;
    private int modelCustomisation;
@@ -66,5 +_,9 @@
 
    public HandSide getMainHand() {
       return this.mainHand;
+   }
+
+   public String getLanguage() {
+      return this.language;
    }
 }
