--- a/net/minecraft/network/play/client/CCustomPayloadPacket.java
+++ b/net/minecraft/network/play/client/CCustomPayloadPacket.java
@@ -1,5 +_,6 @@
 package net.minecraft.network.play.client;
 
+import catserver.server.CatServer;
 import io.netty.buffer.ByteBuf;
 import java.io.IOException;
 import net.minecraft.network.IPacket;
@@ -9,10 +_,10 @@
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 
-public class CCustomPayloadPacket implements IPacket<IServerPlayNetHandler> {
+public class CCustomPayloadPacket implements IPacket<IServerPlayNetHandler>, net.minecraftforge.fml.network.ICustomPacket<CCustomPayloadPacket> {
    public static final ResourceLocation BRAND = new ResourceLocation("brand");
-   private ResourceLocation identifier;
-   private PacketBuffer data;
+   public ResourceLocation identifier;
+   public PacketBuffer data;
 
    public CCustomPayloadPacket() {
    }
@@ -35,10 +_,17 @@
 
    public void write(PacketBuffer p_148840_1_) throws IOException {
       p_148840_1_.writeResourceLocation(this.identifier);
-      p_148840_1_.writeBytes((ByteBuf)this.data);
+      p_148840_1_.writeBytes((ByteBuf)this.data.copy()); //This may be access multiple times, from multiple threads, lets be safe like the S->C packet
    }
 
    public void handle(IServerPlayNetHandler p_148833_1_) {
+      // CatServer start
+      if ("FML|HS".equalsIgnoreCase(identifier.getPath()) || "FML|MP".equalsIgnoreCase(identifier.getPath())) {
+         CatServer.log.warn("Unsupported Forge client or proxy detected! (Player: {} Channel: {})", ((net.minecraft.network.play.ServerPlayNetHandler)p_148833_1_).player.getName().getString(), identifier.getPath());
+         if (this.data != null) this.data.release();
+         return;
+      }
+      // CatServer end
       p_148833_1_.handleCustomPayload(this);
       if (this.data != null) {
          this.data.release();
