--- a/net/minecraft/network/play/client/CPlayerPacket.java
+++ b/net/minecraft/network/play/client/CPlayerPacket.java
@@ -8,14 +_,14 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class CPlayerPacket implements IPacket<IServerPlayNetHandler> {
-   protected double x;
-   protected double y;
-   protected double z;
-   protected float yRot;
-   protected float xRot;
-   protected boolean onGround;
-   protected boolean hasPos;
-   protected boolean hasRot;
+   public double x;
+   public double y;
+   public double z;
+   public float yRot;
+   public float xRot;
+   public boolean onGround;
+   public boolean hasPos;
+   public boolean hasRot;
 
    public CPlayerPacket() {
    }
