--- a/net/minecraft/client/renderer/BlockModelShapes.java
+++ b/net/minecraft/client/renderer/BlockModelShapes.java
@@ -14,6 +_,7 @@
 import net.minecraft.util.registry.Registry;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import net.minecraftforge.client.model.data.EmptyModelData;
 
 @OnlyIn(Dist.CLIENT)
 public class BlockModelShapes {
@@ -24,8 +_,15 @@
       this.modelManager = p_i46245_1_;
    }
 
+   @Deprecated
    public TextureAtlasSprite getParticleIcon(BlockState p_178122_1_) {
-      return this.getBlockModel(p_178122_1_).getParticleIcon();
+      return this.getBlockModel(p_178122_1_).getParticleTexture(net.minecraftforge.client.model.data.EmptyModelData.INSTANCE);
+   }
+
+   public TextureAtlasSprite getTexture(BlockState state, net.minecraft.world.World world, net.minecraft.util.math.BlockPos pos) {
+      net.minecraftforge.client.model.data.IModelData data = net.minecraftforge.client.model.ModelDataManager.getModelData(world, pos);
+      IBakedModel model = this.getBlockModel(state);
+      return model.getParticleTexture(model.getModelData(world, pos, state, data == null ? EmptyModelData.INSTANCE : data));
    }
 
    public IBakedModel getBlockModel(BlockState p_178125_1_) {
