--- a/net/minecraft/entity/item/minecart/CommandBlockMinecartEntity.java
+++ b/net/minecraft/entity/item/minecart/CommandBlockMinecartEntity.java
@@ -19,9 +_,11 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.v1_16_R3.entity.CraftMinecartCommand;
 
 public class CommandBlockMinecartEntity extends AbstractMinecartEntity {
-   private static final DataParameter<String> DATA_ID_COMMAND_NAME = EntityDataManager.defineId(CommandBlockMinecartEntity.class, DataSerializers.STRING);
+   public static final DataParameter<String> DATA_ID_COMMAND_NAME = EntityDataManager.defineId(CommandBlockMinecartEntity.class, DataSerializers.STRING); // private->public CraftBukkit
    private static final DataParameter<ITextComponent> DATA_ID_LAST_OUTPUT = EntityDataManager.defineId(CommandBlockMinecartEntity.class, DataSerializers.COMPONENT);
    private final CommandBlockLogic commandBlock = new CommandBlockMinecartEntity.MinecartCommandLogic();
    private int lastActivated;
@@ -73,6 +_,8 @@
    }
 
    public ActionResultType interact(PlayerEntity p_184230_1_, Hand p_184230_2_) {
+      ActionResultType ret = super.interact(p_184230_1_, p_184230_2_);
+      if (ret.consumesAction()) return ret;
       return this.commandBlock.usedBy(p_184230_1_);
    }
 
@@ -116,5 +_,12 @@
       public CommandSource createCommandSourceStack() {
          return new CommandSource(this, CommandBlockMinecartEntity.this.position(), CommandBlockMinecartEntity.this.getRotationVector(), this.getLevel(), 2, this.getName().getString(), CommandBlockMinecartEntity.this.getDisplayName(), this.getLevel().getServer(), CommandBlockMinecartEntity.this);
       }
+
+      // CraftBukkit start
+      @Override
+      public CommandSender getBukkitSender(CommandSource p0) {
+         return (CraftMinecartCommand) CommandBlockMinecartEntity.this.getBukkitEntity();
+      }
+      // CraftBukkit end
    }
 }
