--- a/net/minecraft/util/text/Color.java
+++ b/net/minecraft/util/text/Color.java
@@ -11,7 +_,7 @@
 
 public final class Color {
    private static final Map<TextFormatting, Color> LEGACY_FORMAT_TO_COLOR = Stream.of(TextFormatting.values()).filter(TextFormatting::isColor).collect(ImmutableMap.toImmutableMap(Function.identity(), (p_240748_0_) -> {
-      return new Color(p_240748_0_.getColor(), p_240748_0_.getName());
+      return new Color(p_240748_0_.getColor(), p_240748_0_.getName(), p_240748_0_); // CraftBukkit
    }));
    private static final Map<String, Color> NAMED_COLORS = LEGACY_FORMAT_TO_COLOR.values().stream().collect(ImmutableMap.toImmutableMap((p_240746_0_) -> {
       return p_240746_0_.name;
@@ -20,15 +_,30 @@
    @Nullable
    private final String name;
 
+   // CraftBukkit start
+   @Nullable
+   public final TextFormatting format;
+
+   private Color(int color, String name, TextFormatting format) {
+      this.value = color;
+      this.name = name;
+      this.format = format;
+   }
+
+   // LoliServer start
    private Color(int p_i232573_1_, String p_i232573_2_) {
       this.value = p_i232573_1_;
       this.name = p_i232573_2_;
+      this.format = null;
    }
+   // LoliServer end
 
    private Color(int p_i232572_1_) {
       this.value = p_i232572_1_;
       this.name = null;
+      this.format = null;
    }
+   // CraftBukkit end
 
    @OnlyIn(Dist.CLIENT)
    public int getValue() {
