--- a/net/minecraft/network/play/server/SChatPacket.java
+++ b/net/minecraft/network/play/server/SChatPacket.java
@@ -12,6 +_,7 @@
 
 public class SChatPacket implements IPacket<IClientPlayNetHandler> {
    private ITextComponent message;
+   public net.md_5.bungee.api.chat.BaseComponent[] components; // Spigot
    private ChatType type;
    private UUID sender;
 
@@ -31,7 +_,13 @@
    }
 
    public void write(PacketBuffer p_148840_1_) throws IOException {
-      p_148840_1_.writeComponent(this.message);
+      // Spigot start
+      if (components != null) {
+         p_148840_1_.writeUtf(net.md_5.bungee.chat.ComponentSerializer.toString(components));
+      } else {
+         p_148840_1_.writeComponent(this.message);
+      }
+      // Spigot end
       p_148840_1_.writeByte(this.type.getIndex());
       p_148840_1_.writeUUID(this.sender);
    }
