--- a/net/minecraft/network/play/server/SMultiBlockChangePacket.java
+++ b/net/minecraft/network/play/server/SMultiBlockChangePacket.java
@@ -5,6 +_,7 @@
 import java.util.function.BiConsumer;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockState;
+import net.minecraft.block.Blocks;
 import net.minecraft.client.network.play.IClientPlayNetHandler;
 import net.minecraft.network.IPacket;
 import net.minecraft.network.PacketBuffer;
@@ -31,7 +_,7 @@
 
       for(short short1 : p_i242085_2_) {
          this.positions[i] = short1;
-         this.states[i] = p_i242085_3_.getBlockState(SectionPos.sectionRelativeX(short1), SectionPos.sectionRelativeY(short1), SectionPos.sectionRelativeZ(short1));
+         this.states[i] = (p_i242085_3_ != null) ? p_i242085_3_.getBlockState(SectionPos.sectionRelativeX(short1), SectionPos.sectionRelativeY(short1), SectionPos.sectionRelativeZ(short1)): Blocks.AIR.defaultBlockState(); // CraftBukkit - SPIGOT-6076, Mojang bug when empty chunk section notified
          ++i;
       }
 
