--- a/net/minecraft/network/play/server/SPlayerListHeaderFooterPacket.java
+++ b/net/minecraft/network/play/server/SPlayerListHeaderFooterPacket.java
@@ -9,8 +_,8 @@
 import net.minecraftforge.api.distmarker.OnlyIn;
 
 public class SPlayerListHeaderFooterPacket implements IPacket<IClientPlayNetHandler> {
-   private ITextComponent header;
-   private ITextComponent footer;
+   public ITextComponent header;
+   public ITextComponent footer;
 
    public void read(PacketBuffer p_148837_1_) throws IOException {
       this.header = p_148837_1_.readComponent();
