--- a/net/minecraft/network/play/server/SWorldBorderPacket.java
+++ b/net/minecraft/network/play/server/SWorldBorderPacket.java
@@ -24,8 +_,10 @@
 
    public SWorldBorderPacket(WorldBorder p_i46921_1_, SWorldBorderPacket.Action p_i46921_2_) {
       this.type = p_i46921_2_;
-      this.newCenterX = p_i46921_1_.getCenterX();
-      this.newCenterZ = p_i46921_1_.getCenterZ();
+      // CraftBukkit start - multiply out nether border
+      this.newCenterX = p_i46921_1_.getCenterX() * p_i46921_1_.world.dimensionType().coordinateScale();
+      this.newCenterZ = p_i46921_1_.getCenterZ() * p_i46921_1_.world.dimensionType().coordinateScale();
+      // CraftBukkit end
       this.oldSize = p_i46921_1_.getSize();
       this.newSize = p_i46921_1_.getLerpTarget();
       this.lerpTime = p_i46921_1_.getLerpRemainingTime();
