--- a/net/minecraft/dispenser/ShulkerBoxDispenseBehavior.java
+++ b/net/minecraft/dispenser/ShulkerBoxDispenseBehavior.java
@@ -7,6 +_,8 @@
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.Direction;
 import net.minecraft.util.math.BlockPos;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.block.BlockDispenseEvent;
 
 public class ShulkerBoxDispenseBehavior extends OptionalDispenseBehavior {
    protected ItemStack execute(IBlockSource p_82487_1_, ItemStack p_82487_2_) {
@@ -16,6 +_,32 @@
          Direction direction = p_82487_1_.getBlockState().getValue(DispenserBlock.FACING);
          BlockPos blockpos = p_82487_1_.getPos().relative(direction);
          Direction direction1 = p_82487_1_.getLevel().isEmptyBlock(blockpos.below()) ? direction : Direction.UP;
+
+
+         // CraftBukkit start
+         org.bukkit.block.Block bukkitBlock = p_82487_1_.getLevel().getWorld().getBlockAt(p_82487_1_.getPos().getX(), p_82487_1_.getPos().getY(), p_82487_1_.getPos().getZ());
+         CraftItemStack craftItem = CraftItemStack.asCraftMirror(p_82487_2_);
+
+         BlockDispenseEvent event = new BlockDispenseEvent(bukkitBlock, craftItem.clone(), new org.bukkit.util.Vector(blockpos.getX(), blockpos.getY(), blockpos.getZ()));
+         if (!DispenserBlock.eventFired) {
+            p_82487_1_.getLevel().getCBServer().getPluginManager().callEvent(event);
+         }
+
+         if (event.isCancelled()) {
+            return p_82487_2_;
+         }
+
+         if (!event.getItem().equals(craftItem)) {
+            // Chain to handler for new item
+            ItemStack eventStack = CraftItemStack.asNMSCopy(event.getItem());
+            IDispenseItemBehavior idispensebehavior = (IDispenseItemBehavior) DispenserBlock.DISPENSER_REGISTRY.get(eventStack);
+            if (idispensebehavior != IDispenseItemBehavior.NOOP && idispensebehavior != this) {
+               idispensebehavior.dispense(p_82487_1_, eventStack);
+               return p_82487_2_;
+            }
+         }
+         // CraftBukkit end
+
          this.setSuccess(((BlockItem)item).place(new DirectionalPlaceContext(p_82487_1_.getLevel(), blockpos, direction, p_82487_2_, direction1)).consumesAction());
       }
 
