--- a/net/minecraft/resources/DataPackRegistries.java
+++ b/net/minecraft/resources/DataPackRegistries.java
@@ -15,7 +_,7 @@
 public class DataPackRegistries implements AutoCloseable {
    private static final CompletableFuture<Unit> DATA_RELOAD_INITIAL_TASK = CompletableFuture.completedFuture(Unit.INSTANCE);
    private final IReloadableResourceManager resources = new SimpleReloadableResourceManager(ResourcePackType.SERVER_DATA);
-   private final Commands commands;
+   public Commands commands;
    private final RecipeManager recipes = new RecipeManager();
    private final NetworkTagManager tagManager = new NetworkTagManager();
    private final LootPredicateManager predicateManager = new LootPredicateManager();
@@ -32,6 +_,7 @@
       this.resources.registerReloadListener(this.lootTables);
       this.resources.registerReloadListener(this.functionLibrary);
       this.resources.registerReloadListener(this.advancements);
+      net.minecraftforge.event.ForgeEventFactory.onResourceReload(this).forEach(resources::registerReloadListener);
    }
 
    public FunctionReloader getFunctionLibrary() {
@@ -81,6 +_,7 @@
 
    public void updateGlobals() {
       this.tagManager.getTags().bindToGlobal();
+      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.TagsUpdatedEvent.CustomTagTypes(tagManager.getTags()));
    }
 
    public void close() {
