--- a/net/minecraft/entity/passive/BatEntity.java
+++ b/net/minecraft/entity/passive/BatEntity.java
@@ -27,6 +_,7 @@
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class BatEntity extends AmbientEntity {
    private static final DataParameter<Byte> DATA_ID_FLAGS = EntityDataManager.defineId(BatEntity.class, DataSerializers.BYTE);
@@ -115,16 +_,24 @@
             }
 
             if (this.level.getNearestPlayer(BAT_RESTING_TARGETING, this) != null) {
+               // CraftBukkit Start - Call BatToggleSleepEvent
+               if (CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
+                  this.setResting(false);
+                  if (!flag) {
+                     this.level.levelEvent((PlayerEntity) null, 1025, blockpos, 0);
+                  }
+               }
+               // CraftBukkit end
+            }
+         } else {
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if(CraftEventFactory.handleBatToggleSleepEvent(this, true)) {
                this.setResting(false);
                if (!flag) {
-                  this.level.levelEvent((PlayerEntity)null, 1025, blockpos, 0);
+                  this.level.levelEvent((PlayerEntity) null, 1025, blockpos, 0);
                }
             }
-         } else {
-            this.setResting(false);
-            if (!flag) {
-               this.level.levelEvent((PlayerEntity)null, 1025, blockpos, 0);
-            }
+            // CraftBukkit End - Call BatToggleSleepEvent
          }
       } else {
          if (this.targetPosition != null && (!this.level.isEmptyBlock(this.targetPosition) || this.targetPosition.getY() < 1)) {
@@ -146,7 +_,11 @@
          this.zza = 0.5F;
          this.yRot += f1;
          if (this.random.nextInt(100) == 0 && this.level.getBlockState(blockpos1).isRedstoneConductor(this.level, blockpos1)) {
+            // CraftBukkit Start - Call BatToggleSleepEvent
+            if(CraftEventFactory.handleBatToggleSleepEvent(this, false)){
             this.setResting(true);
+         }
+            // CraftBukkit End
          }
       }
 
