--- a/net/minecraft/entity/passive/CatEntity.java
+++ b/net/minecraft/entity/passive/CatEntity.java
@@ -67,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class CatEntity extends TameableEntity {
    private static final Ingredient TEMPT_INGREDIENT = Ingredient.of(Items.COD, Items.SALMON);
@@ -402,7 +_,7 @@
             }
          } else if (this.isFood(itemstack)) {
             this.usePlayerItem(p_230254_1_, itemstack);
-            if (this.random.nextInt(3) == 0) {
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) { // CraftBukkit
                this.tame(p_230254_1_);
                this.setOrderedToSit(true);
                this.level.broadcastEntityEvent(this, (byte)7);
@@ -551,7 +_,13 @@
          LootContext.Builder lootcontext$builder = (new LootContext.Builder((ServerWorld)this.cat.level)).withParameter(LootParameters.ORIGIN, this.cat.position()).withParameter(LootParameters.THIS_ENTITY, this.cat).withRandom(random);
 
          for(ItemStack itemstack : loottable.getRandomItems(lootcontext$builder.create(LootParameterSets.GIFT))) {
-            this.cat.level.addFreshEntity(new ItemEntity(this.cat.level, (double)blockpos$mutable.getX() - (double)MathHelper.sin(this.cat.yBodyRot * ((float)Math.PI / 180F)), (double)blockpos$mutable.getY(), (double)blockpos$mutable.getZ() + (double)MathHelper.cos(this.cat.yBodyRot * ((float)Math.PI / 180F)), itemstack));
+            ItemEntity itemEntity = new ItemEntity(this.cat.level, (double)blockpos$mutable.getX() - (double)MathHelper.sin(this.cat.yBodyRot * ((float)Math.PI / 180F)), (double)blockpos$mutable.getY(), (double)blockpos$mutable.getZ() + (double)MathHelper.cos(this.cat.yBodyRot * ((float)Math.PI / 180F)), itemstack);
+            org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(this.cat.getBukkitEntity(), (org.bukkit.entity.Item) itemEntity.getBukkitEntity());
+            itemEntity.level.getCBServer().getPluginManager().callEvent(event);
+            if (event.isCancelled()) {
+               continue;
+            }
+            this.cat.level.addFreshEntity(itemEntity);
          }
 
       }
@@ -579,7 +_,7 @@
 
    static class TemptGoal extends net.minecraft.entity.ai.goal.TemptGoal {
       @Nullable
-      private PlayerEntity selectedPlayer;
+      private LivingEntity selectedPlayer; // CraftBukkit
       private final CatEntity cat;
 
       public TemptGoal(CatEntity p_i50438_1_, double p_i50438_2_, Ingredient p_i50438_4_, boolean p_i50438_5_) {
