--- a/net/minecraft/entity/passive/ChickenEntity.java
+++ b/net/minecraft/entity/passive/ChickenEntity.java
@@ -68,6 +_,11 @@
    }
 
    public void aiStep() {
+      // CraftBukkit start
+      if (this.isChickenJockey()) {
+         this.persistenceRequired = !this.removeWhenFarAway(0);
+      }
+      // CraftBukkit end
       super.aiStep();
       this.oFlap = this.flap;
       this.oFlapSpeed = this.flapSpeed;
@@ -86,7 +_,9 @@
       this.flap += this.flapping * 2.0F;
       if (!this.level.isClientSide && this.isAlive() && !this.isBaby() && !this.isChickenJockey() && --this.eggTime <= 0) {
          this.playSound(SoundEvents.CHICKEN_EGG, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
+         this.forceDrops = true; // CraftBukkit
          this.spawnAtLocation(Items.EGG);
+         this.forceDrops = false; // CraftBukkit
          this.eggTime = this.random.nextInt(6000) + 6000;
       }
 
