--- a/net/minecraft/entity/passive/CowEntity.java
+++ b/net/minecraft/entity/passive/CowEntity.java
@@ -29,6 +_,9 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraft.world.server.ServerWorld;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack;
+import org.bukkit.event.player.PlayerBucketFillEvent;
 
 public class CowEntity extends AnimalEntity {
    public CowEntity(EntityType<? extends CowEntity> p_i48567_1_, World p_i48567_2_) {
@@ -73,8 +_,14 @@
    public ActionResultType mobInteract(PlayerEntity p_230254_1_, Hand p_230254_2_) {
       ItemStack itemstack = p_230254_1_.getItemInHand(p_230254_2_);
       if (itemstack.getItem() == Items.BUCKET && !this.isBaby()) {
+         // CraftBukkit start - Got milk?
+         PlayerBucketFillEvent event = CraftEventFactory.callPlayerBucketFillEvent(p_230254_1_.level, p_230254_1_, this.blockPosition(), this.blockPosition(), null, itemstack, Items.MILK_BUCKET);
+         if (event.isCancelled()) {
+            return ActionResultType.PASS;
+         }
+         // CraftBukkit end
          p_230254_1_.playSound(SoundEvents.COW_MILK, 1.0F, 1.0F);
-         ItemStack itemstack1 = DrinkHelper.createFilledResult(itemstack, p_230254_1_, Items.MILK_BUCKET.getDefaultInstance());
+         ItemStack itemstack1 = DrinkHelper.createFilledResult(itemstack, p_230254_1_, CraftItemStack.asNMSCopy(event.getItemStack()));
          p_230254_1_.setItemInHand(p_230254_2_, itemstack1);
          return ActionResultType.sidedSuccess(this.level.isClientSide);
       } else {
