--- a/net/minecraft/entity/passive/DolphinEntity.java
+++ b/net/minecraft/entity/passive/DolphinEntity.java
@@ -69,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class DolphinEntity extends WaterMobEntity {
    private static final DataParameter<BlockPos> TREASURE_POS = EntityDataManager.defineId(DolphinEntity.class, DataSerializers.BLOCK_POS);
@@ -221,6 +_,12 @@
       if (this.getItemBySlot(EquipmentSlotType.MAINHAND).isEmpty()) {
          ItemStack itemstack = p_175445_1_.getItem();
          if (this.canHoldItem(itemstack)) {
+            // CraftBukkit start - call EntityPickupItemEvent
+            if(CraftEventFactory.callEntityPickupItemEvent(this,p_175445_1_, 0, false).isCancelled()){
+               return;
+            }
+            itemstack = p_175445_1_.getItem(); // update ItemStack from event
+            // CraftBukkit end
             this.onItemPickup(p_175445_1_);
             this.setItemSlot(EquipmentSlotType.MAINHAND, itemstack);
             this.handDropChances[EquipmentSlotType.MAINHAND.getIndex()] = 2.0F;
@@ -485,7 +_,7 @@
       }
 
       public boolean canUse() {
-         return this.dolphin.gotFish() && this.dolphin.getAirSupply() >= 100;
+         return this.dolphin.gotFish() && this.dolphin.getAirSupply() >= 100 && this.dolphin.level.getWorld().canGenerateStructures(); // MC-151364, SPIGOT-5494: hangs if generate-structures=false
       }
 
       public boolean canContinueToUse() {
@@ -582,7 +_,7 @@
       }
 
       public void start() {
-         this.player.addEffect(new EffectInstance(Effects.DOLPHINS_GRACE, 100));
+         this.player.addEffect(new EffectInstance(Effects.DOLPHINS_GRACE, 100), Cause.DOLPHIN); // CraftBukkit - reason
       }
 
       public void stop() {
@@ -599,7 +_,7 @@
          }
 
          if (this.player.isSwimming() && this.player.level.random.nextInt(6) == 0) {
-            this.player.addEffect(new EffectInstance(Effects.DOLPHINS_GRACE, 100));
+            this.player.addEffect(new EffectInstance(Effects.DOLPHINS_GRACE, 100), Cause.DOLPHIN); // CraftBukkit - reason
          }
 
       }
