--- a/net/minecraft/entity/passive/FoxEntity.java
+++ b/net/minecraft/entity/passive/FoxEntity.java
@@ -82,12 +_,13 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.CreatureSpawnEvent;
 
 public class FoxEntity extends AnimalEntity {
    private static final DataParameter<Integer> DATA_TYPE_ID = EntityDataManager.defineId(FoxEntity.class, DataSerializers.INT);
    private static final DataParameter<Byte> DATA_FLAGS_ID = EntityDataManager.defineId(FoxEntity.class, DataSerializers.BYTE);
-   private static final DataParameter<Optional<UUID>> DATA_TRUSTED_ID_0 = EntityDataManager.defineId(FoxEntity.class, DataSerializers.OPTIONAL_UUID);
-   private static final DataParameter<Optional<UUID>> DATA_TRUSTED_ID_1 = EntityDataManager.defineId(FoxEntity.class, DataSerializers.OPTIONAL_UUID);
+   public static final DataParameter<Optional<UUID>> DATA_TRUSTED_ID_0 = EntityDataManager.defineId(FoxEntity.class, DataSerializers.OPTIONAL_UUID);
+   public static final DataParameter<Optional<UUID>> DATA_TRUSTED_ID_1 = EntityDataManager.defineId(FoxEntity.class, DataSerializers.OPTIONAL_UUID);
    private static final Predicate<ItemEntity> ALLOWED_ITEMS = (p_213489_0_) -> {
       return !p_213489_0_.hasPickUpDelay() && p_213489_0_.isAlive();
    };
@@ -325,7 +_,7 @@
       return FoxEntity.Type.byId(this.entityData.get(DATA_TYPE_ID));
    }
 
-   private void setFoxType(FoxEntity.Type p_213474_1_) {
+   public void setFoxType(FoxEntity.Type p_213474_1_) {
       this.entityData.set(DATA_TYPE_ID, p_213474_1_.getId());
    }
 
@@ -409,7 +_,7 @@
       return this.getFlag(32);
    }
 
-   private void setSleeping(boolean p_213485_1_) {
+   public void setSleeping(boolean p_213485_1_) {
       this.setFlag(32, p_213485_1_);
    }
 
@@ -458,7 +_,8 @@
 
    protected void pickUpItem(ItemEntity p_175445_1_) {
       ItemStack itemstack = p_175445_1_.getItem();
-      if (this.canHoldItem(itemstack)) {
+      if (!org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callEntityPickupItemEvent(this, p_175445_1_, itemstack.getCount() - 1, !this.canHoldItem(itemstack)).isCancelled()) { // CraftBukkit - call EntityPickupItemEvent
+         itemstack = p_175445_1_.getItem(); // CraftBukkit - update ItemStack from event
          int i = itemstack.getCount();
          if (i > 1) {
             this.dropItemStack(itemstack.split(i - 1));
@@ -761,7 +_,7 @@
       }
 
       protected void onReachedTarget() {
-         if (FoxEntity.this.level.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) {
+         if (net.minecraftforge.event.ForgeEventFactory.getMobGriefingEvent(FoxEntity.this.level, FoxEntity.this)) {
             BlockState blockstate = FoxEntity.this.level.getBlockState(this.blockPos);
             if (blockstate.is(Blocks.SWEET_BERRY_BUSH)) {
                int i = blockstate.getValue(SweetBerryBushBlock.AGE);
@@ -1001,6 +_,17 @@
       protected void breed() {
          ServerWorld serverworld = (ServerWorld)this.level;
          FoxEntity foxentity = (FoxEntity)this.animal.getBreedOffspring(serverworld, this.partner);
+         final net.minecraftforge.event.entity.living.BabyEntitySpawnEvent event = new net.minecraftforge.event.entity.living.BabyEntitySpawnEvent(animal, partner, foxentity);
+         final boolean cancelled = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(event);
+         foxentity = (FoxEntity) event.getChild();
+         if (cancelled) {
+            //Reset the "inLove" state for the animals
+            this.animal.setAge(6000);
+            this.partner.setAge(6000);
+            this.animal.resetLove();
+            this.partner.resetLove();
+            return;
+         }
          if (foxentity != null) {
             ServerPlayerEntity serverplayerentity = this.animal.getLoveCause();
             ServerPlayerEntity serverplayerentity1 = this.partner.getLoveCause();
@@ -1024,9 +_,10 @@
             this.partner.setAge(6000);
             this.animal.resetLove();
             this.partner.resetLove();
-            foxentity.setAge(-24000);
-            foxentity.moveTo(this.animal.getX(), this.animal.getY(), this.animal.getZ(), 0.0F, 0.0F);
+            serverworld.dummyFreshEntitySpawn = true;
             serverworld.addFreshEntityWithPassengers(foxentity);
+            serverworld.dummyFreshEntitySpawn = false;
+            serverworld.addAllEntities(foxentity, CreatureSpawnEvent.SpawnReason.BREEDING);// CraftBukkit - added SpawnReason
             this.level.broadcastEntityEvent(this.animal, (byte)18);
             if (this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBLOOT)) {
                this.level.addFreshEntity(new ExperienceOrbEntity(this.level, this.animal.getX(), this.animal.getY(), this.animal.getZ(), this.animal.getRandom().nextInt(7) + 1));
