--- a/net/minecraft/entity/passive/OcelotEntity.java
+++ b/net/minecraft/entity/passive/OcelotEntity.java
@@ -48,6 +_,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
 
 public class OcelotEntity extends AnimalEntity {
    private static final Ingredient TEMPT_INGREDIENT = Ingredient.of(Items.COD, Items.SALMON);
@@ -118,7 +_,7 @@
    }
 
    public boolean removeWhenFarAway(double p_213397_1_) {
-      return !this.isTrusting() && this.tickCount > 2400;
+      return !this.isTrusting() /*&& this.ticksExisted > 2400; */; // CraftBukkit
    }
 
    public static AttributeModifierMap.MutableAttribute createAttributes() {
@@ -163,7 +_,8 @@
       if ((this.temptGoal == null || this.temptGoal.isRunning()) && !this.isTrusting() && this.isFood(itemstack) && p_230254_1_.distanceToSqr(this) < 9.0D) {
          this.usePlayerItem(p_230254_1_, itemstack);
          if (!this.level.isClientSide) {
-            if (this.random.nextInt(3) == 0) {
+            // CraftBukkit - added event call and isCancelled check
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) {
                this.setTrusting(true);
                this.spawnTrustingParticles(true);
                this.level.broadcastEntityEvent(this, (byte)41);
