--- a/net/minecraft/entity/passive/PandaEntity.java
+++ b/net/minecraft/entity/passive/PandaEntity.java
@@ -61,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class PandaEntity extends AnimalEntity {
    private static final DataParameter<Integer> UNHAPPY_COUNTER = EntityDataManager.defineId(PandaEntity.class, DataSerializers.INT);
@@ -492,7 +_,7 @@
    }
 
    protected void pickUpItem(ItemEntity p_175445_1_) {
-      if (this.getItemBySlot(EquipmentSlotType.MAINHAND).isEmpty() && PANDA_ITEMS.test(p_175445_1_)) {
+      if(!CraftEventFactory.callEntityPickupItemEvent(this, p_175445_1_, 0, !(this.getItemBySlot(EquipmentSlotType.MAINHAND).isEmpty() && PANDA_ITEMS.test(p_175445_1_))).isCancelled()) {  // CraftBukkit
          this.onItemPickup(p_175445_1_);
          ItemStack itemstack = p_175445_1_.getItem();
          this.setItemSlot(EquipmentSlotType.MAINHAND, itemstack);
@@ -928,7 +_,7 @@
 
       protected void alertOther(MobEntity p_220793_1_, LivingEntity p_220793_2_) {
          if (p_220793_1_ instanceof PandaEntity && ((PandaEntity)p_220793_1_).isAggressive()) {
-            p_220793_1_.setTarget(p_220793_2_);
+            p_220793_1_.setGoalTarget(p_220793_2_, TargetReason.TARGET_ATTACKED_ENTITY, true); // CraftBukkit
          }
 
       }
