--- a/net/minecraft/entity/passive/ParrotEntity.java
+++ b/net/minecraft/entity/passive/ParrotEntity.java
@@ -62,6 +_,8 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ParrotEntity extends ShoulderRidingEntity implements IFlyingAnimal {
    private static final DataParameter<Integer> DATA_VARIANT_ID = EntityDataManager.defineId(ParrotEntity.class, DataSerializers.INT);
@@ -237,7 +_,7 @@
          }
 
          if (!this.level.isClientSide) {
-            if (this.random.nextInt(10) == 0) {
+            if (this.random.nextInt(10) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) { // CraftBukkit
                this.tame(p_230254_1_);
                this.level.broadcastEntityEvent(this, (byte)7);
             } else {
@@ -251,7 +_,7 @@
             itemstack.shrink(1);
          }
 
-         this.addEffect(new EffectInstance(Effects.POISON, 900));
+         this.addEffect(new EffectInstance(Effects.POISON, 900), Cause.FOOD); // CraftBukkit
          if (p_230254_1_.isCreative() || !this.isInvulnerable()) {
             this.hurt(DamageSource.playerAttack(p_230254_1_), Float.MAX_VALUE);
          }
@@ -349,7 +_,7 @@
    }
 
    public boolean isPushable() {
-      return true;
+      return super.isPushable(); // CraftBukkit - collidable API
    }
 
    protected void doPush(Entity p_82167_1_) {
@@ -362,7 +_,7 @@
       if (this.isInvulnerableTo(p_70097_1_)) {
          return false;
       } else {
-         this.setOrderedToSit(false);
+         //this.setOrderedToSit(false); // CraftBukkit - moved into EntityLiving.damageEntity(DamageSource, float)
          return super.hurt(p_70097_1_, p_70097_2_);
       }
    }
