--- a/net/minecraft/entity/passive/SheepEntity.java
+++ b/net/minecraft/entity/passive/SheepEntity.java
@@ -29,6 +_,7 @@
 import net.minecraft.entity.ai.goal.WaterAvoidingRandomWalkingGoal;
 import net.minecraft.entity.item.ItemEntity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.inventory.CraftResultInventory;
 import net.minecraft.inventory.CraftingInventory;
 import net.minecraft.inventory.container.Container;
 import net.minecraft.inventory.container.ContainerType;
@@ -60,8 +_,10 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.SheepRegrowWoolEvent;
+import org.bukkit.inventory.InventoryView;
 
-public class SheepEntity extends AnimalEntity implements IShearable {
+public class SheepEntity extends AnimalEntity implements IShearable, net.minecraftforge.common.IForgeShearable {
    private static final DataParameter<Byte> DATA_WOOL_ID = EntityDataManager.defineId(SheepEntity.class, DataSerializers.BYTE);
    private static final Map<DyeColor, IItemProvider> ITEM_BY_DYE = Util.make(Maps.newEnumMap(DyeColor.class), (p_203402_0_) -> {
       p_203402_0_.put(DyeColor.WHITE, Blocks.WHITE_WOOL);
@@ -216,7 +_,7 @@
 
    public ActionResultType mobInteract(PlayerEntity p_230254_1_, Hand p_230254_2_) {
       ItemStack itemstack = p_230254_1_.getItemInHand(p_230254_2_);
-      if (itemstack.getItem() == Items.SHEARS) {
+      if (false && itemstack.getItem() == Items.SHEARS) { //Forge: Moved to onSheared
          if (!this.level.isClientSide && this.readyForShearing()) {
             this.shear(SoundCategory.PLAYERS);
             itemstack.hurtAndBreak(1, p_230254_1_, (p_213613_1_) -> {
@@ -237,7 +_,9 @@
       int i = 1 + this.random.nextInt(3);
 
       for(int j = 0; j < i; ++j) {
+         this.forceDrops = true; // CraftBukkit
          ItemEntity itementity = this.spawnAtLocation(ITEM_BY_DYE.get(this.getColor()), 1);
+         this.forceDrops = false; // CraftBukkit
          if (itementity != null) {
             itementity.setDeltaMovement(itementity.getDeltaMovement().add((double)((this.random.nextFloat() - this.random.nextFloat()) * 0.1F), (double)(this.random.nextFloat() * 0.05F), (double)((this.random.nextFloat() - this.random.nextFloat()) * 0.1F)));
          }
@@ -323,6 +_,11 @@
    }
 
    public void ate() {
+      // CraftBukkit start
+      SheepRegrowWoolEvent event = new SheepRegrowWoolEvent((org.bukkit.entity.Sheep) this.getBukkitEntity());
+      this.level.getCBServer().getPluginManager().callEvent(event);
+      if (event.isCancelled()) return;
+      // CraftBukkit end
       this.setSheared(false);
       if (this.isBaby()) {
          this.ageUp(60);
@@ -349,16 +_,46 @@
 
    private static CraftingInventory makeContainer(DyeColor p_213611_0_, DyeColor p_213611_1_) {
       CraftingInventory craftinginventory = new CraftingInventory(new Container((ContainerType)null, -1) {
+         // CraftBukkit start
+         @Override
+         public InventoryView getBukkitView() {
+            return null;
+         }
+         // CraftBukkit end
+
          public boolean stillValid(PlayerEntity p_75145_1_) {
             return false;
          }
       }, 2, 1);
       craftinginventory.setItem(0, new ItemStack(DyeItem.byColor(p_213611_0_)));
       craftinginventory.setItem(1, new ItemStack(DyeItem.byColor(p_213611_1_)));
+      craftinginventory.resultInventory = new CraftResultInventory(); // CraftBukkit - add result slot for event
       return craftinginventory;
    }
 
    protected float getStandingEyeHeight(Pose p_213348_1_, EntitySize p_213348_2_) {
       return 0.95F * p_213348_2_.height;
+   }
+
+   @Override
+   public boolean isShearable(@javax.annotation.Nonnull ItemStack item, World world, BlockPos pos) {
+      return readyForShearing();
+   }
+
+   @javax.annotation.Nonnull
+   @Override
+   public java.util.List<ItemStack> onSheared(@Nullable PlayerEntity player, @javax.annotation.Nonnull ItemStack item, World world, BlockPos pos, int fortune) {
+      world.playSound(null, this, SoundEvents.SHEEP_SHEAR, player == null ? SoundCategory.BLOCKS : SoundCategory.PLAYERS, 1.0F, 1.0F);
+      if (!world.isClientSide) {
+         this.setSheared(true);
+         int i = 1 + this.random.nextInt(3);
+
+         java.util.List<ItemStack> items = new java.util.ArrayList<>();
+         for (int j = 0; j < i; ++j) {
+            items.add(new ItemStack(ITEM_BY_DYE.get(this.getColor())));
+         }
+         return items;
+      }
+      return java.util.Collections.emptyList();
    }
 }
