--- a/net/minecraft/entity/passive/SquidEntity.java
+++ b/net/minecraft/entity/passive/SquidEntity.java
@@ -178,7 +_,7 @@
    }
 
    public static boolean checkSquidSpawnRules(EntityType<SquidEntity> p_223365_0_, IWorld p_223365_1_, SpawnReason p_223365_2_, BlockPos p_223365_3_, Random p_223365_4_) {
-      return p_223365_3_.getY() > 45 && p_223365_3_.getY() < p_223365_1_.getSeaLevel();
+      return p_223365_3_.getY() > p_223365_1_.getMinecraftWorld().spigotConfig.squidSpawnRangeMin && p_223365_3_.getY() < p_223365_1_.getSeaLevel(); // Spigot
    }
 
    @OnlyIn(Dist.CLIENT)
