--- a/net/minecraft/entity/passive/StriderEntity.java
+++ b/net/minecraft/entity/passive/StriderEntity.java
@@ -75,7 +_,7 @@
    private static final DataParameter<Integer> DATA_BOOST_TIME = EntityDataManager.defineId(StriderEntity.class, DataSerializers.INT);
    private static final DataParameter<Boolean> DATA_SUFFOCATING = EntityDataManager.defineId(StriderEntity.class, DataSerializers.BOOLEAN);
    private static final DataParameter<Boolean> DATA_SADDLE_ID = EntityDataManager.defineId(StriderEntity.class, DataSerializers.BOOLEAN);
-   private final BoostHelper steering = new BoostHelper(this.entityData, DATA_BOOST_TIME, DATA_SADDLE_ID);
+   public final BoostHelper steering = new BoostHelper(this.entityData, DATA_BOOST_TIME, DATA_SADDLE_ID);
    private TemptGoal temptGoal;
    private PanicGoal panicGoal;
 
@@ -274,7 +_,12 @@
       BlockState blockstate = this.level.getBlockState(this.blockPosition());
       BlockState blockstate1 = this.getBlockStateOn();
       boolean flag = blockstate.is(BlockTags.STRIDER_WARM_BLOCKS) || blockstate1.is(BlockTags.STRIDER_WARM_BLOCKS) || this.getFluidHeight(FluidTags.LAVA) > 0.0D;
-      this.setSuffocating(!flag);
+      // CraftBukkit start
+      if (!flag ^ this.isSuffocating()) {
+         org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, !flag);
+         this.setSuffocating(!flag);
+      }
+      // CraftBukkit end
       super.tick();
       this.floatStrider();
       this.checkInsideBlocks();
