--- a/net/minecraft/entity/passive/WolfEntity.java
+++ b/net/minecraft/entity/passive/WolfEntity.java
@@ -61,6 +_,9 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.v1_16_R3.event.CraftEventFactory;
+import org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason;
+import org.bukkit.event.entity.EntityTargetEvent.TargetReason;
 
 public class WolfEntity extends TameableEntity implements IAngerable {
    private static final DataParameter<Boolean> DATA_INTERESTED_ID = EntityDataManager.defineId(WolfEntity.class, DataSerializers.BOOLEAN);
@@ -110,6 +_,24 @@
       return MobEntity.createMobAttributes().add(Attributes.MOVEMENT_SPEED, (double)0.3F).add(Attributes.MAX_HEALTH, 8.0D).add(Attributes.ATTACK_DAMAGE, 2.0D);
    }
 
+   // CraftBukkit - add overriden version
+   @Override
+   public boolean setGoalTarget(LivingEntity entitylivingbaseIn, TargetReason reason, boolean fireEvent) {
+      if (!super.setGoalTarget(entitylivingbaseIn, reason, fireEvent)) {
+         return false;
+      }
+      entitylivingbaseIn = getTarget();
+      /*
+      if (entitylivingbaseIn == null) {
+         this.setAngry(false);
+      } else if (!this.isTamed()) {
+         this.setAngry(true);
+      }
+       */
+      return true;
+   }
+   // CraftBukkit end
+
    protected void defineSynchedData() {
       super.defineSynchedData();
       this.entityData.define(DATA_INTERESTED_ID, false);
@@ -133,7 +_,8 @@
          this.setCollarColor(DyeColor.byId(p_70037_1_.getInt("CollarColor")));
       }
 
-      this.readPersistentAngerSaveData((ServerWorld)this.level, p_70037_1_);
+      if(!level.isClientSide) //FORGE: allow this entity to be read from nbt on client. (Fixes MC-189565)
+         this.readPersistentAngerSaveData((ServerWorld)this.level, p_70037_1_);
    }
 
    protected SoundEvent getAmbientSound() {
@@ -273,7 +_,7 @@
          return false;
       } else {
          Entity entity = p_70097_1_.getEntity();
-         this.setOrderedToSit(false);
+         // this.setOrderedToSit(false);   // CraftBukkit - moved into EntityLiving.damageEntity(DamageSource, float)
          if (entity != null && !(entity instanceof PlayerEntity) && !(entity instanceof AbstractArrowEntity)) {
             p_70097_2_ = (p_70097_2_ + 1.0F) / 2.0F;
          }
@@ -295,7 +_,7 @@
       super.setTame(p_70903_1_);
       if (p_70903_1_) {
          this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(20.0D);
-         this.setHealth(20.0F);
+         this.setHealth(this.getMaxHealth()); // CraftBukkit - 20.0 -> getMaxHealth()
       } else {
          this.getAttribute(Attributes.MAX_HEALTH).setBaseValue(8.0D);
       }
@@ -316,7 +_,7 @@
                   itemstack.shrink(1);
                }
 
-               this.heal((float)item.getFoodProperties().getNutrition());
+               this.heal((float)item.getFoodProperties().getNutrition(), RegainReason.EATING); // CraftBukkit
                return ActionResultType.SUCCESS;
             }
 
@@ -326,7 +_,7 @@
                   this.setOrderedToSit(!this.isOrderedToSit());
                   this.jumping = false;
                   this.navigation.stop();
-                  this.setTarget((LivingEntity)null);
+                  this.setGoalTarget((LivingEntity)null, TargetReason.FORGOT_TARGET, true); // CraftBukkit - reason
                   return ActionResultType.SUCCESS;
                }
 
@@ -347,7 +_,8 @@
                itemstack.shrink(1);
             }
 
-            if (this.random.nextInt(3) == 0) {
+            // CraftBukkit - added event call and isCancelled check.
+            if (this.random.nextInt(3) == 0 && !net.minecraftforge.event.ForgeEventFactory.onAnimalTame(this, p_230254_1_) && !CraftEventFactory.callEntityTameEvent(this, p_230254_1_).isCancelled()) {
                this.tame(p_230254_1_);
                this.navigation.stop();
                this.setTarget((LivingEntity)null);
