--- a/net/minecraft/inventory/EnderChestInventory.java
+++ b/net/minecraft/inventory/EnderChestInventory.java
@@ -5,12 +_,34 @@
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.ListNBT;
 import net.minecraft.tileentity.EnderChestTileEntity;
+import org.bukkit.Location;
+import org.bukkit.inventory.InventoryHolder;
 
 public class EnderChestInventory extends Inventory {
    private EnderChestTileEntity activeChest;
 
+   // CraftBukkit start
+   private final PlayerEntity owner;
+
+   // CatServer start
+   public InventoryHolder getBukkitOwner(){
+      return owner != null ? owner.getBukkitEntity() : null;
+   }
+   // CatServer end
+
+   @Override
+   public Location getLocation() {
+      return this.activeChest != null ?new Location(this.activeChest.getLevel().getWorld(), this.activeChest.getBlockPos().getX(), this.activeChest.getBlockPos().getY(), this.activeChest.getBlockPos().getZ()) : null;
+   }
+
    public EnderChestInventory() {
+      this(null); // CatServer
+   }
+
+   public EnderChestInventory(PlayerEntity owner) {
       super(27);
+      this.owner = owner;
+      // CraftBukkit end
    }
 
    public void setActiveChest(EnderChestTileEntity p_146031_1_) {
