--- a/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
+++ b/net/minecraft/entity/ai/attributes/ModifiableAttributeInstance.java
@@ -68,7 +_,7 @@
       return this.modifierById.get(p_180374_1_.getId()) != null;
    }
 
-   private void addModifier(AttributeModifier p_111121_1_) {
+   public void addModifier(AttributeModifier p_111121_1_) {
       AttributeModifier attributemodifier = this.modifierById.putIfAbsent(p_111121_1_.getId(), p_111121_1_);
       if (attributemodifier != null) {
          throw new IllegalArgumentException("Modifier is already applied on this attribute!");
