--- a/net/minecraft/entity/ai/attributes/RangedAttribute.java
+++ b/net/minecraft/entity/ai/attributes/RangedAttribute.java
@@ -4,7 +_,7 @@
 
 public class RangedAttribute extends Attribute {
    private final double minValue;
-   private final double maxValue;
+   public double maxValue; // Spigot
 
    public RangedAttribute(String p_i231504_1_, double p_i231504_2_, double p_i231504_4_, double p_i231504_6_) {
       super(p_i231504_1_, p_i231504_2_);
@@ -20,6 +_,7 @@
    }
 
    public double sanitizeValue(double p_111109_1_) {
+      if (p_111109_1_ != p_111109_1_) return getDefaultValue(); // CraftBukkit
       return MathHelper.clamp(p_111109_1_, this.minValue, this.maxValue);
    }
 }
